---
title: "Differential co-expression analysis of interactions between mRNA, miRNA, and methylation expression"
author: "Thanneer Perumal"
date: '`r date()`'
output: html_document
---
```{r knit2synapse, eval = FALSE, include=FALSE}
library(synapseClient)
library(knit2synapse)

synapseLogin()

knit2synapse::knitToFolderEntity(file = 'AnalyseDiffCoExp_All.Rmd',
                                 parentId = "syn5194922",
                                 entityName = 'Differential Coexpression Analysis Features')
```

```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
## It is assumed your working directory is where this file is

# Clear R console screen output
cat("\014")

# Load libraries
library(plyr)
library(dplyr)
library(data.table)
library(stringr)
library(tidyr)
library(reshape2)

library(circlize)
library(tools)
library(ggplot2)
library(RColorBrewer)
library(ComplexHeatmap)
library(matrixStats)
library(biomaRt)
library(fpc)
library(reshape2)
library(EBcoexpress)

library(knitr)
library(knit2synapse)
library(synapseClient)
library(rGithubClient) ## Needs the dev branch

## Requires ggdendro
devtools::source_url("https://raw.githubusercontent.com/chr1swallace/random-functions/master/R/ggplot-heatmap.R")

synapseLogin()

# source utility files from ../R/lib folder
file.sources = list.files('../R/lib',pattern="*.R", full.names = T)
tmp = sapply(file.sources,source,.GlobalEnv)

# source all discordant related files (see for installation: https://github.com/siskac/discordant/archive/master.zip)
dir = getwd()
setwd('/mnt/Github/discordant/discordant-master/')
source('/mnt/Github/discordant/discordant-master//discordant.R')
setwd(dir)

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE)
```

```{r synapseStore.params}
parentId = "syn5194922"
SYNAPSE_STORE = T  
ALL_USED_IDs = c()
```

```{r fxns}
downloadFile <- function(id){
  tmp = fread(synGet(id)@filePath, data.table=F, header=T)
}

# Function to calcualte Fishers Z
fishersZ <- function(C){
  # C is a correlation matrix
  fishersZ = 0.5 * log((1+C)/(1-C), base = exp(1));
  return(fishersZ)
}

# Function to calculate change in Fishers Z
fishersPvalue <- function(fishersZ.group1, fishersZ.group2, n.group1, n.group2){
  # Calculate change in Fishers Z between groups
  result = data.frame(changeFishersZ = abs((fishersZ.group1 - fishersZ.group2)/
                                sqrt((1/(n.group1 - 3)) + (1/(n.group2 - 3)))))
    
  # Estimate p-values for change in Fishers Z between groups
  result$pvalue = pnorm(result$changeFishersZ, lower.tail = F)
  
  # Adjust P-value using FDR
  result$fdr = p.adjust(result$pvalue, method = 'fdr')
  return(result)
}

# Function to calculate change in co-expression analysis using Fishers Z method
calculateCor <- function(metaData, omics1Mat, omics2Mat, var1 = "omics1", var2 = "omics2"){
  ## For group1
  omics1Mat = omics1Mat[,metaData$biologicalSampleName]
  omics2Mat = omics2Mat[,metaData$biologicalSampleName]
  
  # Calcualte correlation between features at group1
  correlation = WGCNA::bicor(t(omics1Mat), t(omics2Mat))  
  
  correlation = reshape2::melt(correlation) %>%
    dplyr::mutate(feature = paste(X1,X2,sep = "_")) %>%
    plyr::rename(c("X1" = var1, "X2" = var2,  "value" = unique(metaData$Diffname_short)))
    
  rownames(correlation) = correlation$feature
  writeLines(paste('Completed',unique(metaData$Diffname_short)))
  return(correlation)
}
  
# Function to calculate change in co-expression analysis using Fishers Z method
changeCoExp <- function(group, metaData, correlationMat, var1 = "omics1", var2 = "omics2", FDR = 0.05, logFC = 1.5){
  results = dplyr::select(correlationMat, one_of(c('feature',var1,var2,group)))
  
  colnames(results) = gsub(group[1], 'correlation.group1', colnames(results))
  colnames(results) = gsub(group[2], 'correlation.group2', colnames(results))

  # Calcualte Fishers Z scores from correlations
  results$fishersZ.group1 = fishersZ(results$correlation.group1)
  results$fishersZ.group2 = fishersZ(results$correlation.group2)
  
  # Estimate p-values for change in Fishers Z
  results = dplyr::bind_cols(results,
                             fishersPvalue(results$fishersZ.group1, 
                                           results$fishersZ.group2, 
                                           dim(metaData %>% dplyr::filter(Diffname_short %in% group[1]))[1],
                                           dim(metaData %>% dplyr::filter(Diffname_short %in% group[2]))[1]));
  
  results$logfc = log(abs(results$correlation.group1/results$correlation.group2))

  results = dplyr::filter(results, fdr <= FDR, abs(logfc) >= logFC)
  writeLines(paste('Completed',group[1],group[2]))
  return(results)
}

# Function to calculate discordant probabilities
discordantAnalysis <- function(group, metaData, omics1Mat, omics2Mat, var1, var2){
  omics1Mat = omics1Mat[,c(metaData[[group[1]]]$biologicalSampleName,
                           metaData[[group[2]]]$biologicalSampleName)]
  omics2Mat = omics2Mat[,c(metaData[[group[1]]]$biologicalSampleName,
                           metaData[[group[2]]]$biologicalSampleName)]
  
  groupLabels = c(rep(1, length(metaData[[group[1]]]$Diffname_short)),
            rep(2, length(metaData[[group[2]]]$Diffname_short)))
  
  # Run discordant analysis
  vectors <- createVectors(omics1Mat, omics2Mat, groups = groupLabels)
  results <- discordantRun(vectors$v1, vectors$v2, omics1Mat, omics2Mat)
  
  rownames(results$discordPPMatrix) = rownames(omics1Mat)
  colnames(results$discordPPMatrix) = rownames(omics2Mat)
  
  results$discordPPMatrix = reshape2::melt(results$discordPPMatrix) %>%
    dplyr::rename(omics1 = X1, omics2 = X2, PP = value) %>%
    dplyr::mutate(feature = paste(omics1,omics2,sep = "_"))
  results$discordPPMatrix$PP = 1 - results$discordPPMatrix$PP
  
  results$discordPPMatrix = results$discordPPMatrix %>%
    plyr::rename(c("omics1" = var1, "omics2" = var2))  
  writeLines(paste('Completed',group[1],group[2]))
  return(results$discordPPMatrix)
}
```


### Download mappings
#### Get TF-DNA mapping from Enrichr genesets
```{r tf.dna.mapping}
# Download TFs from Enrichr genesets
load(synGet('syn4867851')@filePath)
ALL_USED_IDs = c(ALL_USED_IDs,'syn4867851')

# Get unique TF - gene mapping from three data bases: ChEA, TRANSFAC&JASPAR and ENCODE
ind = grep('HUMAN', names(GeneSets$ChEA))
TFsMapping1 = mapply(function(x, y){
  TFname = str_split(x, '-')[[1]][1]
  return(list(data.frame(feature = TFname, target = y)))
}, names(GeneSets$ChEA)[ind], GeneSets$ChEA[ind]) %>%
  rbindlist %>%
  unique

ind = grep('human', names(GeneSets$TRANSFAC_and_JASPAR_PWMs))
TFsMapping2 = mapply(function(x, y){
  TFname = str_split(x, ' ')[[1]][1]
  return(list(data.frame(feature = TFname, target = y)))
}, names(GeneSets$TRANSFAC_and_JASPAR_PWMs)[ind], GeneSets$TRANSFAC_and_JASPAR_PWMs[ind]) %>%
  rbindlist %>%
  unique

ind = grep('hg19', names(GeneSets$"ENCODE_TF_ChIP-seq_2015"))
TFsMapping3 = mapply(function(x, y){
  TFname = str_split(x, '_')[[1]][1]
  return(list(data.frame(feature = TFname, target = y)))
}, names(GeneSets$"ENCODE_TF_ChIP-seq_2015")[ind], GeneSets$"ENCODE_TF_ChIP-seq_2015"[ind]) %>%
  rbindlist %>%
  unique

TFsMapping = rbindlist(list(TFsMapping1, TFsMapping2, TFsMapping3)) %>% unique             
```


### Download expression matrices and calculate median values at each diff state
#### Download mRNA counts matrix
```{r mrna.counts}
mrnaCountsId = 'syn5011095'
ALL_USED_IDs = c(ALL_USED_IDs, mrnaCountsId)
mrnaCovariatesId = 'syn5011149'
ALL_USED_IDs = c(ALL_USED_IDs, mrnaCovariatesId)

# Get count matrix
mrnaCounts = downloadFile(mrnaCountsId)
row.names(mrnaCounts) = mrnaCounts[,1]
mrnaCounts = as.matrix(mrnaCounts[,-(1)])

# Get sample weights
mrnaCovariates = downloadFile(mrnaCovariatesId)
row.names(mrnaCovariates) = mrnaCovariates[,1]
mrnaCovariates = mrnaCovariates[,-(1)]

mrnaCovariates = split(mrnaCovariates, mrnaCovariates$Diffname_short)
mrnaCounts = sapply(mrnaCovariates, function(covariates, counts){
  tmp = rowMedians(counts[, rownames(covariates)], na.rm=T)
  names(tmp) = rownames(counts)
  return(tmp)
}, mrnaCounts)

mrnaCounts = t(apply(mrnaCounts, 1, function(x){x = x/max(x, na.rm=T)}))
mrnaCounts = rownameToFirstColumn(mrnaCounts, 'mrna')
colnames(mrnaCounts)[-(1)] = paste(colnames(mrnaCounts)[-(1)],'mrna',sep='.')
```
#### Download miRNA counts matrix
```{r mirna.counts}
mirnaCountsId = 'syn5014454'
ALL_USED_IDs = c(ALL_USED_IDs, mirnaCountsId)
mirnaCovariatesId = 'syn5014460'
ALL_USED_IDs = c(ALL_USED_IDs, mirnaCovariatesId)

# Get count matrix
mirnaCounts = downloadFile(mirnaCountsId)
row.names(mirnaCounts) = mirnaCounts[,1]
mirnaCounts = as.matrix(mirnaCounts[,-(1)])

# Get sample weights
mirnaCovariates = downloadFile(mirnaCovariatesId)
row.names(mirnaCovariates) = mirnaCovariates[,1]
mirnaCovariates = mirnaCovariates[,-(1)]

mirnaCovariates = split(mirnaCovariates, mirnaCovariates$Diffname_short)
mirnaCounts = sapply(mirnaCovariates, function(covariates, counts){
  tmp = rowMedians(counts[, rownames(covariates)], na.rm=T)
  names(tmp) = rownames(counts)
  return(tmp)
}, mirnaCounts)

mirnaCounts = t(apply(mirnaCounts, 1, function(x){x = x/max(x, na.rm=T)}))
mirnaCounts = rownameToFirstColumn(mirnaCounts, 'mirna')
colnames(mirnaCounts)[-(1)] = paste(colnames(mirnaCounts)[-(1)],'mirna',sep='.')
```
#### Download methylation beta matrix
```{r methyl.counts}
methylCountsId = 'syn4487642'
ALL_USED_IDs = c(ALL_USED_IDs, methylCountsId)
methylCovariatesId = 'syn4487669'
ALL_USED_IDs = c(ALL_USED_IDs, methylCovariatesId)

# Get count matrix
methylCounts = downloadFile(methylCountsId)
row.names(methylCounts) = methylCounts[,1]
methylCounts = as.matrix(methylCounts[,-(1)])

# Get sample weights
methylCovariates = downloadFile(methylCovariatesId)
row.names(methylCovariates) = methylCovariates[,1]
methylCovariates = methylCovariates[,-(1)]

methylCovariates = split(methylCovariates, methylCovariates$Diffname_short)
methylCounts = sapply(methylCovariates, function(covariates, counts){
  tmp = rowMedians(counts[, rownames(covariates)], na.rm=T)
  names(tmp) = rownames(counts)
  return(tmp)
}, methylCounts)

methylCounts = t(apply(methylCounts, 1, function(x){x = x/max(x, na.rm=T)}))
methylCounts = rownameToFirstColumn(methylCounts, 'methyl')
colnames(methylCounts)[-(1)] = paste(colnames(methylCounts)[-(1)],'methyl',sep='.')
```


### Download differential expression results from synapse
```{r diffExp}
diffExp.ids = c(TF = 'syn5521879', nonTF = 'syn5521883',
                mirna = 'syn5521885', methyl = 'syn5521887')

diffExp = lapply(diffExp.ids, downloadFile)
```


### Download coexpression and metadata from synapse
#### Download mRNA-miRNA coexpression and metadata
```{r download.mRNA.miRNA.data}
# Get coexpression
mrna.mirna.coexp = downloadFile('syn5578560')
ALL_USED_IDs = c(ALL_USED_IDs, 'syn5578560')

TF.mirna.coexp = dplyr::filter(mrna.mirna.coexp, mrna %in% TFsMapping$feature)
nonTF.mirna.coexp = dplyr::filter(mrna.mirna.coexp, !(mrna %in% TFsMapping$feature))

# Get mRNA-miRNA metadata
mrna.mirna.metadata = downloadFile('syn5578554')
ALL_USED_IDs <- c(ALL_USED_IDs, 'syn5578554')
```
#### Download mRNA-methyl coexpression and metadata
```{r download.mRNA.methyl.data}
# Get coexpression
mrna.methyl.coexp = downloadFile('syn5578562')
ALL_USED_IDs = c(ALL_USED_IDs, 'syn5578562')

TF.methyl.coexp = dplyr::filter(mrna.methyl.coexp, mrna %in% TFsMapping$feature)
nonTF.methyl.coexp = dplyr::filter(mrna.methyl.coexp, !(mrna %in% TFsMapping$feature))

# Get mRNA-methyl metadata
mrna.methyl.metadata = downloadFile('syn5578556')
ALL_USED_IDs <- c(ALL_USED_IDs, 'syn5578556')
```
#### Download miRNA-methyl coexpression and metadata
```{r download.miRNA.methyl.data}
# Get coexpression
mirna.methyl.coexp = downloadFile('syn5578564')
ALL_USED_IDs = c(ALL_USED_IDs, 'syn5578564')

# Get mRNA-methyl metadata
mirna.methyl.metadata = downloadFile('syn5578558')
ALL_USED_IDs <- c(ALL_USED_IDs, 'syn5578558')
```


### Threshold coexpression
Coexpression matrices are raised to a power, while retaining their sign. This is performed to upweight higher correlations and down weight lower correlation values
```{r threshold}
beta = 1
TF.mirna.coexp[,-(1:3)] = sign(TF.mirna.coexp[,-(1:3)]) * (abs(TF.mirna.coexp[,-(1:3)]))^beta
TF.methyl.coexp[,-(1:3)] = sign(TF.methyl.coexp[,-(1:3)]) * (abs(TF.methyl.coexp[,-(1:3)]))^beta
mirna.methyl.coexp[,-(1:3)] = sign(mirna.methyl.coexp[,-(1:3)]) * (abs(mirna.methyl.coexp[,-(1:3)]))^beta
```


### Calculate differential co-expression analysis between TF-miRNA features at each diffstate
Performing analysis with `r dim(TF.mirna.coexp)[1]` TF-miRNAs calculated from `r dim(mrna.mirna.metadata)[1]` samples
```{r coexpp.fisher.TF.miRNA, fig.height=15, fig.width=15, echo=FALSE, results='asis'}
# Get all possible diff state comparisons
all.comparisons = t(combn(setdiff(unique(mrna.mirna.metadata$Diffname_short), c()),2))#'MESO15','MESO30'

# Perform Fisher analysis across all possible diff state comparisons
TF.mirna.fisher = apply(all.comparisons, 1, changeCoExp, 
                        mrna.mirna.metadata, TF.mirna.coexp, 
                        var1 = 'mrna', var2 = 'mirna', FDR = 0.05, logFC = 0)
names(TF.mirna.fisher) = apply(all.comparisons, 1, paste, collapse = '_vs_')

# Extract co-expression values from differential coexpression
D = mapply(function(x,y){
  group = str_split(y,'_vs_')[[1]]
  x = plyr::rename(x, c(correlation.group1 = group[1], correlation.group2 = group[2])) %>%
    dplyr::select(one_of(c('feature','mrna','mirna',group)))
}, TF.mirna.fisher, names(TF.mirna.fisher), SIMPLIFY = F) %>%
  plyr::join_all(type = "full", by = c("feature", "mrna", "mirna"))

writeLines("Number of differentially co-expressed interactions at an FDR <= 0.05 with either the TF or the miRNA are differentially expressed are")
tmp = sapply(TF.mirna.fisher, function(x, diffexp){
  mRNA = (intersect(x$mrna, diffexp[['TF']]$feature))
  miRNA = (intersect(x$mirna, diffexp[['mirna']]$feature))
  x = dplyr::filter(x, mrna %in% mRNA | mirna %in% miRNA)
  dim(x)[1]
}, diffExp, simplify = F) %>% 
  ldply(function(x){data.frame(value = x)}, .id = 'Comparison')
kable(tmp)

# Extract co-expression matrix at each differentiation stage
uniqueFeatures = lapply(TF.mirna.fisher, function(x, diffexp){
  mRNA = (intersect(x$mrna, diffexp[['TF']]$feature))
  miRNA = (intersect(x$mirna, diffexp[['mirna']]$feature))
  x = dplyr::filter(x, mrna %in% mRNA | mirna %in% miRNA) %>%
    dplyr::select(feature) %>%
    unlist %>% unique
}, diffExp) 
writeLines(paste("Unique number of differentially co-expressed interactions at an FDR <= 0.05 with either a differentially expressed TF or miRNA are", length(uniqueFeatures %>% unlist %>% unique)))

tmp = ldply(uniqueFeatures, function(x){
  data.frame(value = paste(x, collapse = ';'))
}, .id = 'Comparison')
writeLines("They are")
kable(tmp)

TF.mirna.fisher = ldply(TF.mirna.fisher, .id = "Comparison")
```

```{r cluster.TF.miRNA, fig.height=25, fig.width=15, results='asis', echo=FALSE}
tmp = D %>%
  dplyr::filter(feature %in% (uniqueFeatures %>% unlist %>% unique)) %>%
  dplyr::select(feature, SC, DE, MESO5, ECTO, MESO15, MESO30, EB)

tmp.cor = t(apply(tmp[,-(1)], 1, function(x){
  x[x>=0] = 0; 
  x[is.na(x)] = 0; 
  x = x/max(abs(x), na.rm = T)
}))
rownames(tmp.cor) = tmp$feature
tmp.cor = na.omit(tmp.cor)

writeLines(paste('Clustering',dim(tmp.cor)[1],'TF-miRNA interactions'))
clust.TF.miRNA = pamk(tmp.cor, 1:20, seed = 12345, usepam= T)
set.seed(12345); clust.TF.miRNA = kmeans(tmp.cor, 7)#clust.TF.miRNA$nc)
clust.TF.miRNA$cluster = clust.TF.miRNA$cluster %>%
  factor(levels = c(6, 7, 4, 1, 3, 2, 5))
  
# Plot raw correlation values based on processed value based clustering
tmp = D %>%
  dplyr::filter(feature %in% rownames(tmp.cor)) %>%
  dplyr::select(feature, mrna, mirna, SC, DE, MESO5, ECTO, MESO15, MESO30, EB) %>%
  left_join(mrnaCounts) %>%
  left_join(mirnaCounts)

tmp.coExp = dplyr::select(tmp, SC, DE, MESO5, ECTO, MESO15, MESO30, EB)
rownames(tmp.coExp) = tmp$feature
tmp.mrna = dplyr::select(tmp, SC.mrna, DE.mrna, MESO5.mrna, ECTO.mrna, MESO15.mrna, MESO30.mrna, EB.mrna)
rownames(tmp.mrna) = tmp$feature
tmp.mirna = dplyr::select(tmp, SC.mirna, DE.mirna, MESO5.mirna, ECTO.mirna, MESO15.mirna, MESO30.mirna, EB.mirna)   
rownames(tmp.mirna) = tmp$feature

col = rev(brewer.pal(5, "RdPu"))
figure_size = list(title_gp = gpar(fontsize = 18), labels_gp = gpar(fontsize = 14))

h.TF.miRNA = Heatmap(tmp.coExp, name = 'Corr', 
        cluster_columns = F, cluster_rows=T, split = clust.TF.miRNA$cluster, 
        # col = colorRamp2(c(-1,-0.75,-0.5,-0.25,0), col),
        column_title_gp = gpar(fontsize = 22, fontface = "bold"), 
        row_title_gp = gpar(fontsize = 22, fontface = "bold"),
        column_names_gp = gpar(fontsize = 18), combined_name_fun = NULL,
        row_names_gp = gpar(fontsize = 10),
        show_row_names = T, row_names_side = 'right', show_column_names = T, show_row_hclust = F,
        heatmap_legend_param = figure_size)
draw(h.TF.miRNA, heatmap_legend_side = "left", column_title = 'Transcription Factors-miRNA',
     column_title_gp = gpar(fontsize = 22, fontface = "bold"))
WGCNA::collectGarbage()
```

### Calculate differential co-expression analysis between TF-methylation interactions at each diffstate
Performing analysis with `r dim(TF.methyl.coexp)[1]` TF-methylation interactions calculated from `r dim(mrna.methyl.metadata)[1]` samples
```{r coexpp.fisher.TF.methyl, fig.height=15, fig.width=15, echo=FALSE, results='asis'}
# Get all possible diff state comparisons
all.comparisons = t(combn(setdiff(unique(mrna.methyl.metadata$Diffname_short), c()),2))#'MESO15','MESO30'

# Perform Fisher analysis across all possible diff state comparisons
TF.methyl.fisher = apply(all.comparisons, 1, changeCoExp, 
                         mrna.methyl.metadata, TF.methyl.coexp, 
                         var1 = 'mrna', var2 = 'methyl', FDR = 0.05, logFC = 0)
names(TF.methyl.fisher) = apply(all.comparisons, 1, paste, collapse = '_vs_')

# Extract co-expression values from differential coexpression
D = mapply(function(x,y){
  group = str_split(y,'_vs_')[[1]]
  x = plyr::rename(x, c(correlation.group1 = group[1], correlation.group2 = group[2])) %>%
    dplyr::select(one_of(c('feature','mrna','methyl',group)))
}, TF.methyl.fisher, names(TF.methyl.fisher), SIMPLIFY = F) %>%
  plyr::join_all(type = "full", by = c("feature", "mrna", "methyl"))

writeLines("Number of differentially co-expressed interactions at an FDR <= 0.05 with either the TF or the methylation probe are differentially expressed are")
tmp = sapply(TF.methyl.fisher, function(x, diffexp){
  mRNA = (intersect(x$mrna, diffexp[['TF']]$feature))
  methyl = (intersect(x$methyl, diffexp[['methyl']]$feature))
  x = dplyr::filter(x, mrna %in% mRNA | methyl %in% methyl)
  dim(x)[1]
}, diffExp, simplify = F) %>% 
  ldply(function(x){data.frame(value = x)}, .id = 'Comparison')
kable(tmp)

# Extract co-expression matrix at each differentiation stage
uniqueFeatures = lapply(TF.methyl.fisher, function(x, diffexp){
  mRNA = (intersect(x$mrna, diffexp[['TF']]$feature))
  methyl = (intersect(x$methyl, diffexp[['methyl']]$feature))
  x = dplyr::filter(x, mrna %in% mRNA | methyl %in% methyl) %>%
    dplyr::select(feature) %>%
    unlist %>% unique
}, diffExp) 
writeLines(paste("Unique number of differentially co-expressed interactions at an FDR <= 0.05 with either a differentially expressed TF or methyl probe are", length(uniqueFeatures %>% unlist %>% unique)))

tmp = ldply(uniqueFeatures, function(x){
  data.frame(value = paste(x, collapse = ';'))
}, .id = 'Comparison')
writeLines("They are")
kable(tmp)

TF.methyl.fisher = ldply(TF.methyl.fisher, .id = "Comparison")
```

```{r cluster.TF.methyl, fig.height=25, fig.width=15, results='asis', echo=FALSE}
tmp = D %>%
  dplyr::filter(feature %in% (uniqueFeatures %>% unlist %>% unique)) %>%
  dplyr::select(feature, SC, DE, MESO5, ECTO, EB)
rownames(tmp) = tmp$feature

col = rev(brewer.pal(5, "RdPu"))
figure_size = list(title_gp = gpar(fontsize = 18), labels_gp = gpar(fontsize = 14))

h.TF.methyl = Heatmap(tmp[,-(1)], name = 'Corr', 
        cluster_columns = F, cluster_rows=F,  
        column_title_gp = gpar(fontsize = 22, fontface = "bold"), 
        row_title_gp = gpar(fontsize = 22, fontface = "bold"),
        column_names_gp = gpar(fontsize = 18), combined_name_fun = NULL,
        row_names_gp = gpar(fontsize = 10),
        show_row_names = T, row_names_side = 'left', show_column_names = T, show_row_hclust = F,
        heatmap_legend_param = figure_size)
draw(h.TF.methyl, heatmap_legend_side = "right", column_title = 'Transcription Factors-methylation Probes',
     column_title_gp = gpar(fontsize = 22, fontface = "bold"))
WGCNA::collectGarbage()
```

### Calculate differential co-expression analysis between miRNA-methylation interactions at each diffstate
Performing analysis with `r dim(mirna.methyl.coexp)[1]` miRNA-methlation interactions calculated from `r dim(mirna.methyl.metadata)[1]` samples
```{r coexpp.fisher.miRNA.methyl, fig.height=15, fig.width=15, echo=FALSE, results='asis'}
# Get all possible diff state comparisons
all.comparisons = t(combn(setdiff(unique(mirna.methyl.metadata$Diffname_short), c()),2))#'MESO15','MESO30'

# Perform Fisher analysis across all possible diff state comparisons
mirna.methyl.fisher = apply(all.comparisons, 1, changeCoExp, 
                            mirna.methyl.metadata, mirna.methyl.coexp, 
                            var1 = 'mirna', var2 = 'methyl', FDR = 0.05, logFC = 0)
names(mirna.methyl.fisher) = apply(all.comparisons, 1, paste, collapse = '_vs_')

# Extract co-expression values from differential coexpression
D = mapply(function(x,y){
  group = str_split(y,'_vs_')[[1]]
  x = plyr::rename(x, c(correlation.group1 = group[1], correlation.group2 = group[2])) %>%
    dplyr::select(one_of(c('feature','mirna','methyl',group)))
}, mirna.methyl.fisher, names(mirna.methyl.fisher), SIMPLIFY = F) %>%
  plyr::join_all(type = "full", by = c("feature", "mirna", "methyl"))

writeLines("Number of differentially co-expressed interactions at an FDR <= 0.05 with either the miRNA or the methylation probe that are differentially expressed are")
tmp = sapply(mirna.methyl.fisher, function(x, diffexp){
  miRNA = (intersect(x$mrna, diffexp[['mirna']]$feature))
  methylProbe = (intersect(x$mirna, diffexp[['methyl']]$feature))
  x = dplyr::filter(x, mirna %in% miRNA | methyl %in% methylProbe)
  dim(x)[1]
}, diffExp, simplify = F) %>% 
  ldply(function(x){data.frame(value = x)}, .id = 'Comparison')
kable(tmp)

# Extract co-expression matrix at each differentiation stage
uniqueFeatures = lapply(mirna.methyl.fisher, function(x, diffexp){
  miRNA = (intersect(x$mrna, diffexp[['mirna']]$feature))
  methylProbe = (intersect(x$mirna, diffexp[['methyl']]$feature))
  x = dplyr::filter(x, mirna %in% miRNA | methyl %in% methylProbe) %>%
    dplyr::select(feature) %>%
    unlist %>% unique
}, diffExp) 
writeLines(paste("Unique number of differentially co-expressed interactions at an FDR <= 0.05 with either a differentially expressed miRNA or methylation probe are", length(uniqueFeatures %>% unlist %>% unique)))

tmp = ldply(uniqueFeatures, function(x){
  data.frame(value = paste(x, collapse = ';'))
}, .id = 'Comparison')
writeLines("They are")
kable(tmp)

mirna.methyl.fisher = ldply(mirna.methyl.fisher, .id = "Comparison")
```


### Store results in synapse
```{r synapsestore, echo=FALSE, include=FALSE, eval=TRUE, cache = F}
ActivityName <- 'Analyse integrated differential co-expression results between assays'
  
ThisFileName <- 'AnalyseDiffCoExp_All.Rmd'

# Github link
ThisRepo <- getRepo(repository = "th1vairam/pcbc_c4_analysis", 
                    ref="branch", 
                    refName='discordant_anal')

ThisFile <- getPermlink(repository = ThisRepo,
                        repositoryPath=paste0('code/Rmd/', ThisFileName))    

# Populate wiki with results
CODE <- Folder(name = 'Differential Coexpression Analysis Features',parentId = parentId)
CODE <- synStore(CODE)

# Store coexpression
changeCoExp <- list(TF_methyl = TF.methyl.fisher,
                    TF_methyl = TF.methyl.fisher,
                    mirna_methyl = mirna.methyl.fisher)
changeCoExp_obj = mapply(function(x,y, parentId, used, executed, activityName){
  write.table(x, file=paste0('changeCoExp_',y,'.tsv'), sep = '\t', quote=F, row.names=F)
  obj = File(paste0('changeCoExp_',y,'.tsv'), name = paste('Fishers Change in Coexpression',y), parentId = parentId)
  obj = synStore(obj, used = used, executed = executed, activityName = activityName)
}, changeCoExp, names(changeCoExp), MoreArgs = list(CODE$properties$id, ALL_USED_IDs, ThisFile, ActivityName),
SIMPLIFY = F)
```
### Source Code
[Source R Markdown](`r ThisFile`)