---
title: "Calculate correlation between TFs, mirna, and methylation expressions at each differentiation state"
author: "Thanneer Perumal"
date: '`r date()`'
output: html_document
---
```{r knit2synapse, eval = FALSE, include=FALSE}
library(sysnapseClient)
library(knit2synapse)

synapseLogin()

knit2synapse::knitToFolderEntity(file = 'CalculateCoExp_All.Rmd',
                                 parentId = "syn5194922",
                                 entityName = 'Coexpression Between Different Assays At Each Differentiation Stage')
```

```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
## It is assumed your working directory is where this file is

# Clear R console screen output
cat("\014")

# Load libraries
library(plyr)
library(dplyr)
library(data.table)
library(stringr)
library(tidyr)
library(reshape2)

library(circlize)
library(tools)
library(ggplot2)
library(RColorBrewer)
library(ComplexHeatmap)
library(matrixStats)
library(biomaRt)
library(fpc)
library(reshape2)
library(EBcoexpress)

library(knitr)
library(knit2synapse)
library(synapseClient)
library(rGithubClient) ## Needs the dev branch

## Requires ggdendro
devtools::source_url("https://raw.githubusercontent.com/chr1swallace/random-functions/master/R/ggplot-heatmap.R")

synapseLogin()

# source utility files from ../R/lib folder
file.sources = list.files('../R/lib',pattern="*.R", full.names = T)
tmp = sapply(file.sources,source,.GlobalEnv)

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE)
```

```{r synapseStore.params}
parentId = "syn5194922"
SYNAPSE_STORE = T  
ALL_USED_IDs = c()
```

```{r fxns}
downloadFile <- function(id){
  tmp = fread(synGet(id)@filePath, data.table=F, header=T)
}

# Function to calculate change in co-expression analysis using Fishers Z method
calculateCor <- function(metaData, omics1Mat, omics2Mat, features, var1 = "omics1", var2 = "omics2"){
  ## For group1
  omics1Mat = omics1Mat[,metaData$biologicalSampleName]
  omics2Mat = omics2Mat[,metaData$biologicalSampleName]
  
  # Calcualte correlation between features at group1
  correlation = apply(features, 1, function(x, omics1Mat, omics2Mat){
    WGCNA::bicor(t(omics1Mat[as.character(x[1]),]), t(omics2Mat[as.character(x[2]),]), use = "pairwise.complete.obs")  
  }, omics1Mat, omics2Mat)
  
  correlation = data.frame(value = correlation) %>%
    bind_cols(features) %>%
    dplyr::mutate(feature = paste(omics1,omics2,sep = "_")) %>%
    plyr::rename(c("omics1" = var1, "omics2" = var2,  "value" = unique(metaData$Diffname_short)))
  
  rownames(correlation) = correlation$feature
  writeLines(paste('Completed',unique(metaData$Diffname_short)))
  return(correlation)
}
```

### Download raw counts from synapse
#### Download mrna data
```{r download.mrna.data}
# Get metadata
mrna_metdata_id <- "syn3156503"
ALL_USED_IDs = c(ALL_USED_IDs, mrna_metdata_id)

mrna_metadata_obj <- synTableQuery(sprintf('SELECT * FROM %s', mrna_metdata_id))
mrna_metadata <- mrna_metadata_obj@values

mrna_metadata[mrna_metadata == 'N/A'] = NA

# Get mrna raw counts
mrna_id <- "syn5011097"
ALL_USED_IDs <- c(ALL_USED_IDs, mrna_id)

mrna_mat <- downloadFile(mrna_id)

# Filter mrna metadata - will use UIDs from these to subset matrices
mrna_metadata_filtered <- mrna_metadata %>%
  filter(public, pass_qc, !exclude,
         UID %in% colnames(mrna_mat),
         Diffname_short != "",
         Cell_Type == "PSC",
         C4_Karyotype_Result != "abnormal")
```
#### Download mirna data
```{r download.mirna.data}
# Get metadata
mirna_metdata_id <- "syn3219876"
ALL_USED_IDs <- c(ALL_USED_IDs, mirna_metdata_id)

mirna_metadata_obj <- synTableQuery(sprintf('SELECT * FROM %s', mirna_metdata_id))
mirna_metadata <- mirna_metadata_obj@values

mirna_metadata[mirna_metadata == 'N/A'] = NA

# Get mirna raw counts
mirna_id <- "syn5014456"
ALL_USED_IDs = c(ALL_USED_IDs, mirna_id)  

mirna_mat <- downloadFile(mirna_id)

# Filter mirna metadata - will use UIDs from these to subset matrices
mirna_metadata_filtered <- mirna_metadata %>%
  filter(public, pass_qc, !exclude,
         UID %in% colnames(mirna_mat),
         Diffname_short != "",
         Cell_Type == "PSC",
         C4_Karyotype_Result != "abnormal")
```
#### Download DNA methylation data
```{r download.methyl.data}
# Get metadata
methyl_metdata_id <- "syn3156828"; 
ALL_USED_IDs <- c(ALL_USED_IDs, methyl_metdata_id)

methyl_metadata_obj <- synTableQuery(sprintf('SELECT * FROM %s', methyl_metdata_id))
methyl_metadata <- methyl_metadata_obj@values

methyl_metadata[methyl_metadata == 'N/A'] = NA

# Get methyl raw counts
methyl_id <- "syn4487642"
ALL_USED_IDs = c(ALL_USED_IDs, methyl_id)  

methyl_mat <- downloadFile(methyl_id)

# Filter methyl metadata - will use UIDs from these to subset matrices
methyl_metadata_filtered <- methyl_metadata %>%
  filter(UID %in% colnames(methyl_mat),
         Diffname_short != "")
```


### Download comparison reference and mapping files from synapse
#### Get TF-DNA mapping from Enrichr genesets
```{r tf.dna.mapping}
# Download TFs from Enrichr genesets
load(synGet('syn4867851')@filePath)
ALL_USED_IDs = c(ALL_USED_IDs,'syn4867851')

# Get unique TF - gene mapping from three data bases: ChEA, TRANSFAC&JASPAR and ENCODE
ind = grep('HUMAN', names(GeneSets$ChEA))
TFsMapping1 = mapply(function(x, y){
  TFname = str_split(x, '-')[[1]][1]
  return(list(data.frame(feature = TFname, target = y)))
}, names(GeneSets$ChEA)[ind], GeneSets$ChEA[ind]) %>%
  rbindlist %>%
  unique

ind = grep('human', names(GeneSets$TRANSFAC_and_JASPAR_PWMs))
TFsMapping2 = mapply(function(x, y){
  TFname = str_split(x, ' ')[[1]][1]
  return(list(data.frame(feature = TFname, target = y)))
  }, names(GeneSets$TRANSFAC_and_JASPAR_PWMs)[ind], GeneSets$TRANSFAC_and_JASPAR_PWMs[ind]) %>%
  rbindlist %>%
  unique

ind = grep('hg19', names(GeneSets$"ENCODE_TF_ChIP-seq_2015"))
TFsMapping3 = mapply(function(x, y){
  TFname = str_split(x, '_')[[1]][1]
  return(list(data.frame(feature = TFname, target = y)))
  }, names(GeneSets$"ENCODE_TF_ChIP-seq_2015")[ind], GeneSets$"ENCODE_TF_ChIP-seq_2015"[ind]) %>%
  rbindlist %>%
  unique

TFsMapping = rbindlist(list(TFsMapping1, TFsMapping2, TFsMapping3)) %>% unique   
rm(list= 'GeneSets')
WGCNA::collectGarbage()
```
#### Get mRNA-miRNA mapping from synapse (Lorena version)
```{r miRNA.mRNA.mapping}
# Get miRNA mapping files
miRNA.mRNA.id = 'syn3461627'
ALL_USED_IDs <- c(ALL_USED_IDs, miRNA.mRNA.id)
miRNA.mRNA = fread(synGet(miRNA.mRNA.id)@filePath, data.table=F, header=F)
setnames(miRNA.mRNA, c('V1','V2'), c('hsa_id','ensembl_gene_id'))

# Get human related mapping
Hs = useMart("ENSEMBL_MART_ENSEMBL", host="www.ensembl.org") # use this one when biomart.org is down
Hs = useDataset("hsapiens_gene_ensembl", Hs)
human_ensg2symbol = getBM(attributes = c("ensembl_gene_id","hgnc_symbol"),
                          filters = "ensembl_gene_id",                         
                          values = unique(miRNA.mRNA$ensembl_gene_id),
                          mart = Hs)

miRNA.mRNA <- left_join(miRNA.mRNA, human_ensg2symbol, by = 'ensembl_gene_id') %>%
  dplyr::rename(feature=hsa_id, target = hgnc_symbol) %>% 
  dplyr::select(feature, target)
```
#### Get mRNA-miRNA mapping from synapse (exp. validated version)
```{r miRNA.mRNA.exp.mapping}
# Get miRNA mapping files
miRNA.mRNA2.id = 'syn5049680'
ALL_USED_IDs <- c(ALL_USED_IDs, miRNA.mRNA2.id)
miRNA.mRNA2 = fread(synGet(miRNA.mRNA2.id)@filePath, data.table=F, header=T)

miRNA.mRNA2 <- miRNA.mRNA2 %>%
  dplyr::select(one_of(c("miRNA", "Target Gene"))) %>%
  plyr::rename(c("miRNA" = "feature", "Target Gene" = "target")) %>%
  unique

miRNA.mRNA = unique(bind_rows(miRNA.mRNA, miRNA.mRNA2))
```
#### Get miRNA-methylation mapping from synapse (Lorena version)
```{r methyl.mirna.mapping}
# Get miRNA methyl mapping files
methyl.mirna.id = 'syn4895962'
ALL_USED_IDs <- c(ALL_USED_IDs, methyl.mirna.id)
methyl.mirna = fread(synGet(methyl.mirna.id)@filePath, data.table=F, header=F)

methyl.mirna <- methyl.mirna %>%  
  plyr::rename(c("V1" = "feature", "V2" = "target")) %>%
  unique
```
#### Get methylation-mrna mapping
```{r methyl.mrna.mapping}
# Get mRNA methyl mapping
methyl.mrna <- get450KProbeMapping(methyl_mat$methProbeID)$Annotation %>%  
  dplyr::select(methProbeIDs, nearestTx) %>%
  dplyr::rename(feature = methProbeIDs, target = nearestTx) %>%
  unique
```


#### Calculate co-expression between TF-TF features at each diffstate
Combine samples to unique biological sample name (take median values for replicates)
```{r combine.samples.TF.TF}
# Match up biological sample between assays
biosampleInBoth <- mrna_metadata_filtered$biologicalSampleName

# Filter metadata
mrna_metadata <- mrna_metadata_filtered 

# Take the median across multiple biological samples per feature
mrna_mat_median <- mrna_mat %>%
  dplyr::select(GeneName, one_of(mrna_metadata$UID)) %>%
  melt %>%
  dplyr::rename(UID = variable, expression = value) %>% 
  left_join(mrna_metadata %>% dplyr::select(UID, biologicalSampleName)) %>% 
  dplyr::group_by(GeneName, biologicalSampleName) %>% 
  dplyr::summarize(median_expression=median(expression)) %>%
  reshape2::dcast(GeneName ~ biologicalSampleName)

rownames(mrna_mat_median) = mrna_mat_median$GeneName
mrna_mat_median$GeneName = NULL

# Get unique metadata
metadata.TF.TF <- mrna_metadata_filtered %>% 
  dplyr::select(biologicalSampleName, Diffname_short, Originating_Lab, Cell_Type, 
                Cell_Line_Type, Cell_Line_of_Origin, Tissue_of_Origin,
                Reprogramming_Gene_Combination, Culture_Conditions,
                Cell_Type_of_Origin_Level2, Reprogramming_Vector_Type_Level2) %>%
  dplyr::filter(biologicalSampleName %in% biosampleInBoth) %>%
  dplyr::mutate(Diffname_short = gsub('-','',Diffname_short)) %>% 
  unique
mrna_mat_median = mrna_mat_median[,metadata.TF.TF$biologicalSampleName]

TF_mat_median = mrna_mat_median[(rownames(mrna_mat_median) %in% unique(TFsMapping$feature)),]

# Split metadata in terms of diffstate
metadata.TF.TF <- split(metadata.TF.TF, metadata.TF.TF$Diffname_short)
```
Calculating coexpression between `r dim(TF_mat_median)[1]` TFs in `r dim(TF_mat_median)[2]` samples
```{r coexpp.TF.TF}
features = data.frame(omics1 = TFsMapping$feature, omics2 = TFsMapping$target) %>%
  dplyr::filter(omics1 %in% rownames(TF_mat_median), omics2 %in% rownames(TF_mat_median))

TF.TF.correlation = lapply(metadata.TF.TF, calculateCor, TF_mat_median, TF_mat_median, features, 'TF.regulator', 'TF.regulated')
TF.TF.correlation = join_all(TF.TF.correlation, by = c("feature","TF.regulator","TF.regulated")) %>%
  dplyr::select(feature, TF.regulator, TF.regulated, DE, EB, ECTO, MESO15, MESO30, MESO5, SC)

metadata.TF.TF = rbindlist(metadata.TF.TF)
```


#### Calculate co-expression between TF-nonTF features at each diffstate
Combine samples to unique biological sample name (take median values for replicates)
```{r combine.samples.TF.nonTF}
# Match up biological sample between assays
biosampleInBoth <- mrna_metadata_filtered$biologicalSampleName

# Filter metadata
mrna_metadata <- mrna_metadata_filtered 

# Take the median across multiple biological samples per feature
mrna_mat_median <- mrna_mat %>%
  dplyr::select(GeneName, one_of(mrna_metadata$UID)) %>%
  melt %>%
  dplyr::rename(UID = variable, expression = value) %>% 
  left_join(mrna_metadata %>% dplyr::select(UID, biologicalSampleName)) %>% 
  dplyr::group_by(GeneName, biologicalSampleName) %>% 
  dplyr::summarize(median_expression=median(expression)) %>%
  reshape2::dcast(GeneName ~ biologicalSampleName)

rownames(mrna_mat_median) = mrna_mat_median$GeneName
mrna_mat_median$GeneName = NULL

# Get unique metadata
metadata.TF.nonTF <- mrna_metadata_filtered %>% 
  dplyr::select(biologicalSampleName, Diffname_short, Originating_Lab, Cell_Type, 
                Cell_Line_Type, Cell_Line_of_Origin, Tissue_of_Origin,
                Reprogramming_Gene_Combination, Culture_Conditions,
                Cell_Type_of_Origin_Level2, Reprogramming_Vector_Type_Level2) %>%
  dplyr::filter(biologicalSampleName %in% biosampleInBoth) %>%
  dplyr::mutate(Diffname_short = gsub('-','',Diffname_short)) %>% 
  unique
mrna_mat_median = mrna_mat_median[,metadata.TF.nonTF$biologicalSampleName]

nonTF_mat_median = mrna_mat_median[!(rownames(mrna_mat_median) %in% unique(TFsMapping$feature)),]
TF_mat_median = mrna_mat_median[(rownames(mrna_mat_median) %in% unique(TFsMapping$feature)),]

# Split metadata in terms of diffstate
metadata.TF.nonTF <- split(metadata.TF.nonTF, metadata.TF.nonTF$Diffname_short)
```
Calculating coexpression between `r dim(TF_mat_median)[1]` TFs and  `r dim(nonTF_mat_median)[1]` nonTFs in `r dim(TF_mat_median)[2]` samples
```{r coexpp.TF.nonTF}
features = data.frame(omics1 = TFsMapping$feature, omics2 = TFsMapping$target) %>%
  dplyr::filter(omics1 %in% rownames(TF_mat_median), omics2 %in% rownames(nonTF_mat_median))

TF.nonTF.correlation = lapply(metadata.TF.nonTF, calculateCor, TF_mat_median, nonTF_mat_median, features, 'TF', 'nonTF')
TF.nonTF.correlation = join_all(TF.nonTF.correlation, by = c("feature","TF","nonTF")) %>%
  dplyr::select(feature, TF, nonTF, DE, EB, ECTO, MESO15, MESO30, MESO5, SC)

metadata.TF.nonTF = rbindlist(metadata.TF.nonTF)
```


#### Calculate co-expression between mirna-mrna features at each diffstate
Combine samples to unique biological sample name (take median values for replicates)
```{r combine.samples.mirna.mrna}
# Match up biological sample between assays
biosampleInBoth <- intersect(mirna_metadata_filtered$biologicalSampleName, 
                             mrna_metadata_filtered$biologicalSampleName)

# Filter metadata
mirna_metadata <- mirna_metadata_filtered %>%
  filter(biologicalSampleName %in% biosampleInBoth)

mrna_metadata <- mrna_metadata_filtered %>%
  filter(biologicalSampleName %in% biosampleInBoth)

# Take the median across multiple biological samples per feature
mirna_mat_median <- mirna_mat %>%
  dplyr::select(GeneName, one_of(mirna_metadata$UID)) %>%
  melt %>%
  dplyr::rename(UID = variable, expression = value) %>% 
  left_join(mirna_metadata %>% dplyr::select(UID, biologicalSampleName)) %>% 
  group_by(GeneName, biologicalSampleName) %>% 
  summarize(median_expression=median(expression)) %>% 
  dcast(GeneName ~ biologicalSampleName)

rownames(mirna_mat_median) = mirna_mat_median$GeneName
mirna_mat_median$GeneName = NULL

# Take the median across multiple biological samples per feature
mrna_mat_median <- mrna_mat %>%
  dplyr::select(GeneName, one_of(mrna_metadata$UID)) %>%
  melt %>%
  dplyr::rename(UID = variable, expression = value) %>% 
  left_join(mrna_metadata %>% dplyr::select(UID, biologicalSampleName)) %>% 
  dplyr::group_by(GeneName, biologicalSampleName) %>% 
  dplyr::summarize(median_expression=median(expression)) %>%
  reshape2::dcast(GeneName ~ biologicalSampleName)

rownames(mrna_mat_median) = mrna_mat_median$GeneName
mrna_mat_median$GeneName = NULL

# Get unique metadata
metadata.mirna.mrna <- mrna_metadata_filtered %>% 
  dplyr::select(biologicalSampleName, Diffname_short, Originating_Lab, Cell_Type, 
                Cell_Line_Type, Cell_Line_of_Origin, Tissue_of_Origin,
                Reprogramming_Gene_Combination, Culture_Conditions,
                Cell_Type_of_Origin_Level2, Reprogramming_Vector_Type_Level2) %>%
  dplyr::filter(biologicalSampleName %in% biosampleInBoth) %>%
  dplyr::mutate(Diffname_short = gsub('-','',Diffname_short)) %>% 
  unique
mrna_mat_median = mrna_mat_median[,metadata.mirna.mrna$biologicalSampleName]
mirna_mat_median = mirna_mat_median[,metadata.mirna.mrna$biologicalSampleName]

# Split metadata in terms of diffstate
metadata.mirna.mrna <- split(metadata.mirna.mrna, metadata.mirna.mrna$Diffname_short)
```
Calculating coexpression between `r dim(mrna_mat_median)[1]` mrnas and  `r dim(mirna_mat_median)[1]` mirnas in `r dim(mirna_mat_median)[2]` samples
```{r coexpp.mirna.mrna}
features = data.frame(omics1 = miRNA.mRNA$feature, omics2 = miRNA.mRNA$target) %>%
  dplyr::filter(omics1 %in% rownames(mirna_mat_median), omics2 %in% rownames(mrna_mat_median))

mirna.mrna.correlation = lapply(metadata.mirna.mrna, calculateCor, mirna_mat_median, mrna_mat_median, features, 'mirna', 'mrna')
mirna.mrna.correlation = join_all(mirna.mrna.correlation, by = c("feature","mirna","mrna")) %>%
  dplyr::select(feature, mirna, mrna, DE, EB, ECTO, MESO15, MESO30, MESO5, SC)

metadata.mirna.mrna = rbindlist(metadata.mirna.mrna)
```


#### Calculate co-expression between methylation-mrna features at each diffstate
Combine samples to unique biological sample name (take median values for replicates)
```{r combine.samples.methyl.mrna}
# Match up biological sample between assays
biosampleInBoth <- intersect(methyl_metadata_filtered$biologicalSampleName, 
                             mrna_metadata_filtered$biologicalSampleName)

# Filter metadata
methyl_metadata <- methyl_metadata_filtered %>%
  filter(biologicalSampleName %in% biosampleInBoth)

mrna_metadata <- mrna_metadata_filtered %>%
  filter(biologicalSampleName %in% biosampleInBoth)

# Take the median across multiple biological samples per feature
rownames(methyl_metadata) = methyl_metadata$UID
methyl_mat_median <- methyl_mat[,methyl_metadata$UID]
rownames(methyl_mat_median) = methyl_mat$methProbeID
colnames(methyl_mat_median) = methyl_metadata[colnames(methyl_mat_median), 'biologicalSampleName']

# Take the median across multiple biological samples per feature
mrna_mat_median <- mrna_mat %>%
  dplyr::select(GeneName, one_of(mrna_metadata$UID)) %>%
  melt %>%
  dplyr::rename(UID = variable, expression = value) %>% 
  left_join(mrna_metadata %>% dplyr::select(UID, biologicalSampleName)) %>% 
  dplyr::group_by(GeneName, biologicalSampleName) %>% 
  dplyr::summarize(median_expression=median(expression)) %>%
  reshape2::dcast(GeneName ~ biologicalSampleName)

rownames(mrna_mat_median) = mrna_mat_median$GeneName
mrna_mat_median$GeneName = NULL

# Get unique metadata
metadata.methyl.mrna <- mrna_metadata_filtered %>% 
  dplyr::select(biologicalSampleName, Diffname_short, Originating_Lab, Cell_Type, 
                Cell_Line_Type, Cell_Line_of_Origin, Tissue_of_Origin,
                Reprogramming_Gene_Combination, Culture_Conditions,
                Cell_Type_of_Origin_Level2, Reprogramming_Vector_Type_Level2) %>%
  dplyr::filter(biologicalSampleName %in% biosampleInBoth) %>%
  dplyr::mutate(Diffname_short = gsub('-','',Diffname_short)) %>% 
  unique
mrna_mat_median = mrna_mat_median[,metadata.methyl.mrna$biologicalSampleName]
methyl_mat_median = methyl_mat_median[,metadata.methyl.mrna$biologicalSampleName]

# Split metadata in terms of diffstate
metadata.methyl.mrna <- split(metadata.methyl.mrna, metadata.methyl.mrna$Diffname_short)
```
Performing analysis with `r dim(mrna_mat_median)[1]` mrnas and  `r dim(methyl_mat_median)[1]` methylation probes in `r dim(methyl_mat_median)[2]` samples
```{r coexpp.methyl.mrna}
features = data.frame(omics1 = methyl.mrna$feature, omics2 = methyl.mrna$target) %>%
  dplyr::filter(omics1 %in% rownames(methyl_mat_median), omics2 %in% rownames(mrna_mat_median))

methyl.mrna.correlation = lapply(metadata.methyl.mrna, 
                                 calculateCor, 
                                 methyl_mat_median, 
                                 mrna_mat_median, 
                                 features, 'methyl', 'mrna')
methyl.mrna.correlation = join_all(methyl.mrna.correlation, by = c("feature","methyl","mrna")) %>%
  dplyr::select(feature, methyl, mrna, DE, EB, ECTO, MESO5, SC)

metadata.methyl.mrna = rbindlist(metadata.methyl.mrna)
```


#### Calculate co-expression between methylation-mirna features at each diffstate
Combine samples to unique biological sample name (take median values for replicates)
```{r combine.samples.methyl.mirna}
# Match up biological sample between assays
biosampleInBoth <- intersect(methyl_metadata_filtered$biologicalSampleName,
                             mirna_metadata_filtered$biologicalSampleName)

# Filter metadata
methyl_metadata <- methyl_metadata_filtered %>%
  filter(biologicalSampleName %in% biosampleInBoth)

mirna_metadata <- mirna_metadata_filtered %>%
  filter(biologicalSampleName %in% biosampleInBoth)

# Take the median across multiple biological samples per feature
rownames(methyl_metadata) = methyl_metadata$UID
methyl_mat_median <- methyl_mat[,methyl_metadata$UID]
rownames(methyl_mat_median) = methyl_mat$methProbeID
colnames(methyl_mat_median) = methyl_metadata[colnames(methyl_mat_median), 'biologicalSampleName']

# Take the median across multiple biological samples per feature
mirna_mat_median <- mirna_mat %>%
  dplyr::select(GeneName, one_of(mirna_metadata$UID)) %>%
  melt %>%
  dplyr::rename(UID = variable, expression = value) %>% 
  left_join(mirna_metadata %>% dplyr::select(UID, biologicalSampleName)) %>% 
  dplyr::group_by(GeneName, biologicalSampleName) %>% 
  dplyr::summarize(median_expression=median(expression)) %>%
  reshape2::dcast(GeneName ~ biologicalSampleName)

rownames(mirna_mat_median) = mirna_mat_median$GeneName
mirna_mat_median$GeneName = NULL

# Get unique metadata
metadata.methyl.mirna <- mirna_metadata_filtered %>% 
  dplyr::select(biologicalSampleName, Diffname_short, Originating_Lab, Cell_Type, 
                Cell_Line_Type, Cell_Line_of_Origin, Tissue_of_Origin,
                Reprogramming_Gene_Combination, Culture_Conditions,
                Cell_Type_of_Origin_Level2, Reprogramming_Vector_Type_Level2) %>%
  dplyr::filter(biologicalSampleName %in% biosampleInBoth) %>%
  dplyr::mutate(Diffname_short = gsub('-','',Diffname_short)) %>% 
  unique
mirna_mat_median = mirna_mat_median[,metadata.methyl.mirna$biologicalSampleName]
methyl_mat_median = methyl_mat_median[,metadata.methyl.mirna$biologicalSampleName]

# Split metadata in terms of diffstate
metadata.methyl.mirna <- split(metadata.methyl.mirna, metadata.methyl.mirna$Diffname_short)
```
Performing analysis with `r dim(mirna_mat_median)[1]` mirnas and  `r dim(methyl_mat_median)[1]` methylation probes in `r dim(methyl_mat_median)[2]` samples
```{r coexpp.methyl.mirna}
features = data.frame(omics1 = methyl.mirna$feature, omics2 = methyl.mirna$target) %>%
  dplyr::filter(omics1 %in% rownames(methyl_mat_median), omics2 %in% rownames(mirna_mat_median))

methyl.mirna.correlation = lapply(metadata.methyl.mirna, 
                                 calculateCor, 
                                 methyl_mat_median, 
                                 mirna_mat_median, 
                                 features, 'methyl', 'mirna')
methyl.mirna.correlation = join_all(methyl.mirna.correlation, by = c("feature","methyl","mirna")) %>%
  dplyr::select(feature, methyl, mirna, DE, EB, ECTO, MESO5, SC)

metadata.methyl.mirna = rbindlist(metadata.methyl.mirna)
```


### Store results in synapse
```{r synapsestore, echo=FALSE, include=FALSE, eval=TRUE, cache=FALSE}
ActivityName <- 'Calculate co-expression between assays'

ThisFileName <- 'CalculateCoExp_All.Rmd'

# Github link
ThisRepo <- getRepo(repository = "th1vairam/pcbc_c4_analysis", 
                    ref="branch", 
                    refName='discordant_anal')

ThisFile <- getPermlink(repository = ThisRepo,
                        repositoryPath=paste0('code/Rmd/', ThisFileName))    

# Populate wiki with results
CODE <- Folder(name = 'Coexpression Between Different Assays At Each Differentiation Stage', parentId = parentId)
CODE <- synStore(CODE)

# Store metadata
metadata <- list(TF_TF = metadata.TF.TF,
                 TF_nonTF = metadata.TF.nonTF,
                 mirna_mrna = metadata.mirna.mrna, 
                 methyl_mrna = metadata.methyl.mrna,
                 methyl_mirna = metadata.methyl.mirna)
metadata_obj = mapply(function(x,y, parentId, used, executed, activityName){
  write.table(x, file=paste0('Metadata_',y,'.tsv'), sep = '\t', quote=F, row.names=F)
  obj = File(paste0('Metadata_',y,'.tsv'), name = paste('Metadata',y), parentId = parentId)
  obj = synStore(obj, used = used, executed = executed, activityName = activityName)
}, metadata, names(metadata), MoreArgs = list(CODE$properties$id, ALL_USED_IDs, ThisFile, ActivityName),
SIMPLIFY = F)


# Store coexpression
CoExp <- list(TF_TF = TF.TF.correlation,
              TF_nonTF = TF.nonTF.correlation,
              mirna_mrna = mirna.mrna.correlation, 
              methyl_mrna = methyl.mrna.correlation,
              methyl_mirna = methyl.mirna.correlation)
CoExp_obj = mapply(function(x,y, parentId, used, executed, activityName){
  write.table(x, file=paste0('CoExp_',y,'.tsv'), sep = '\t', quote=F, row.names=F)
  obj = File(paste0('CoExp_',y,'.tsv'), name = paste('Coexpression',y), parentId = parentId)
  obj = synStore(obj, used = used, executed = executed, activityName = activityName)
}, CoExp, names(CoExp), MoreArgs = list(CODE$properties$id, ALL_USED_IDs, ThisFile, ActivityName),
SIMPLIFY = F)
```

### Source Code
[Source R Markdown](`r ThisFile`)