---
title: "Coexpression Network Analysis for eXpress aligned mRNA data with only technical covariates"
author: "Thanneer Perumal"
date: "`r date()`"
output: html_document
---
```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
## It is assumed your working directory is where this file is

# Clear R console screen output
cat("\014")  

# Load required libraries
library(synapseClient)
library(data.table)
library(dplyr)
library(knitr)
library(stringr)
library(RColorBrewer)
library(ggplot2)
library(gplots)
library(limma)
library(edgeR)
library(ctv)
library(psych)
library(reshape2)
library(vcd)
library(erer)
library(fpc)
library(WGCNA)
library(Rclusterpp)
library(circlize)
library(knit2synapse)
# library(ComplexHeatmap)
library(vbsr)

## Needs the dev branch
library(rGithubClient)

## Requires ggdendro
devtools::source_url("https://raw.githubusercontent.com/chr1swallace/random-functions/master/R/ggplot-heatmap.R")

# source('/mnt/Github/knit2synapse-1/R/knitFile2SynapseFolder.R')
# knitToFolderEntity(file = "./CoexppNet_mRNA_TechCovar.Rmd", 
#                    entityName = 'Coexpression Network Analysis mRNA All Technical Covariates Model', 
#                    parentId = 'syn3256504', 
#                    overwrite=F)

synapseLogin()

# source utility files from ../R/lib folder
file.sources = list.files('../R/lib',pattern="*.R")
file.sources = sapply(file.sources,function(x){return(paste('../R/lib',x,sep='/'))})
tmp = sapply(file.sources,source,.GlobalEnv)

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = FALSE)
```
### Download data
```{r setup, include=FALSE}
# Input Parameters
COUNT_ID = 'syn3446250'
METADATA_ID = 'syn3156503'

SYNAPSE_STORE = T
parentId = 'syn3256504'

# Specify factor and continuous covarites pool (adjusted covariates will only be the subset of these covariates)
FactorCovariates = c('Diffname_short', 'run', 'lane', 'Cell_Line_Type', 'Cell_Line_of_Origin', 'Tissue_of_Origin', 'Reprogramming_Gene_Combination', 'Culture_Conditions', 'Donor_Life_Stage', 'Race', 'Ethnicity' , 'Gender', 'Disease', 'Originating_Lab', 'Donor_ID', 'Cell_Type_of_Origin_Level2', 'Reprogramming_Vector_Type')
ContCovariates = c('PassageAtThaw', 'PassageAtHarvest')
```
Synapse id of count matrix used for the analysis is `r COUNT_ID` and the synapse id of meta data table used for the analysis is `r METADATA_ID`. 

Factor covariates considered for analysis are `r paste(gsub('_','\\\\_',FactorCovariates),collapse=',')`, and continuous covariates considered for the analysis are `r paste(gsub('_','\\\\_',ContCovariates),collapse=',')`.

Obtain count matrix and metadata from synapse.
```{r getdata, cache=TRUE, include=FALSE}
# Get count matrix
COUNT_OBJ = synGet(COUNT_ID)
ALL_USED_IDs = COUNT_OBJ$properties$id
COUNT = fread(getFileLocation(COUNT_OBJ), data.table=FALSE)
row.names(COUNT) = COUNT[,1]
COUNT = COUNT[,-(1)]

# Get metadata
METADATA_OBJ = synTableQuery(paste('SELECT * FROM',METADATA_ID,sep=' '))
ALL_USED_IDs[length(ALL_USED_IDs)+1] = METADATA_OBJ@schema
METADATA = METADATA_OBJ@values
```
Preprocess counts matrix and metadata.
```{r preprocessing, include=FALSE}
# Preprocess metadata
METADATA[METADATA == 'N/A'] = NA

# Replace all special characters with blank
myFix <- function(x) str_replace_all(x, '[^[:alnum:]]', '')
METADATA <- METADATA %>%
  dplyr::mutate_each(funs(myFix), -UID, -C4_Cell_Line_ID, -biologicalSampleName) # fix them but don't touch some columns

# Set rownames
rownames(METADATA) = METADATA$UID
```
### Preprocess data
* Remove somatic samples and samples with not type.
* Remove samples that failed QC and samples classified as exclude.
* Remove samples with abnormal karyotypes.
```{r filtering, echo=TRUE}
#### Pre processing mRNA expression counts and metadata ####
metadata_filtered <- 
  METADATA %>%
  filter(Diffname_short != "") %>%
  filter(UID %in% colnames(COUNT)) %>%
  filter(Cell_Type == "PSC") %>%  
  filter(pass_qc == "TRUE") %>%
  filter(exclude != "TRUE") %>%
  filter(C4_Karyotype_Result != "abnormal")

REMOVED_UID <- setdiff(colnames(COUNT), metadata_filtered$UID)
METADATA <- METADATA[metadata_filtered$UID,]
COUNT <- COUNT[, METADATA$UID]
```
The following `r length(REMOVED_UID)` samples were removed:
  
`r paste(gsub('_','\\\\_',REMOVED_UID), collapse= ',')` 

### CPM Normalisation
Determine design matrix for normalisation and differential expression analysis. 

Remove genes that have less than 1 cpm counts in at least 50% of each of the individual differentiation stages.
```{r cpmnormalisation}
tmp <- tapply(colnames(COUNT),
              factor(METADATA$Diffname_short),
              function(cols,COUNT){PROCESSED_COUNTS = getGeneFilteredGeneExprMatrix(COUNT[,cols])},
              COUNT)

ugenes <- c()
for (i in unique(METADATA$Diffname_short)) {
  ugenes <- unique(c(ugenes,tmp[[i]]$filteredExprMatrix$genes[,1]))
}

COUNT <- COUNT[ugenes,,drop=F]
PROCESSED_COUNTS = getGeneFilteredGeneExprMatrix(COUNT,MIN_GENE_CPM=0,
                                                 MIN_SAMPLE_PERCENT_WITH_MIN_GENE_CPM=0)
```
`r dim(COUNT)[1]` genes and `r dim(COUNT)[2]` samples are considered for the analysis
### Covariates
```{r covariates.clustering}
# Find inter relation between factor covariates
COVARIATES = METADATA[,c(FactorCovariates,ContCovariates)]

# Convert factor covariates to factors and continuous to numeric
COVARIATES[,FactorCovariates] = lapply(COVARIATES[,FactorCovariates], factor)
COVARIATES[,ContCovariates] = lapply(COVARIATES[,ContCovariates], as.numeric)
```
### Normalisation (with technical covariates only)
Initial normalisation usign voom (residualised technical covariates with donor id as random variable)
```{r technical.voom.normalisation, fig.height=14, fig.width=20}
# Design matrix with only technical covariates
covars = c("Diffname_short","Gender", "Cell_Type_of_Origin_Level2", "Donor_Life_Stage", "lane", "Culture_Conditions", "run")
DESIGN = getDesignMatrix(COVARIATES[,covars], Intercept = F)
DESIGN = DESIGN$design[,linColumnFinder(DESIGN$design)$indepCols]

# Initial normalisation of gene expression
VOOM.GENE_EXPRESSION = voom(PROCESSED_COUNTS$filteredExprMatrix, design=DESIGN, plot=T)
    
# Calculate correlation between donors
correlation <- duplicateCorrelation(VOOM.GENE_EXPRESSION, block=COVARIATES$Donor_ID)

# Normalisation of gene expression with random effects
VOOM.GENE_EXPRESSION = voom(PROCESSED_COUNTS$filteredExprMatrix, 
                            design=DESIGN, 
                            plot=T, 
                            block=COVARIATES$Donor_ID, 
                            correlation = correlation$cor)

# Fit linear model using mixed effects design
FIT = lmFit(VOOM.GENE_EXPRESSION, block=COVARIATES$Donor_ID, correlation = correlation$cor)

# Calculate residuals
EXPR = residuals.MArrayLM(FIT, VOOM.GENE_EXPRESSION)
varsToAddBackIn = c("Diffname_shortDE", "Diffname_shortEB", "Diffname_shortECTO", "Diffname_shortMESO15", "Diffname_shortMESO30", "Diffname_shortMESO5", "Diffname_shortSC", "Genderfemale", "Cell_Type_of_Origin_Level2blood", "Cell_Type_of_Origin_Level2fibroblast", "Cell_Type_of_Origin_Level2innercellmass", "Donor_Life_Stageadult", "Donor_Life_Stageblastocyst", "Donor_Life_Stagechildhood", "Donor_Life_Stagefetal", "Donor_Life_Stagenewborn", "Culture_ConditionsDMEMF12KOSR01mMBME01mMNEAA01mMLgluatime20ngmlbFGF", "Culture_ConditionsE8", "Culture_ConditionsFeederandKOSRcultureconditions", "Culture_ConditionsStandardTESR1MatrigelorKOSRMEFs", "Culture_ConditionsStromalnonprimed", "Culture_ConditionsStromalprimed", "Culture_ConditionsiPSCsweregeneratedandintiallymaintainedusingMEFKSRculturesAftertransfertoWiCelliPSCsexpandedinFeederIndependentTeSR1Medium")
EXPR = EXPR + VOOM.GENE_EXPRESSION$weights * (FIT$coefficients[,varsToAddBackIn] %*% t(VOOM.GENE_EXPRESSION$design[,varsToAddBackIn]))

# Design matrix with all covariates
DESIGN = getDesignMatrix(COVARIATES, Intercept = F)
DESIGN = DESIGN$design[,linColumnFinder(DESIGN$design)$indepCols]
colnames(DESIGN) = str_sub(colnames(DESIGN), 1, 30)

# Correlation between covariates and expression residuals
residualCovaraitesCorrelation = runPCAandPlotCorrelations(EXPR, DESIGN, dataName = 'residual covariates relation')
print(residualCovaraitesCorrelation$PC_res[[2]])
```
Sample clustering to check for outliers
```{r sample.clustering, fig.height=15, fig.width=20}
sampleTrees = flashClust::hclust(dist(t(EXPR)), method = "average")
tree = cutreeStatic(sampleTrees, cutHeight = 200, minSize=3)

plotDendroAndColors(sampleTrees, 
                    labels2colors(cbind(tree,COVARIATES)),
                    groupLabels = c('Cluster',colnames(COVARIATES)))
collectGarbage()
```
Sample pca to check for outliers
```{r residual.clustering, fig.height=10, fig.width=10}
# Find principal components of expression to plot
PC <- prcomp(EXPR)

# Determine number of clusters automatically using pamk
pam.cluster <- pamk(t(EXPR), krange=2:10)

# Plot first 2 PCs
plotdata <- data.frame(UID=rownames(PC$rotation), 
                       PC1=PC$rotation[,1], 
                       PC2=PC$rotation[,2], 
                       cluster=pam.cluster$pamobject$clustering)

plotdata <- merge(plotdata, METADATA, by="UID")
plotdata <- mutate(plotdata, labels = '');#ifelse(Diffname_short == '',as.character(UID),'')) 

p <- ggplot(plotdata, aes(x=PC1, y=PC2))
p <- p + geom_point(aes(color=factor(cluster), shape=Diffname_short, size=Gender))
p <- p + theme_bw() + scale_size_manual(values = c(4, 2)) + scale_shape_manual(values=1:7) + theme(legend.position="top") 
p <- p + geom_text(aes(label= labels), size=4, hjust=0)
p
```
### Coexpression networks
```{r wgcna.power, fig.height=8, fig.width=8}
# Allow multi-threading within WGCNA. 
# Caution: skip this line if you run RStudio or other third-party R environments.
enableWGCNAThreads()

# Choose a set of soft-thresholding powers
powers = c(seq(2,10,by=1), seq(12,30, by=2));

# Call the network topology analysis function
EXPR = t(EXPR)
powerTables = pickSoftThreshold((EXPR), 
                                powerVector=powers,
                                corFnc = 'bicor',
                                networkType = 'unsigned',
                                verbose = 2)[[2]]
collectGarbage();

# Plot the results
colors = c("black", "red")

# Will plot these columns of the returned scale free analysis tables
plotCols = c(2,5,6,7)
colNames = c("Scale Free Topology Model Fit", "Mean connectivity", "Median connectivity", "Max connectivity");

# Plot the quantities in the chosen columns vs. the soft thresholding power
p <- list()
for (i in 1:length(plotCols)){
  name = colnames(powerTables)[plotCols[i]]
  p[[name]] <- ggplot(powerTables, aes_string(x = 'Power', y = name)) + geom_point()
  p[[name]] <- p[[name]] + geom_text(aes(label= Power), size=4, hjust=1, vjust = -1)
  p[[name]] <- p[[name]] + ggtitle(colNames[i])
}

multiplot(plotlist=p, cols=2)
softPower = 18
collectGarbage()
save(list=c('powerTables'), file = 'CoexppNet_mRNA_technicalCovariates.RData')
```
Beta is chosen to be `r softPower`
```{r wgcna.adj, fig.width=15, fig.height=10}
# Get adjacency matrix
Adj = bicor(EXPR, nThreads = 2)
Adj = abs(Adj^softPower);#Adj/abs(Adj) * abs(Adj^softPower)
collectGarbage()

# Get topological overlap matrix
TOM = TOMdist(Adj, TOMType='unsigned', verbose=3)

# Hierarchichal clustering
tree = Rclusterpp.hclust(as.dist(TOM), method = "average", distance = "euclidean")
collectGarbage()

# We set the minimum module size relatively high
minModuleSize = 30;

# Module identification using dynamic tree cut:
unmergedNetModules = cutreeDynamic(dendro = tree, 
                                   method = "hybrid",
                                   deepSplit = 4,
                                   minClusterSize = minModuleSize,                                   
                                   distM = TOM,
                                   pamRespectsDendro = TRUE,
                                   verbose=3)

unmergedNetModules.colors = labels2colors(unmergedNetModules)
collectGarbage()

save(list=c('powerTables', 'Adj', 'TOM','tree', 'unmergedNetModules.colors'), 
     file = 'CoexppNet_mRNA_technicalCovariates.RData')
```
Dendrogram tree of gene clustering
```{r dendro.modules, fig.height=12, fig.width=20}
plotDendroAndColors(tree, unmergedNetModules.colors, "Dynamic Tree Cut",
                    dendroLabels = FALSE, hang = 0.03,
                    addGuide = TRUE, guideHang = 0.05)
```
### Module analysis
```{r module.analysis, fig.height=5, fig.width=10}
# Merge modules with similar expression profiles
# Calculate module eigengenes
unmergedMEs = moduleEigengenes(EXPR, 
                               colors = unmergedNetModules.colors,
                               nPC = 10)

# Calculate dissimilarity between module eigen genes
MEDiss = 1 - abs(cor(unmergedMEs$eigengenes, use = "p"))

# Cluster module eigen genes
METree = hclust(as.dist(MEDiss), method = "average");
collectGarbage()

# Plot the results
par(mfrow = c(1,1))
plot(METree, main = "Clustering of module eigengenes",
     xlab = "", sub = "")
abline(h=0.15, col = "red")

# Merge modules
merge = mergeCloseModules(fixDataStructure(EXPR), unmergedNetModules.colors, cutHeight = 0.15, verbose = 3)

# Numeric module labels
mergedNetModules = merge$colors

# Convert labels to colors
mergedNetModules.colors = labels2colors(mergedNetModules)
mergedNetModules.colors = matchLabels(unmergedNetModules.colors, mergedNetModules.colors)

# Eigengenes of the new merged modules:
mergedMEs = merge$newMEs
MEs = orderMEs(mergedMEs[[1]]$data)

tmp = corr.test(MEs)
tmp1 = tmp$r
tmp1[tmp$pval >= 0.05] = 0
tmp1 = melt(tmp1)

# Plot correlation between eigen genes
writeLines('Correlation between eigen genes')
p = ggplot(tmp1, aes(x=X1,y=X2)) + geom_tile(aes(fill = value), colour = "white") + 
  scale_fill_gradient2(low = "green", high = "red", limits = c(-1,1))
p = p + theme(axis.text.x = element_text(angle = 90, hjust = 1))
p

# Plot dendrogram and module colors
writeLines('Module dendrogram and color assignment')
plotDendroAndColors(tree,
                    cbind(unmergedNetModules.colors, mergedNetModules.colors),
                    c("Unmerged", "Merged"),
                    dendroLabels = FALSE, hang = 0.03,
                    addGuide = TRUE, guideHang = 0.05)
save(list=c('powerTables', 'Adj', 'TOM','tree', 'unmergedNetModules.colors',
            'unmergedMEs', 'mergedNetModules.colors', 'mergedMEs'), 
     file = 'CoexppNet_mRNA_technicalCovariates.RData')
```
### Compare eigen genes and covariates
```{r compare.covars, fig.height=10, fig.width=15}
# Set up variables to contain the module-trait correlations
DESIGN = getDesignMatrix(COVARIATES[,FactorCovariates], Intercept = F)$design
DESIGN[is.na(DESIGN)] = 0

DESIGN.SPLIT = lapply(FactorCovariates, function(x, DESIGN){ ind = grep(x, colnames(DESIGN)); return(DESIGN[,ind])}, DESIGN)
names(DESIGN.SPLIT) = FactorCovariates

# Also obtain comparisons
ALL.COMP = mapply(function(x, y){
  CONT.NAMES <- colnames(x)
  CONT.NAMES <- combn(CONT.NAMES,2)
  CONT.NAMES <- apply(CONT.NAMES,2,paste,collapse='-')
  
  CONT <- makeContrasts(contrasts=CONT.NAMES,levels=colnames(x))
  colnames(CONT) = gsub(y,'',colnames(CONT))
  
  ALL.CONT = x %*% CONT
  
  colnames(x) = gsub(y, '', colnames(x))
  
  return(ALL.CONT)
  }, DESIGN.SPLIT, FactorCovariates)
ALL.COMP = do.call(cbind, ALL.COMP)
ALL.COMP = ALL.COMP[,!duplicated(colnames(ALL.COMP))]
```
Correlation between module eigen genes and molecular traits
```{r calculate.corr, fig.height=15, fig.width=10}
# Calculate the correlations
g = corr.test(mergedMEs[[1]]$data, ALL.COMP, adjust = 'bonferroni')

moduleTraitCor = g$r
moduleTraitPvalue = g$p
moduleTraitCor[moduleTraitPvalue > 0.05] = 0

moduleTraitCor = moduleTraitCor[, colSums(moduleTraitCor) != 0]
colnames(moduleTraitCor) = str_sub(colnames(moduleTraitCor),1,25)

tmp = reshape2::melt(moduleTraitCor)
setnames(tmp, c('X1','X2'), c('MEs','Trait'))
p = ggplot(tmp, aes(x=MEs,y=Trait)) + geom_tile(aes(fill = value), colour = "white") + 
  scale_fill_gradient2(low = "green", high = "red", limits = c(-1,1))
p = p + theme(axis.text.x = element_text(angle = 90, hjust = 1))
p

# Heatmap(moduleTraitCor, 
#         column_title = 'Covariates', 
#         row_title = 'Module Eigengene', 
#         name = 'Correlation',
#         col = colorRamp2(c(-1,0,1), c("green", "white", "red")))

save(list=c('powerTables', 'Adj', 'TOM','tree', 'unmergedNetModules.colors',
            'unmergedMEs', 'mergedNetModules.colors', 'mergedMEs','g','ALL.COMP'), 
     file = 'CoexppNet_mRNA_technicalCovariates.RData')
```
vbsr between MEs and traits
```{r vbsr, fig.height=15, fig.width=10}
associationStats = sapply(MEs, function(x, ALL.COMP){
  tmp = vbsr(x,ALL.COMP)$z
  names(tmp) = colnames(ALL.COMP)
  return(tmp)
}, ALL.COMP)
rownames(associationStats) = str_sub(rownames(associationStats), 1, 25)

associationStats = associationStats[rowSums(abs(associationStats) >= 5) != 0,colSums(abs(associationStats) >= 5) != 0]

tmp = reshape2::melt(associationStats)
setnames(tmp, c('X1','X2','value'), c('Trait', 'MEs','Zscores'))
p = ggplot(tmp, aes(x=MEs,y=Trait)) + geom_tile(aes(fill = Zscores), colour = "white") + 
  scale_fill_gradient2(low = "green", high = "red", limits = c(-70,70))
p = p + theme(axis.text.x = element_text(angle = 90, hjust = 1))
p
```

### Synapse Store
```{r synapse.store, cache=FALSE, include=FALSE}
parentId = 'syn3256504'

activityName='Coexpression Network Analysis of eXpress aligned mRNA data with technical covariates model'
  
thisFileName <- 'CoexppNet_mRNA_TechCovar.Rmd'
  
# Github link
thisRepo <- getRepo(repository = "th1vairam/pcbc_c4_analysis", 
                    ref="branch", 
                    refName='mRNANet')

thisFile <- getPermlink(repository = thisRepo,
                        repositoryPath=paste0('code/Rmd/', thisFileName))
    
# Create folder to tore the results and populate analysis wiki
CODE <- Folder(name = 'Coexpression Network Analysis mRNA All Technical Covariates Model',parentId = parentId)
CODE <- synStore(CODE)
      
# Store Results
COEXPP_OBJ <- File('./CoexppNet_mRNA_technicalCovariates.RData', name = 'Coexpression Networks', parentId = CODE$properties$id)
COEXPP_OBJ <- synStore(COEXPP_OBJ, used = ALL_USED_IDs, activityName = activityName, executed=thisFile)

# Store Results
MOD = data.frame(GeneNames = colnames(EXPR) , merged = mergedNetModules.colors, unmerged = unmergedNetModules.colors)
write.table(MOD, file='modules.tsv', sep='\t', quote=F, row.names=F)
MOD_OBJ <- File('modules.tsv', name = 'Coexpression Network Modules', parentId = CODE$properties$id)
MOD_OBJ <- synStore(MOD_OBJ, used = ALL_USED_IDs, activityName = activityName, executed=thisFile)
```