---
title: "Covariate Analysis using eXpress aligned mRNA"
author: "Thanneer Perumal"
date: "03/26/2015"
output: html_document
---

```{r libs, echo=FALSE, warning=FALSE, message=FALSE}
## It is assumed your working directory is where this file is
# Load required libraries
library('synapseClient')
library('limma')
library('edgeR')
library('RColorBrewer')
library('ctv')
library('ggplot2')
library('psych')
library('reshape2')
library('gplots')
library('vcd')
library('psych')
library('erer')
## Requires ggdendro
devtools::source_url("https://raw.githubusercontent.com/chr1swallace/random-functions/master/R/ggplot-heatmap.R")

# devtools::source_gist('2866ef5c0aeb64d265ed')
# knit2synapse(file = "./CovariateAnalysis_Express.Rmd", owner = 'syn3382563', wikiName = "Covariate Analysis Express",overwrite=T)

synapseLogin()

# source utility files from ../R/lib folder
file.sources = list.files('../R/lib',pattern="*.R")
file.sources = sapply(file.sources,function(x){return(paste('../R/lib',x,sep='/'))})
tmp = sapply(file.sources,source,.GlobalEnv)

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE)
```

```{r setup}
# Input Parameters
COUNT_ID = 'syn3354716';
METADATA_ID = 'syn3156503';
OUTLIERS_ID = 'NULL';

SYNAPSE_STORE = T;
parentId = 'syn3276099';

# Specify factor and continuous covarites pool (adjusted covariates will only be the subset of these covariates)
FactorCovariates = c('Diffname_short','run','lane','index','Cell_Type_of_Origin',
                     'Tissue_of_Origin', 'Reprogramming_Vector_Type','Reprogramming_Gene_Combination',
                     'Donor_Life_Stage','Originating_Lab','Gender','High_Confidence_Donor_ID')
ContCovariates = c('numberReads','mappedReads')
```
Synapse id of count matrix used for the analysis is `r COUNT_ID` and the synapse id of meta data table used for the analysis is `r METADATA_ID`. Outliers were extracted from `r OUTLIERS_ID`

Factor covariates considered for analysis are `r FactorCovariates`, and continuous covariates considered for the analysis are `r ContCovariates`.

Obtain count matrix and metadata from synapse.
```{r getdata, cache=TRUE}
# Get count matrix
COUNT_OBJ = synGet(COUNT_ID);
ALL_USED_IDs = list(COUNT_OBJ$properties$id);
COUNT = read.table(getFileLocation(COUNT_OBJ),sep=',',header=T,row.names=1, check.names=FALSE)

# Get metadata
METADATA_OBJ = synTableQuery(paste('SELECT * FROM',METADATA_ID,sep=' '))
ALL_USED_IDs[length(ALL_USED_IDs)+1] = METADATA_OBJ@schema;
METADATA = METADATA_OBJ@values
```

Preprocess counts matrix and metadata 
```{r preprocessing}
# Preprocess metadata
METADATA[METADATA == 'N/A'] = NA
METADATA[METADATA == 'unknown'] = NA

# Assign new values for MESO-5, 15 and 30 samples
METADATA$Diffname_short[METADATA$Diffname_short == 'MESO-5'] = 'MESO_EARLY'
METADATA$Diffname_short[METADATA$Diffname_short == 'MESO-15' | METADATA$Diffname_short == 'MESO-30'] = 'MESO_LATE'

# Arrange count and metadata
RowsPresent = match(colnames(COUNT), METADATA$UID);
METADATA = METADATA[RowsPresent,]
rownames(METADATA) = METADATA$UID

#### Pre processing mRNA expression counts and metadata ####
# Remove somatic samples and samples with type NA
ind = METADATA$Cell_Type == "PSC" & !is.na(METADATA$Cell_Type)
COUNT = COUNT[,ind]
METADATA = METADATA[ind,]

# Remove samples that failed QC and samples classified as bad_lines
ind <- which(!METADATA$pass_qc | METADATA$exclude)

# Remove samples with abnormal karyotypes
ind = unique(c(ind, which(METADATA$C4_Karyotype_Result == "abnormal" | is.na(METADATA$C4_Karyotype_Result))))

REMOVED_UID <- METADATA$UID[ind]

COUNT <- COUNT[,-(ind)]
METADATA <- METADATA[-(ind),]
```
Following samples `r length(REMOVED_UID)` were removed `r REMOVED_UID` 

### Normalisation
Determine design matrix for normalisation and differential expression analysis. 

Remove genes that have less than 1 cpm counts in at least 50% of each of the individual differentiation stages.
```{r cpmnormalisation}
tmp <- tapply(colnames(COUNT),factor(METADATA$Diffname_short),function(cols,COUNT){PROCESSED_COUNTS = getGeneFilteredGeneExprMatrix(COUNT[,cols])},COUNT)
ugenes <- c()
for (i in unique(METADATA$Diffname_short))
  ugenes <- unique(c(ugenes,tmp[[i]]$filteredExprMatrix$genes[,1]))

COUNT <- COUNT[ugenes,,drop=F]
PROCESSED_COUNTS = getGeneFilteredGeneExprMatrix(COUNT,MIN_GENE_CPM=0,MIN_SAMPLE_PERCENT_WITH_MIN_GENE_CPM=0)
```

### Covariate clustering
Determine relationship between covariates. 

```{r covariates.clustering, fig.width=12, fig.height=10}
# Find inter relation between factor covariates
COVARIATES = METADATA[,c(FactorCovariates,ContCovariates)]

# Convert factor to numeric matrix 
COVARIATES[,FactorCovariates] = apply(COVARIATES[,FactorCovariates],2,function(cols){cols=as.numeric(unclass(factor(cols)))})

heatmap.2(as.matrix(COVARIATES), col = brewer.pal(11,'Spectral'), scale='column',
          Rowv = F, cex.lab = 1.5, cex.axis = 1.5, dendrogram = 'column',
          labRow=rep(as.character(' '),dim(COVARIATES)[1]),
          cex.main = 1.5, mar=c(18,10), xlab = "Covariates", ylab = "Samples", 
          main = "Clustering of covariates", keysize=1)
```

### Covariate correlation

```{r covariates.correlation, fig.width=12, fig.height=10}
COVARIATES.CORRELATION = getAssociationStatistics(COVARIATES,FactorCovariates,ContCovariates,PVAL=1)
ggheatmap.show(COVARIATES.CORRELATION$plot, col.width=0.3, row.width=0.15)
```

Initial normalisation usign voom (with NULL design)

```{r initial.voom.normalisation}
VOOM.GENE_EXPRESSION = voom(PROCESSED_COUNTS$filteredExprMatrix, design=NULL, plot=T)

# Find PC of gene expression and significant covariates that are highly correlated with PCs
DM = getDesignMatrix(METADATA[,c(FactorCovariates,ContCovariates)],FactorCovariates)
```

Correlation between pca of unadjusted mRNA expression and covariates is used to find significant covariates

```{r preAdjusted.covariates}
preAdjustedSigCovars = runPCAandPlotCorrelations(VOOM.GENE_EXPRESSION$E, DM$design,'NULL design(voom-normalized)', isKeyPlot=TRUE)

# Find significant covariates
adjustCovars = designMatVarsToCovars(getCovariatesMap(DM),preAdjustedSigCovars$significantCovars)
```

Significant covariates to adjust at FDR 0.1 are `r adjustCovars`

```{r preAdjustedSigCovars.NULL, fig.width=20, fig.height=10}
preAdjustedSigCovars[["PC_res"]][[1]]$plotData
```

```{r preAdjustedSigCovars.NULL.ALL, fig.width=20, fig.height=10}
preAdjustedSigCovars[["PC_res"]][[2]]$plotData
```

Re-normalise with COVARIATES design matrix.

```{r adjusted.voom.normalisation}
DM1 = getDesignMatrix(METADATA[,adjustCovars],FactorCovariates)
VOOM.ADJUSTED.GENE_EXPRESSION = voom(PROCESSED_COUNTS$filteredExprMatrix, design=DM1$design, plot=T)

# Find PC of re-adjusted gene expression and significant covariates that are highly correlated with PCs
postAdjustedSigCovars = runPCAandPlotCorrelations(VOOM.ADJUSTED.GENE_EXPRESSION$E, DM$design,'all adjusted design(voom-normalized)',isKeyPlot=TRUE)

postAdjustCovars = designMatVarsToCovars(getCovariatesMap(DM),postAdjustedSigCovars$significantCovars)
```

Significant covariates to adjust at FDR 0.1 after fixing covariates are `r postAdjustCovars`

```{r postAdjustedSigCovars, fig.width=20, fig.height=10}
postAdjustedSigCovars[["PC_res"]][[1]]$plotData
```

```{r postAdjustedSigCovars.ALL, fig.width=20, fig.height=10}
postAdjustedSigCovars[["PC_res"]][[2]]$plotData
```

Sanity check: Residualise significant covariates using a linear model and find correlation between PCA of residuals with covariates

```{r calculate.residuals}
RESIDUAL.GENE_EXPRESSION = calcResiduals(VOOM.ADJUSTED.GENE_EXPRESSION$E, DM1$design)

# Find PC of re-adjusted gene expression and significant covariates that are highly correlated with PCs
residualSigCovars = runPCAandPlotCorrelations(RESIDUAL.GENE_EXPRESSION, DM$design,'residual (voom-normalized)',isKeyPlot=TRUE)
```

```{r residualSigCovars, fig.width=20, fig.height=10}
residualSigCovars[["PC_res"]][[1]]$plotData
```

```{r residualSigCovars.ALL, fig.width=20, fig.height=10}
residualSigCovars[["PC_res"]][[2]]$plotData
```

Get differentially expressed genes using limma package after fixing for covariates

```{r differential.expression}
colnames(DM1$design) <- gsub('[[:punct:]]','_',colnames(DM1$design))
colnames(DM1$design) <- gsub(' ','_',colnames(DM1$design))
colnames(DM1$design)[1] <- 'Intercept'

# Fit linear model to normalised expression using limma package
FIT <- lmFit(VOOM.ADJUSTED.GENE_EXPRESSION$E, design = DM1$design)
    
# Make contrast to check differential expression between different differentiation stages
CONT.NAMES <- c(colnames(DM1$design)[1],colnames(DM1$design)[grep('Diffname_short',colnames(DM1$design))])
CONT.NAMES <- combn(CONT.NAMES,2)
CONT.NAMES <- apply(CONT.NAMES,2,paste,collapse='-')
  
CONT <- makeContrasts(contrasts=CONT.NAMES,levels=colnames(FIT$coefficients))

# Refit contrasts
FIT.CONTRAST <- contrasts.fit(FIT,CONT)

# Estimate moderated t-statistics
FIT.CONTRAST <- eBayes(FIT.CONTRAST)

# Obtain all the differential expession combinations
DIFF.EXP <- list()
DIFF.EXP$logFC <- data.frame(row.names = rownames(VOOM.ADJUSTED.GENE_EXPRESSION$E))
DIFF.EXP$adj.P.Val <- data.frame(row.names = rownames(VOOM.ADJUSTED.GENE_EXPRESSION$E))

for (i in CONT.NAMES){
  tmp <- topTable(FIT.CONTRAST, coef=i, number=dim(VOOM.ADJUSTED.GENE_EXPRESSION$E)[1])
  colNames <- gsub('Diffname_short','',gsub('Intercept','DE',i))
  DIFF.EXP$logFC[,paste(colNames,'logFC',sep='-')] <- tmp[rownames(DIFF.EXP$logFC),'logFC']
  DIFF.EXP$adj.P.Val[,paste(colNames,'adj.P.Val',sep='-')] <- tmp[rownames(DIFF.EXP$adj.P.Val),'adj.P.Val']
}
tmp <- cbind(DIFF.EXP$logFC,DIFF.EXP$adj.P.Val)
write.table(DIFF.EXP,file='./DifferentialExpression_DiffnameShort_Express.txt',sep='\t',row.names=T,col.names=T,quote=F)

# Get number of significantly differentialy expressed genes with adj.P.Val <= 0.01 and abs(logFC) >= 2
SIG.EXPR <- DIFF.EXP$adj.P.Val<=0.01 & abs(DIFF.EXP$logFC) >= 2
colnames(SIG.EXPR) <- gsub('-adj.P.Val','',colnames(SIG.EXPR))
SIG.GSETS <- list()
for (i in colnames(SIG.EXPR))
  SIG.GSETS[[i]] <- rownames(SIG.EXPR)[SIG.EXPR[,i]]

print(sapply(SIG.GSETS,length))
```

```{r synapsestore}
if (SYNAPSE_STORE){
  COVARIATES_CODE <- File('./CovariateAnalysis_Express.Rmd',name = 'Covariate Analysis Express',parentId = parentId)
  COVARIATES_CODE <- synStore(COVARIATES_CODE, used = c(COUNT_OBJ$properties$id,METADATA_OBJ@schema),activityName='Covariate Analysis from eXpress', executed='https://github.com/th1vairam/pcbc_c4_analysis/blob/cov_anal/code/Rmd/CovariateAnalysis_Express.Rmd')
  DE <- File('./DifferentialExpression_DiffnameShort_Express.txt',name = 'Differential Expression DiffnameShort Express',parentId = parentId)
  DE <- synStore(DE, used = c(COUNT_OBJ$properties$id,METADATA_OBJ@schema),activityName='Differential Expression Analysis Express', executed=COVARIATES_CODE)
}
```
