---
title: "Differential expression analysis of minfi curated methylation data with mixed effect modeling at each differentiation stages"
author: "Thanneer Perumal"
date: '`r date()`'
output: html_document
---
```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
## It is assumed your working directory is where this file is

# Clear R console screen output
cat("\014")

# Load required libraries
library('synapseClient')
library('RColorBrewer')
library('ggplot2')
library('gplots')
library('limma')
library('edgeR')
library('ctv')
library('psych')
library('reshape2')
library('vcd')
library('erer')
library('fpc')
library('data.table')
library('dplyr')
library('plyr')
library('knitr')
library('stringr')
library('minfi')
library('org.Hs.eg.db')
library('knit2synapse')

## Needs the dev branch
library(rGithubClient)

## Requires ggdendro
devtools::source_url("https://raw.githubusercontent.com/chr1swallace/random-functions/master/R/ggplot-heatmap.R")
# knitToFolderEntity(file = "./DiffExpAnal_methylation_mixedEffects_EachDiffState.Rmd", 
#                    parentId = 'syn4231339',
#                    entityName = 'Differential Expression Analysis Methylation Mixed Effects Model EachDiffState', 
#                    overwrite=F)

synapseLogin()

# source utility files from ../R/lib folder
file.sources = list.files('../R/lib',pattern="*.R")
file.sources = sapply(file.sources,function(x){return(paste('../R/lib',x,sep='/'))})
tmp = sapply(file.sources,source,.GlobalEnv)

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE)
```
### Download data
```{r setup, include=FALSE}
# Input Parameters
COVARIATES_FOLDER = 'syn4589417'

parentId = 'syn4231339'

# Diffstate Names
DiffState = list(DE = 'DE', EB = 'EB', ECTO = 'ECTO', MESO5 = 'MESO5', SC = 'SC')

# Specify factor and continuous covarites pool (adjusted covariates will only be the subset of these covariates)
FactorCovariates = c("Diffname_short", "BeadChip", "Row", "Column", "Cell_Line_Type", 
                     "Tissue_of_Origin", "Reprogramming_Gene_Combination", "Culture_Conditions",  
                     "Other_Conditions_During_Reprogramming", "Donor_Life_Stage", "Gender", 
                     "Originating_Lab", "Donor_ID", "Cell_Type_of_Origin_Level2",
                     "Reprogramming_Vector_Type" )
ContCovariates = c("PassageAtThaw", "PassageAtDNAHarvest")
```
Factor covariates considered for analysis are `r paste(gsub('_','\\\\_',FactorCovariates),collapse=',')`, and continuous covariates considered for the analysis are `r paste(gsub('_','\\\\_',ContCovariates),collapse=',')`

Obtain filtered beta values, covariates and design matrices from synapse
```{r getdata, cache=TRUE, include=FALSE}
All.Files = synQuery(paste0('select * from file where parentId == "',COVARIATES_FOLDER,'"'))

# Function to download expression counts, 
getData <- function(DiffState, All.Files, FactorCovariates, ContCovariates){
  writeLines('------------------------------------------------------------------------------')
  writeLines(paste('Downloading differentiation state:',as.character(DiffState)))
  writeLines('------------------------------------------------------------------------------')# Get filtered beta values
    
  EXP_OBJ = synGet(All.Files$file.id[grep(paste('Filtered Beta Values',DiffState), All.Files$file.name)])
  ALL_USED_IDs = EXP_OBJ$properties$id
  EXP = fread(getFileLocation(EXP_OBJ))
  tmp = read.table(getFileLocation(EXP_OBJ), sep = '\t', header=T, check.names=F, nrow=1)
  setnames(EXP, colnames(EXP), colnames(tmp))
  EXP = as.data.frame(EXP)
  rownames(EXP) = EXP[,1]
  
  writeLines(paste(dim(EXP)[1],'probes and',dim(EXP)[2],'samples'))

  # Get filtered metadata
  COVARIATES_OBJ = synGet(All.Files$file.id[grep(paste('Adjusted Covariates',DiffState), All.Files$file.name)])
  ALL_USED_IDs[length(ALL_USED_IDs)+1] = COVARIATES_OBJ$properties$id
  COVARIATES = read.table(getFileLocation(COVARIATES_OBJ), sep='\t', header=T, stringsAsFactors = F)

  # Get design matrix
  DESIGN_OBJ = synGet(All.Files$file.id[grep(paste('Adjusted Design',DiffState), All.Files$file.name)])
  ALL_USED_IDs[length(ALL_USED_IDs)+1] = DESIGN_OBJ$properties$id
  DESIGN = read.table(getFileLocation(DESIGN_OBJ), sep='\t', header=T, stringsAsFactors = F, row.names=1)
  DESIGN = data.matrix(DESIGN)

  # Get probe related metadata
  fData = get450KProbeMapping(EXP[['methProbeID']])
  
  return(list(ALL_USED_IDs = ALL_USED_IDs,
              EXP = EXP,
              COVARIATES = COVARIATES,
              DESIGN = DESIGN))
}

# Get data from synapse
DATA = lapply(DiffState, getData, All.Files, FactorCovariates, ContCovariates)
```
### Differential Methylation Analysis At Each Differentiation Stage
Obtain differentially expressed genes with the following covariates (as obtained from covariates analysis) in the model
* Differentation Stage: SC
  * Fixed Effects: Gender, Culture_Conditions
  * Random Effects: Donor ID

Perform differential expression
```{r diff.exp, fig.width=6, fig.height=6}
# Function to calculate differential expression at each differentiation stages
getDifferentialExpression <- function(DiffState, DATA){
  writeLines('------------------------------------------------------------------------------')
  writeLines(paste('Performing differential expression for',as.character(DiffState)))
  writeLines('------------------------------------------------------------------------------')
  
  # Post adjusted design matrix
  DESIGN = DATA[[DiffState]]$DESIGN
  
  if (dim(DESIGN)[1] == 0){
    return(list(DIFF.EXP = c(),
                FIT = c(),
                FIT.CONTRAST = c(),
                CORRELATION = c()))    
  }
    
  # Remove first column from beta matrix
  EXP = DATA[[DiffState]]$EXP[,-(1)]

  # Estimate correlation of random effects
  correlation <- duplicateCorrelation(EXP, design = DESIGN, block=DATA[[DiffState]]$COVARIATES$Donor_ID)
  
  # Fit linear model
  FIT = lmFit(EXP, 
              design = DESIGN, 
              block = DATA[[DiffState]]$COVARIATES$Donor_ID, 
              correlation = correlation$cor)

  # Make contrast to check differential expression between different differentiation stages
  CONT.TO.CHECK = setdiff(colnames(DATA[[DiffState]]$COVARIATES), 'Row')
  CONT.TO.CHECK = CONT.TO.CHECK[!!sapply(CONT.TO.CHECK, function(x, DESIGN) length(grep(x,colnames(DESIGN))), DESIGN)]
    
  if (length(CONT.TO.CHECK) == 0){
    return(list(DIFF.EXP = c(),
                FIT = FIT,
                FIT.CONTRAST = c(),
                CORRELATION = correlation))    
  }
  
  ALL.CONT = c()
  for (name in CONT.TO.CHECK){
    CONT.NAMES <- colnames(DESIGN)[grep(name,colnames(DESIGN))]
    CONT.NAMES <- combn(CONT.NAMES,2)
    CONT.NAMES <- apply(CONT.NAMES,2,paste,collapse='-')
    
    CONT <- makeContrasts(contrasts=CONT.NAMES,levels=colnames(FIT$coefficients))
    colnames(CONT) <- sapply(colnames(CONT),function(x, name){x <- gsub(name,'',x);
                                                              x <- gsub('-','_vs_',x);
                                                              x <- paste(DiffState,x,sep='__')}, name)
    ALL.CONT = cbind(ALL.CONT, CONT)
  }

  # Refit contrasts
  FIT.CONTRAST <- contrasts.fit(FIT,ALL.CONT)
    
  # Estimate moderated t-statistics
  FIT.CONTRAST <- eBayes(FIT.CONTRAST)
  
  # Obtain all the differential expession combinations
  DIFF.EXP <- list()
  DIFF.EXP$logFC <- data.frame(row.names = rownames(EXP))
  DIFF.EXP$adj.P.Val <- data.frame(row.names = rownames(EXP))
  DIFF.EXP$SIG.SETS <- data.frame()

  for (name in colnames(ALL.CONT)){
    tmp <- topTable(FIT.CONTRAST, coef=name, number=dim(EXP)[1])    
    DIFF.EXP$logFC[,name] <- tmp[rownames(DIFF.EXP$logFC),'logFC']
    DIFF.EXP$adj.P.Val[,name] <- tmp[rownames(DIFF.EXP$adj.P.Val),'adj.P.Val'] 
    
    DIFF.EXP$SIG.SETS <- rbind(DIFF.EXP$SIG.SETS,
                               getUpDownGenes(DIFF.EXP$adj.P.Val[,name], 
                                              DIFF.EXP$logFC[,name], 
                                              rownames(DIFF.EXP$logFC), 
                                              name, 
                                              FC_CUTOFF = 0))
    
  }
  colnames(DIFF.EXP$SIG.SETS)[1] = 'methProbeIDs'  
  DIFF.EXP$SIG.SETS = merge(DIFF.EXP$SIG.SETS, fData$Annotation, by = 'methProbeIDs', all.x=T, all.y = F)

  # Get number of significantly differentialy expressed genes with adj.P.Val <= 0.05 and log2FC >= 0.5
  DIFF.EXP$SIG.EXP.POS <- DIFF.EXP$adj.P.Val<=0.05 & DIFF.EXP$logFC >= 0.5
  DIFF.EXP$NUM.SIG.EXP.POS <- colSums(DIFF.EXP$SIG.EXP.POS, na.rm =T)
    
  # Get number of significantly differentialy expressed genes with adj.P.Val <= 0.05 and logFC <= -0.5
  DIFF.EXP$SIG.EXP.NEG <- DIFF.EXP$adj.P.Val<=0.05 & DIFF.EXP$logFC <= -0.5
  DIFF.EXP$NUM.SIG.EXP.NEG <- colSums(DIFF.EXP$SIG.EXP.NEG, na.rm=T)

  return(list(DIFF.EXP = DIFF.EXP,
              FIT = FIT,
              FIT.CONTRAST = FIT.CONTRAST,
              CORRELATION = correlation))
}
# Compute differential expression at each differentiation stages
DEXP = lapply(DiffState, getDifferentialExpression, DATA)
```
Number of differentially expressed genes at SC stage for adj.P.Val <= 0.05 and logFC >= 0.5 or <= -0.5
```{r print.imp.genes.DE}
tmp1 = rownameToFirstColumn(as.data.frame(DEXP$SC$DIFF.EXP$NUM.SIG.EXP.POS),'ContrastNames')
colnames(tmp1) = c('Contrast Names','Up Regulated')
tmp2 = rownameToFirstColumn(as.data.frame(DEXP$SC$DIFF.EXP$NUM.SIG.EXP.NEG),'ContrastNames')
colnames(tmp2) = c('Contrast Names','Down Regulated')

kable(merge(tmp1,tmp2,by = 'Contrast Names', all = T), row.names = F)
```
```{r merge.results}
logFC = join_all(lapply(DEXP, function(x){rownameToFirstColumn(x$DIFF.EXP$logFC, 'methProbeIDs')}), 
                 by = 'methProbeIDs', match = 'all')
adjPVal = join_all(lapply(DEXP, function(x){rownameToFirstColumn(x$DIFF.EXP$adj.P.Val, 'methProbeIDs')}), 
                 by = 'methProbeIDs', match = 'all')
comparisonSet = rbindlist(lapply(DEXP, function(x){x$DIFF.EXP$SIG.SETS}))
ALL_USED_IDs = as.character(unlist(lapply(DATA, function(x){return(x$ALL_USED_IDs)})))
```

### Store files in synapse
```{r synapse.store, include = FALSE, eval=FALSE, cache=FALSE}
activityName='Differential Expression Analysis of minfi processed methylation data with mixed effects model at each differentiation state'
  
thisFileName <- 'DiffExpAnal_methylation_mixedEffects_EachDiffState.Rmd'
  
# Github link
thisRepo <- getRepo(repository = "th1vairam/pcbc_c4_analysis", 
                    ref="branch", 
                    refName='methyl')

thisFile <- getPermlink(repository = thisRepo,
                        repositoryPath=paste0('code/Rmd/', thisFileName))
    
# Make folder and populate with wiki
CODE <- Folder(name = 'Differential Expression Analysis Methylation Mixed Effects Model EachDiffState', parentId = parentId)
CODE <- synStore(CODE)

# Store results: logFC
write.table(logFC,file='./DiffExpAnal_methylation_mixedEffects_EachDiffState_logFC.tsv',sep='\t',row.names=F,quote=F)
FC_OBJ <- File('./DiffExpAnal_methylation_mixedEffects_EachDiffState_logFC.tsv', name = 'logFC', parentId = CODE$properties$id)
FC_OBJ <- synStore(FC_OBJ, used = ALL_USED_IDs, activityName = activityName, executed=thisFile)

# Store results: adjPVal
write.table(adjPVal,file='./DiffExpAnal_methylation_mixedEffects_EachDiffState_adjPVal.tsv',sep='\t',row.names=F,quote=F)
PVAL_OBJ <- File('./DiffExpAnal_methylation_mixedEffects_EachDiffState_adjPVal.tsv', name = 'Adjusted PValue', parentId = CODE$properties$id)
PVAL_OBJ <- synStore(PVAL_OBJ, used = ALL_USED_IDs, activityName = activityName, executed=thisFile)

# Store results: gctFormat
write.table(Comparisons, file='./DiffExpAnal_methylation_mixedEffects_EachDiffState_comparisons.tsv',sep='\t',row.names=F,quote=F)
COMP_OBJ <- File('./DiffExpAnal_methylation_mixedEffects_EachDiffState_comparisons.tsv', name = 'Differentially Expressed Genes comparison list', parentId = CODE$properties$id)
COMP_OBJ <- synStore(COMP_OBJ, used = ALL_USED_IDs, activityName = activityName, executed=thisFile)
```
|  *Results*                                           |  *SynapseID*                                                                 |
|  --------------------------------------------------- |   -------------------------------------------------------------------------- | 
|  logFC                                               |  `r paste(FC_OBJ$properties$id,FC_OBJ$properties$versionNumber,sep='.')`     |
|  Adjusted P Value                                    |  `r paste(PVAL_OBJ$properties$id,PVAL_OBJ$properties$versionNumber,sep='.')` |
|  Differentially Expressed Genes (comparison list)    |  `r paste(COMP_OBJ$properties$id,COMP_OBJ$properties$versionNumber,sep='.')` |


### Executed: Source code
`r paste0('[Source R Markdown](',print(thisFile),')')`