---
title: "Differential expression analysis for eXpress aligned data with mixed effects modeling"
author: "Thanneer Perumal"
date: "`r date()`"
output: html_document
---

```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
## It is assumed your working directory is where this file is
# Load required libraries
library('synapseClient')
library('limma')
library('edgeR')
library('RColorBrewer')
library('ctv')
library('ggplot2')
library('psych')
library('reshape2')
library('gplots')
library('vcd')
library('erer')
library('fpc')
library('data.table')
library('dplyr')
library('stringr')
library('knitr')
library('goseq')
library('plyr')

## Needs the dev branch
library(rGithubClient)

## Requires ggdendro
devtools::source_url("https://raw.githubusercontent.com/chr1swallace/random-functions/master/R/ggplot-heatmap.R")

# source('/mnt/Github/knit2synapse-1//R/knitFile2Synapse.R')
# knitFile2Synapse(file = "./DiffExpAnal_Express_mixedEffects.Rmd", owner = 'syn3972059', name = "Differential Expression Analysis Express Mixed Effects",overwrite=F)

synapseLogin()

# source utility files from ../R/lib folder
file.sources = list.files('../R/lib',pattern="*.R")
file.sources = sapply(file.sources,function(x){return(paste('../R/lib',x,sep='/'))})
tmp = sapply(file.sources,source,.GlobalEnv)

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE)
```

### Download data

```{r setup, include=FALSE}
# Input Parameters
DESIGN_ID = 'syn3943971';
EXPR_ID = 'syn3943951';
WEIGHTS_ID = 'syn3943964';
COVARIATES_ID = 'syn3943968';

SYNAPSE_STORE = T
parentId = 'syn3471223'

# Specify factor and continuous covarites pool (adjusted covariates will only be the subset of these covariates)
FactorCovariates = c('Diffname_short','run','lane','index','Cell_Type_of_Origin', 'Cell_Line_of_Origin', 'Tissue_of_Origin', 'Reprogramming_Vector_Type','Reprogramming_Gene_Combination', 'Donor_Life_Stage','Originating_Lab','Gender','Donor_ID','Disease','Race','Ethnicity','Cell_Line_Type','PassageAtHarvest')
ContCovariates = NULL
```
Synapse id of design, expression, weights and covariates matrix used for the analysis are `r DESIGN_ID`,`r EXPR_ID`,`r WEIGHTS_ID`, and `r COVARIATES_ID`

```{r getdata, cache=TRUE, include=FALSE}
# Get design matrix
DESIGN_OBJ = synGet(DESIGN_ID);
ALL_USED_IDs = c(DESIGN_OBJ$properties$id);
DESIGN = read.table(getFileLocation(DESIGN_OBJ),sep='\t',header=T,row.names=1, check.names=FALSE)
colnames(DESIGN) = gsub("[[:punct:]]", "_",colnames(DESIGN))
colnames(DESIGN) = gsub(" ", "_",colnames(DESIGN))

# Get expression matrix
EXPR_OBJ = synGet(EXPR_ID);
ALL_USED_IDs = c(ALL_USED_IDs,EXPR_OBJ$properties$id);
EXPR = read.table(getFileLocation(EXPR_OBJ),sep='\t',header=T,row.names=1, check.names=FALSE)

# Get weighting matrix
WEIGHTS_OBJ = synGet(WEIGHTS_ID);
ALL_USED_IDs = c(ALL_USED_IDs,WEIGHTS_OBJ$properties$id);
WEIGHTS = read.table(getFileLocation(WEIGHTS_OBJ),sep='\t',header=T,row.names=1, check.names=FALSE)

# Get weighting matrix
COVARIATES_OBJ = synGet(COVARIATES_ID);
ALL_USED_IDs = c(ALL_USED_IDs,COVARIATES_OBJ$properties$id);
COVARIATES = read.table(getFileLocation(COVARIATES_OBJ),sep='\t',header=T,row.names=1, check.names=FALSE)
COVARIATES <- as.data.frame(lapply(COVARIATES, function(cols){ if(is.factor(cols)){levels(cols) <- str_replace_all(as.character(levels(cols)),'[^[:alnum:]]|\\s','_')};return(cols)}))
```

Get differentially expressed genes using limma package with following co-efficients `r colnames(DESIGN)` in linear model

Degenerate columns in the new design are `r linColumnFinder(DESIGN)$relations`

### Differential expression 

##### Differentiation stages
```{r diffstate, cache=TRUE, include=FALSE}
# Calculate correlation between donors
correlation <- duplicateCorrelation(EXPR, design=DESIGN, block=COVARIATES$Donor_ID)

# Fit linear model using mixed effects design
FIT = lmFit(EXPR, design=DESIGN, weights = WEIGHTS, block=COVARIATES$Donor_ID, correlation = correlation$cor)

# Make contrast to check differential expression between different differentiation stages
CONT.NAMES <- colnames(DESIGN)[grep('Diffname_short',colnames(DESIGN))]
CONT.NAMES <- combn(CONT.NAMES,2)
CONT.NAMES <- apply(CONT.NAMES,2,paste,collapse='-')

CONT <- makeContrasts(contrasts=CONT.NAMES,levels=colnames(FIT$coefficients))
colnames(CONT) <- sapply(colnames(CONT),function(x){x <- gsub('Diffname_short','',x);})

# Refit contrasts
FIT.CONTRAST <- contrasts.fit(FIT,CONT)
  
# Estimate moderated t-statistics
FIT.CONTRAST <- eBayes(FIT.CONTRAST)
  
# Obtain all the differential expession combinations
DIFFNAME <- list()
DIFFNAME$logFC <- data.frame(row.names = rownames(EXPR))
DIFFNAME$adj.P.Val <- data.frame(row.names = rownames(EXPR))
  
for (i in colnames(CONT)){
  tmp <- topTable(FIT.CONTRAST, coef=i, number=dim(EXPR)[1])    
  DIFFNAME$logFC[,str_replace_all(i,"Diffname_short","")] <- tmp[rownames(DIFFNAME$logFC),'logFC']
  DIFFNAME$adj.P.Val[,str_replace_all(i,"Diffname_short","")] <- tmp[rownames(DIFFNAME$adj.P.Val),'adj.P.Val']
}

# Get number of significantly differentialy expressed genes with adj.P.Val <= 0.05 and logFC >= 1
DIFFNAME$SIG.EXP.POS <- DIFFNAME$adj.P.Val<=0.05 & DIFFNAME$logFC >= 1   
DIFFNAME$NUM.SIG.EXP.POS <- colSums(DIFFNAME$SIG.EXP.POS)

# Get number of significantly differentialy expressed genes with adj.P.Val <= 0.05 and logFC <= -1
DIFFNAME$SIG.EXP.NEG <- DIFFNAME$adj.P.Val<=0.05 & DIFFNAME$logFC <= -1   
DIFFNAME$NUM.SIG.EXP.NEG <- colSums(DIFFNAME$SIG.EXP.NEG)

save(DIFFNAME,file='DifferentialExpressionDiffName_mixedEffects.RData')
```

Number of differentially expressed genes at FDR <= 0.05 and logFC >= 1 or logFC <= -1
```{r significant.sets1,cache=TRUE}
tmp <- cbind(as.data.frame(DIFFNAME$NUM.SIG.EXP.POS),as.data.frame(DIFFNAME$NUM.SIG.EXP.NEG))
colnames(tmp) <- c('NUMBER OF SIG. EXP. GENES (+ve)','NUMBER OF SIG. EXP. GENES (-ve)')
kable(tmp)
```

#### Gender 
```{r gender, cache=TRUE}
# Post adjusted design matrix
DESIGN = getDesignMatrix(COVARIATES[,c("Gender","Diffname_short","Donor_Life_Stage","run","Reprogramming_Vector_Type"),drop=F],Intercept = F)
DESIGN = DESIGN$design[,linColumnFinder(DESIGN$design)$indepCols]
colnames(DESIGN) <- str_replace_all(colnames(DESIGN),'[[:punct:]]','_')
colnames(DESIGN) <- str_replace_all(colnames(DESIGN),' ','_')

# Calculate correlation between donors
correlation <- duplicateCorrelation(EXPR, design=DESIGN, block=COVARIATES$Donor_ID)

# Fit linear model using mixed effects design
FIT = lmFit(EXPR, design=DESIGN, weights = WEIGHTS, block=COVARIATES$Donor_ID, correlation = correlation$cor)

# Make contrast to check differential expression between different differentiation stages
CONT.NAMES <- colnames(DESIGN)[grep('Gender',colnames(DESIGN))]
CONT.NAMES <- combn(CONT.NAMES,2)
CONT.NAMES <- apply(CONT.NAMES,2,paste,collapse='-')

CONT <- makeContrasts(contrasts=CONT.NAMES,levels=colnames(FIT$coefficients))
colnames(CONT) <- sapply(colnames(CONT),function(x){x <- gsub('Gender','',x);})

# Refit contrasts
FIT.CONTRAST <- contrasts.fit(FIT,CONT)
  
# Estimate moderated t-statistics
FIT.CONTRAST <- eBayes(FIT.CONTRAST)
  
# Obtain all the differential expession combinations
GENDER <- list()
GENDER$logFC <- data.frame(row.names = rownames(EXPR))
GENDER$adj.P.Val <- data.frame(row.names = rownames(EXPR))
  
for (i in colnames(CONT)){
  tmp <- topTable(FIT.CONTRAST, coef=i, number=dim(EXPR)[1])    
  GENDER$logFC[,str_replace_all(i,"Gender","")] <- tmp[rownames(GENDER$logFC),'logFC']
  GENDER$adj.P.Val[,str_replace_all(i,"Gender","")] <- tmp[rownames(GENDER$adj.P.Val),'adj.P.Val']
}

# Get number of significantly differentialy expressed genes with adj.P.Val <= 0.05 and logFC >= 1
GENDER$SIG.EXP.POS <- GENDER$adj.P.Val<=0.05 & GENDER$logFC >= 1
GENDER$NUM.SIG.EXP.POS <- colSums(GENDER$SIG.EXP.POS)

# Get number of significantly differentialy expressed genes with adj.P.Val <= 0.05 and logFC <= -1
GENDER$SIG.EXP.NEG <- GENDER$adj.P.Val<=0.05 & GENDER$logFC <= -1
GENDER$NUM.SIG.EXP.NEG <- colSums(GENDER$SIG.EXP.NEG)

save(GENDER,file='DifferentialExpressionGender_mixedEffects.RData')
```

Number of differentially expressed genes at FDR <= 0.05 and absolute logFC >= 1 or logFC <= -1
```{r significant.sets2,cache=TRUE}
tmp <- cbind(as.data.frame(GENDER$NUM.SIG.EXP.POS),as.data.frame(GENDER$NUM.SIG.EXP.NEG))
colnames(tmp) <- c('NUMBER OF SIG. EXP. GENES (+ve)','NUMBER OF SIG. EXP. GENES (-ve)')
kable(tmp)
```

#### Differentiation state and gender

```{r diffstate.gender, cache=TRUE}
# Post adjusted design matrix
COVARIATES.NEW <- COVARIATES
COVARIATES.NEW <- mutate(COVARIATES.NEW,Diff.Gender = paste(Diffname_short,Gender,sep='.'))
COVARIATES.NEW$Diff.Gender <- factor(COVARIATES.NEW$Diff.Gender)

DESIGN = getDesignMatrix(COVARIATES.NEW[,c("Diff.Gender","run","Donor_Life_Stage","Reprogramming_Vector_Type"),drop=F],Intercept = F)
DESIGN = DESIGN$design[,linColumnFinder(DESIGN$design)$indepCols]
colnames(DESIGN) <- str_replace_all(colnames(DESIGN),'[[:punct:]]','_')
colnames(DESIGN) <- str_replace_all(colnames(DESIGN),' ','_')

# Calculate correlation between donors
correlation <- duplicateCorrelation(EXPR, design=DESIGN, block=COVARIATES.NEW$Donor_ID)

# Fit linear model using mixed effects design
FIT = lmFit(EXPR, design=DESIGN, weights = WEIGHTS, block=COVARIATES.NEW$Donor_ID, correlation = correlation$cor)

# Make contrast to check differential expression between different differentiation stages
CONT.NAMES <- colnames(DESIGN)[grep('female',colnames(DESIGN))]
CONT.NAMES <- combn(CONT.NAMES,2)
CONT.NAMES1 <- apply(CONT.NAMES,2,paste,collapse='-')
CONT.NAMES2 <- sapply(CONT.NAMES1,str_replace_all,'female','male')
CONT.NAMES <- paste0('(',CONT.NAMES1,')-(',CONT.NAMES2,')')

CONT <- makeContrasts(contrasts=CONT.NAMES,levels=colnames(FIT$coefficients))
colnames(CONT) <- sapply(colnames(CONT),function(x){x <- gsub('Diff_Gender','',x);})

# Refit contrasts
FIT.CONTRAST <- contrasts.fit(FIT,CONT)
  
# Estimate moderated t-statistics
FIT.CONTRAST <- eBayes(FIT.CONTRAST)
  
# Obtain all the differential expession combinations
DIFFNAME.GENDER <- list()
DIFFNAME.GENDER$logFC <- data.frame(row.names = rownames(EXPR))
DIFFNAME.GENDER$adj.P.Val <- data.frame(row.names = rownames(EXPR))
  
for (i in colnames(CONT)){
  tmp <- topTable(FIT.CONTRAST, coef=i, number=dim(EXPR)[1])    
  DIFFNAME.GENDER$logFC[,str_replace_all(i,"DIFFNAME.GENDER","")] <- tmp[rownames(DIFFNAME.GENDER$logFC),'logFC']
  DIFFNAME.GENDER$adj.P.Val[,str_replace_all(i,"DIFFNAME.GENDER","")] <- tmp[rownames(DIFFNAME.GENDER$adj.P.Val),'adj.P.Val']
}

# Get number of significantly differentialy expressed genes with adj.P.Val <= 0.05 and logFC >= 1
DIFFNAME.GENDER$SIG.EXP.POS <- DIFFNAME.GENDER$adj.P.Val<=0.05 & DIFFNAME.GENDER$logFC >= 1   
DIFFNAME.GENDER$NUM.SIG.EXP.POS <- colSums(DIFFNAME.GENDER$SIG.EXP.POS)

# Get number of significantly differentialy expressed genes with adj.P.Val <= 0.05 and logFC <= -1
DIFFNAME.GENDER$SIG.EXP.NEG <- DIFFNAME.GENDER$adj.P.Val<=0.05 & DIFFNAME.GENDER$logFC <= -1   
DIFFNAME.GENDER$NUM.SIG.EXP.NEG <- colSums(DIFFNAME.GENDER$SIG.EXP.NEG)

save(DIFFNAME.GENDER,file='DifferentialExpressionDiffnameGender_mixedEffects.RData')
```

Number of differentially expressed genes at FDR <= 0.05 and absolute logFC >= 1
```{r significant.sets3,cache=TRUE}
tmp <- cbind(as.data.frame(DIFFNAME.GENDER$NUM.SIG.EXP.POS),as.data.frame(DIFFNAME.GENDER$NUM.SIG.EXP.NEG))
colnames(tmp) <- c('NUMBER OF SIG. EXP. GENES (+ve)','NUMBER OF SIG. EXP. GENES (-ve)')
kable(tmp)
```

#### Reprogramming vector type

```{r reprogramming.vector, cache=TRUE}
# Get design matrix
NEW.DESIGN <- getDesignMatrix(dplyr::select(COVARIATES,one_of('Reprogramming_Vector_Type','Diffname_short','run','Donor_Life_Stage','Gender')), Intercept = F)
tmp <- linColumnFinder(NEW.DESIGN$design)
NEW.DESIGN$design <- NEW.DESIGN$design[,tmp$indepCols]
colnames(NEW.DESIGN$design) = gsub("[[:punct:]]", "_",colnames(NEW.DESIGN$design))
colnames(NEW.DESIGN$design) = gsub(" ", "_",colnames(NEW.DESIGN$design))

# Calculate correlation between donors
correlation <- duplicateCorrelation(EXPR, design=NEW.DESIGN$design, block=COVARIATES$Donor_ID)
  
# Fit linear model using mixed effects design
FIT = lmFit(EXPR, design=NEW.DESIGN$design, weights = WEIGHTS, block=COVARIATES$Donor_ID, correlation = correlation$cor)
                  
# Make contrast to check differential expression between different differentiation stages
CONT.NAMES <- colnames(NEW.DESIGN$design)[grep('Reprogramming_Vector_Type',colnames(NEW.DESIGN$design))]
CONT.NAMES <- combn(CONT.NAMES,2)
CONT.NAMES <- apply(CONT.NAMES,2,paste,collapse='-')

CONT <- makeContrasts(contrasts=CONT.NAMES,levels=colnames(FIT$coefficients))
colnames(CONT) <- sapply(colnames(CONT),function(x){x <- gsub('Reprogramming_Vector_Type','',x);})

# Refit contrasts
FIT.CONTRAST <- contrasts.fit(FIT,CONT)

# Estimate moderated t-statistics
FIT.CONTRAST <- eBayes(FIT.CONTRAST)

# Obtain all the differential expession combinations
REPVECTOR <- list()
REPVECTOR$logFC <- data.frame(row.names = rownames(EXPR))
REPVECTOR$adj.P.Val <- data.frame(row.names = rownames(EXPR)) 

# Get differential expression
for (i in colnames(CONT)){
  tmp <- topTable(FIT.CONTRAST, coef=i, number=dim(EXPR)[1])
  REPVECTOR$logFC[,i] <- tmp[rownames(REPVECTOR$logFC),'logFC']
  REPVECTOR$adj.P.Val[,i] <- tmp[rownames(REPVECTOR$adj.P.Val),'adj.P.Val']
}

# Get number of significantly differentialy expressed genes with adj.P.Val <= 0.05 and logFC >= 1
REPVECTOR$SIG.EXP.POS <- REPVECTOR$adj.P.Val<=0.05 & REPVECTOR$logFC >= 1   
REPVECTOR$NUM.SIG.EXP.POS <- colSums(REPVECTOR$SIG.EXP.POS)

# Get number of significantly differentialy expressed genes with adj.P.Val <= 0.05 and logFC <= -1
REPVECTOR$SIG.EXP.NEG <- REPVECTOR$adj.P.Val<=0.05 & REPVECTOR$logFC <= -1   
REPVECTOR$NUM.SIG.EXP.NEG <- colSums(REPVECTOR$SIG.EXP.NEG)

save(REPVECTOR,file='DifferentialExpressionReprogrammingVectorType.RData')
```

Number of differentially expressed genes at FDR <= 0.05 and logFC >= 1 or logFC <= -1
```{r significant.sets4,cache=TRUE}
tmp <- cbind(as.data.frame(REPVECTOR$NUM.SIG.EXP.POS),as.data.frame(REPVECTOR$NUM.SIG.EXP.NEG))
colnames(tmp) <- c('NUMBER OF SIG. EXP. GENES (+ve)','NUMBER OF SIG. EXP. GENES (-ve)')
kable(tmp)
```

#### Reprogramming vector type and gender

```{r reprogramming.vector.gender, cache=TRUE}
# Post adjusted design matrix
COVARIATES.NEW <- COVARIATES
COVARIATES.NEW <- mutate(COVARIATES.NEW,Repvect.Gender = paste(Reprogramming_Vector_Type,Gender,sep='.'))
COVARIATES.NEW$Repvect.Gender <- factor(COVARIATES.NEW$Repvect.Gender)

# Get design matrix
NEW.DESIGN <- getDesignMatrix(dplyr::select(COVARIATES.NEW,one_of('Repvect.Gender','Diffname_short','run','Donor_Life_Stage')), Intercept = F)
tmp <- linColumnFinder(NEW.DESIGN$design)
NEW.DESIGN$design <- NEW.DESIGN$design[,tmp$indepCols]
colnames(NEW.DESIGN$design) = gsub("[[:punct:]]", "_",colnames(NEW.DESIGN$design))
colnames(NEW.DESIGN$design) = gsub(" ", "_",colnames(NEW.DESIGN$design))

# Calculate correlation between donors
correlation <- duplicateCorrelation(EXPR, design=NEW.DESIGN$design, block=COVARIATES$Donor_ID)
  
# Fit linear model using mixed effects design
FIT = lmFit(EXPR, design=NEW.DESIGN$design, weights = WEIGHTS, block=COVARIATES$Donor_ID, correlation = correlation$cor)
                  
# Make contrast to check differential expression between different reprogramming vectors with male and female cells
CONT.NAMES1 <- colnames(NEW.DESIGN$design)[grep('_female',colnames(NEW.DESIGN$design))]
CONT.NAMES1 <- combn(CONT.NAMES1,2)
CONT.NAMES1 <- apply(CONT.NAMES1,2,paste,collapse='-')
CONT.NAMES2 <- sapply(CONT.NAMES1,str_replace_all,'female','male')
CONT.NAMES <- paste0('(',CONT.NAMES1,')-(',CONT.NAMES2,')')

CONT <- makeContrasts(contrasts=CONT.NAMES,levels=colnames(FIT$coefficients))
colnames(CONT) <- sapply(colnames(CONT),function(x){x <- gsub('Repvect_Gender','',x);})

# Refit contrasts
FIT.CONTRAST <- contrasts.fit(FIT,CONT)

# Estimate moderated t-statistics
FIT.CONTRAST <- eBayes(FIT.CONTRAST)

# Obtain all the differential expession combinations
REPVECTOR.GENDER <- list()
REPVECTOR.GENDER$logFC <- data.frame(row.names = rownames(EXPR))
REPVECTOR.GENDER$adj.P.Val <- data.frame(row.names = rownames(EXPR)) 

# Get differential expression
for (i in colnames(CONT)){
  tmp <- topTable(FIT.CONTRAST, coef=i, number=dim(EXPR)[1])
  REPVECTOR.GENDER$logFC[,i] <- tmp[rownames(REPVECTOR.GENDER$logFC),'logFC']
  REPVECTOR.GENDER$adj.P.Val[,i] <- tmp[rownames(REPVECTOR.GENDER$adj.P.Val),'adj.P.Val']
}

# Get number of significantly differentialy expressed genes with adj.P.Val <= 0.05 and logFC >= 1
REPVECTOR.GENDER$SIG.EXP.POS <- REPVECTOR.GENDER$adj.P.Val<=0.05 & REPVECTOR.GENDER$logFC >= 1   
REPVECTOR.GENDER$NUM.SIG.EXP.POS <- colSums(REPVECTOR.GENDER$SIG.EXP.POS)

# Get number of significantly differentialy expressed genes with adj.P.Val <= 0.05 and logFC <= -1
REPVECTOR.GENDER$SIG.EXP.NEG <- REPVECTOR.GENDER$adj.P.Val<=0.05 & REPVECTOR.GENDER$logFC <= -1   
REPVECTOR.GENDER$NUM.SIG.EXP.NEG <- colSums(REPVECTOR.GENDER$SIG.EXP.NEG)

save(REPVECTOR.GENDER,file='DifferentialExpressionReprogrammingVectorType.RData')
```

Number of differentially expressed genes at FDR <= 0.05 and logFC >= 1 or logFC <= -1
```{r significant.sets5,cache=TRUE}
tmp <- cbind(as.data.frame(REPVECTOR.GENDER$NUM.SIG.EXP.POS),as.data.frame(REPVECTOR.GENDER$NUM.SIG.EXP.NEG))
colnames(tmp) <- c('NUMBER OF SIG. EXP. GENES (+ve)','NUMBER OF SIG. EXP. GENES (-ve)')
kable(tmp)
```

#### Get pvalue and fold changes of some important genes

```{r print.imp.genes}
FC <- join_all(list(rownameToFirstColumn(DIFFNAME$logFC,'GeneNames'),
                    rownameToFirstColumn(DIFFNAME.GENDER$logFC,'GeneNames'),
                    rownameToFirstColumn(REPVECTOR$logFC,'GeneNames'),
                    rownameToFirstColumn(REPVECTOR.GENDER$logFC,'GeneNames')),
               by = 'GeneNames',
               match = 'all')

PVAL <- join_all(list(rownameToFirstColumn(DIFFNAME$adj.P.Val,'GeneNames'),
                      rownameToFirstColumn(DIFFNAME.GENDER$adj.P.Val,'GeneNames'),
                      rownameToFirstColumn(REPVECTOR$adj.P.Val,'GeneNames'),
                      rownameToFirstColumn(REPVECTOR.GENDER$adj.P.Val,'GeneNames')),
                 by = 'GeneNames',
                 match = 'all')

SIG <- join_all(list(rownameToFirstColumn(DIFFNAME$SIG.EXP.POS+DIFFNAME$SIG.EXP.NEG,'GeneNames'),
                     rownameToFirstColumn(DIFFNAME.GENDER$SIG.EXP.POS+DIFFNAME.GENDER$SIG.EXP.NEG,'GeneNames'),
                     rownameToFirstColumn(REPVECTOR$SIG.EXP.POS+REPVECTOR$SIG.EXP.NEG,'GeneNames'),
                     rownameToFirstColumn(REPVECTOR.GENDER$SIG.EXP.POS+REPVECTOR.GENDER$SIG.EXP.NEG,'GeneNames')),
                by = 'GeneNames',
                match = 'all')

print('Fold change:')
kable(filter(FC, GeneNames %in% c('LARGE','FRG1','FRG1B','PEG10','SPEG')))
print('Adjusted Pvalue:')
kable(filter(PVAL, GeneNames %in% c('LARGE','FRG1','FRG1B','PEG10','SPEG')))
```

#### Gender effect in differentially expressed genes

ECDFs of Ccoefficient of variation between male and female cells at each differentiation stages
```{r gender.effect, cache=TRUE,fig.height=4,fig.width=4}
ind <- which(rowSums(SIG[,2:16]) != 0)

GENDER.VAR <- list()
for (Diffname in levels(COVARIATES$Diffname_short)){
  for (Gender in levels(COVARIATES$Gender)){
    TEMP.EXPR <- EXPR[ind, COVARIATES$Diffname_short == Diffname & COVARIATES$Gender == Gender]
    GENDER.VAR[[Diffname]][[Gender]] <- apply(TEMP.EXPR,1,mean)/apply(TEMP.EXPR,1,sd)
  }
  plotdata <- as.data.frame(rbind(cbind(GENDER.VAR[[Diffname]]$male,'male'),cbind(GENDER.VAR[[Diffname]]$female,'female')))
  plotdata[,1] <- as.numeric(as.character(plotdata[,1]))
  colnames(plotdata) <- c('CV','Gender')
  GENDER.VAR[[Diffname]]$plot <- ggplot(plotdata, aes(x = CV)) + stat_ecdf(aes(colour = Gender)) + ggtitle(Diffname)  
  print(GENDER.VAR[[Diffname]]$plot)
}
```
No significant differnece between male vs female stem cells at each of the differentiation stages in terms of their variance

Perform enrichment analysis for differentiation
```{r enrichment.analysis, include=FALSE}
# Write input files to folder
inputFolderName <- paste0(getwd(),'/GOElite_input')
if (file.exists(inputFolderName))
  unlink(inputFolderName, recursive = T, force = T)
dir.create(inputFolderName, showWarnings = TRUE, recursive = FALSE)

# Function to write input files for GO-Elite
writeInputFiles <- function(cont.name, FC, PVAL, inputFolderName){
  input <- FC[(FC[,cont.name]>=1 & PVAL[,cont.name]<=0.05),c('GeneNames',cont.name),drop=F]
  input <- dplyr::mutate(input,SystemCode = 'Sy')
  input <- input[,c(1,3,2)]
  colnames(input)[3] <- 'logFC'
  
  Genes <- cbind(paste(str_replace_all(cont.name,'[^[:alnum:]]','_'),'up',sep='_'),
                 paste(input$GeneNames,collapse=','))
  
  FName <- paste(str_replace_all(cont.name,'[^[:alnum:]]','_'),'up','txt',sep='.')    
  FName <- write.table(input,paste(inputFolderName,FName,sep='/'),sep='\t',row.names = F,col.names = T, quote=F)
  
  input <- FC[(FC[,cont.name]<=-1 & PVAL[,cont.name]<=0.05),c('GeneNames',cont.name),drop=F]
  input <- dplyr::mutate(input,SystemCode = 'Sy')
  input <- input[,c(1,3,2)]
  colnames(input)[3] <- 'logFC'
  
  Genes <- rbind(Genes,cbind(paste(str_replace_all(cont.name,'[^[:alnum:]]','_'),'down',sep='_'),
                             paste(input$GeneNames,collapse=',')))
  
  FName <- paste(str_replace_all(cont.name,'[^[:alnum:]]','_'),'down','txt',sep='.')  
  FName <- write.table(input,paste(inputFolderName,FName,sep='/'),sep='\t',row.names = F,col.names = T, quote=F)
  
  return(Genes)
}
DEXPP.GENES <- lapply(colnames(FC)[-(1)],writeInputFiles, FC, PVAL, inputFolderName)
DEXPP.GENES <- plyr::ldply(DEXPP.GENES) 
colnames(DEXPP.GENES) <- c('Contrast.Names','Gene.Symbols')

# Write denominator file to folder
denominatorFolderName <- paste0(getwd(),'/GOElite_denominator')
if (file.exists(denominatorFolderName))
  unlink(denominatorFolderName, recursive = T, force = T)
dir.create(denominatorFolderName, showWarnings = TRUE, recursive = FALSE)
tmp <- FC[,c('GeneNames'),drop=F] 
tmp <- dplyr::mutate(tmp,SystemCode = 'Sy')
FName <- 'Denominaor.txt'
write.table(tmp,paste(denominatorFolderName,FName,sep='/'),sep='\t',row.names = F,col.names = T, quote=F)

# Create output dir
outputFolderName <- paste0(getwd(),'/GOElite_output')
if (file.exists(outputFolderName))
  unlink(outputFolderName, recursive = T, force = T)
dir.create(outputFolderName, showWarnings = TRUE, recursive = FALSE)

# system(paste('python',
#              '/mnt/data/GO-Elite_v.1.2.5-Py/GO_Elite.py',
#              '--species','Hs',
#              '--version','EnsMart72Plus',
#              '--input', inputFolderName,
#              '--denom',denominatorFolderName,
#              '--output',outputFolderName))
```


Store files in synapse
 
```{r synapse.store, include = FALSE, eval=TRUE}
if(SYNAPSE_STORE){
  activityName='Differential Expression Analysis of eXpress aligned data with mixed effects model'
  
  thisFileName <- 'DiffExpAnal_Express_mixedEffects.Rmd'
  
  # Github link
  thisRepo <- getRepo(repository = "th1vairam/pcbc_c4_analysis", 
                      ref="branch", 
                      refName='diff_exp')

  thisFile <- getPermlink(repository = thisRepo,
                          repositoryPath=paste0('code/Rmd/', thisFileName))
  
  
  CODE <- File('./DiffExpAnal_Express_mixedEffects.Rmd',name = 'Differential Expression Analysis Express Mixed Effects',parentId = parentId)
  CODE <- synStore(CODE, used = ALL_USED_IDs,activityName = activityName, executed=thisFile)
      
  write.table(FC,file='./DiffnameShort_FoldChange_Express_mixedEffects.tsv',sep='\t',row.names=F,quote=F)
  FC <- File('./DiffnameShort_FoldChange_Express_mixedEffects.tsv',name = 'Differential Expression Analysis Express Mixed Effects logFC',parentId = parentId)
  FC <- synStore(FC, used = ALL_USED_IDs, activityName = activityName, executed=CODE$properties$id)
    
  write.table(PVAL,file='./DiffnameShort_Pvalue_Express_mixedEffects.tsv',sep='\t',row.names=F,quote=F)
  PVAL <- File('./DiffnameShort_Pvalue_Express_mixedEffects.tsv',name = 'Differential Expression Analysis Express Mixed Effects pvalue',parentId = parentId)
  PVAL <- synStore(PVAL, used = ALL_USED_IDs, activityName = activityName, executed=CODE$properties$id)
  
  write.table(DEXPP.GENES,file='./DiffnameShort_DiffGenes_Express_mixedEffects.tsv',sep='\t',row.names=F,quote=F)
  DEXPP <- File('./DiffnameShort_DiffGenes_Express_mixedEffects.tsv',name = 'Differential Expression Analysis Express Mixed Effects genelist',parentId = parentId)
  annotations(DEXPP) <- list(Pvalue='0.05',Foldchage='+/-1')
  DEXPP <- synStore(DEXPP, used = ALL_USED_IDs, activityName = activityName, executed=CODE$properties$id)
}
```
|  *Results*           |  *SynapseID*                |
|  -------             |   ---------                 |
|  Log fold change     |  `r FC$properties$id`       |
|  Adjusted pvalues    |  `r PVAL$properties$id`     |
|  Differentially expressed genes  |  `r DEXPP$properties$id`  |