---
title: "Differential expression analysis for eXpress aligned data with mixed effects modeling"
author: "Thanneer Perumal"
date: "`r date()`"
output: html_document
---

```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
## It is assumed your working directory is where this file is
# Load required libraries
library('synapseClient')
library('limma')
library('edgeR')
library('RColorBrewer')
library('ctv')
library('ggplot2')
library('psych')
library('reshape2')
library('gplots')
library('vcd')
library('erer')
library('fpc')
library('data.table')
library('dplyr')
library('stringr')
library('knitr')
library('goseq')

## Requires ggdendro
devtools::source_url("https://raw.githubusercontent.com/chr1swallace/random-functions/master/R/ggplot-heatmap.R")

# source('/mnt/Github/knit2synapse-1//R/knitFile2Synapse.R')
# knitFile2Synapse(file = "./DiffExpAnal_Express_mixedEffects.Rmd", owner = 'syn3972059', name = "Differential Expression Analysis Express Mixed Effects",overwrite=F)

synapseLogin()

# source utility files from ../R/lib folder
file.sources = list.files('../R/lib',pattern="*.R")
file.sources = sapply(file.sources,function(x){return(paste('../R/lib',x,sep='/'))})
tmp = sapply(file.sources,source,.GlobalEnv)

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE)
```

### Download data

```{r setup, include=FALSE}
# Input Parameters
DESIGN_ID = 'syn3943971';
EXPR_ID = 'syn3943951';
WEIGHTS_ID = 'syn3943964';
COVARIATES_ID = 'syn3943968';

SYNAPSE_STORE = T
parentId = 'syn3471223'

# Specify factor and continuous covarites pool (adjusted covariates will only be the subset of these covariates)
FactorCovariates = c('Diffname_short','run','lane','index','Cell_Type_of_Origin', 'Cell_Line_of_Origin', 'Tissue_of_Origin', 'Reprogramming_Vector_Type','Reprogramming_Gene_Combination', 'Donor_Life_Stage','Originating_Lab','Gender','Donor_ID','Disease','Race','Ethnicity','Cell_Line_Type','PassageAtHarvest')
ContCovariates = NULL
```
Synapse id of design, expression, weights and covariates matrix used for the analysis are `r DESIGN_ID`,`r EXPR_ID`,`r WEIGHTS_ID`, and `r COVARIATES_ID`

```{r getdata, cache=TRUE, include=FALSE}
# Get design matrix
DESIGN_OBJ = synGet(DESIGN_ID);
ALL_USED_IDs = c(DESIGN_OBJ$properties$id);
DESIGN = read.table(getFileLocation(DESIGN_OBJ),sep='\t',header=T,row.names=1, check.names=FALSE)
colnames(DESIGN) = gsub("[[:punct:]]", "_",colnames(DESIGN))
colnames(DESIGN) = gsub(" ", "_",colnames(DESIGN))

# Get expression matrix
EXPR_OBJ = synGet(EXPR_ID);
ALL_USED_IDs = c(ALL_USED_IDs,EXPR_OBJ$properties$id);
EXPR = read.table(getFileLocation(EXPR_OBJ),sep='\t',header=T,row.names=1, check.names=FALSE)

# Get weighting matrix
WEIGHTS_OBJ = synGet(WEIGHTS_ID);
ALL_USED_IDs = c(ALL_USED_IDs,WEIGHTS_OBJ$properties$id);
WEIGHTS = read.table(getFileLocation(WEIGHTS_OBJ),sep='\t',header=T,row.names=1, check.names=FALSE)

# Get weighting matrix
COVARIATES_OBJ = synGet(COVARIATES_ID);
ALL_USED_IDs = c(ALL_USED_IDs,COVARIATES_OBJ$properties$id);
COVARIATES = read.table(getFileLocation(COVARIATES_OBJ),sep='\t',header=T,row.names=1, check.names=FALSE)
```

Get differentially expressed genes using limma package with following co-efficients `r colnames(DESIGN)` in linear model

Degenerate columns in the new design are `r linColumnFinder(DESIGN)$relations`

### Differential expression 

##### Differentiation stages
```{r diffstate, cache=TRUE, include=FALSE}
# Calculate correlation between donors
correlation <- duplicateCorrelation(EXPR, design=DESIGN, block=COVARIATES$Donor_ID)

# Fit linear model using mixed effects design
FIT = lmFit(EXPR, design=DESIGN, weights = WEIGHTS, block=COVARIATES$Donor_ID, correlation = correlation$cor)

# Make contrast to check differential expression between different differentiation stages
CONT.NAMES <- colnames(DESIGN)[grep('Diffname_short',colnames(DESIGN))]
CONT.NAMES <- combn(CONT.NAMES,2)
CONT.NAMES <- apply(CONT.NAMES,2,paste,collapse='-')

CONT <- makeContrasts(contrasts=CONT.NAMES,levels=colnames(FIT$coefficients))
colnames(CONT) <- sapply(colnames(CONT),function(x){x <- gsub('Diffname_short','',x);})

# Refit contrasts
FIT.CONTRAST <- contrasts.fit(FIT,CONT)
  
# Estimate moderated t-statistics
FIT.CONTRAST <- eBayes(FIT.CONTRAST)
  
# Obtain all the differential expession combinations
DIFFNAME <- list()
DIFFNAME$logFC <- data.frame(row.names = rownames(EXPR))
DIFFNAME$adj.P.Val <- data.frame(row.names = rownames(EXPR))
  
for (i in colnames(CONT)){
  tmp <- topTable(FIT.CONTRAST, coef=i, number=dim(EXPR)[1])    
  DIFFNAME$logFC[,str_replace_all(i,"Diffname_short","")] <- tmp[rownames(DIFFNAME$logFC),'logFC']
  DIFFNAME$adj.P.Val[,str_replace_all(i,"Diffname_short","")] <- tmp[rownames(DIFFNAME$adj.P.Val),'adj.P.Val']
}

# Get number of significantly differentialy expressed genes with adj.P.Val <= 0.05 and abs(logFC) >= 1
DIFFNAME$SIG.EXP <- DIFFNAME$adj.P.Val<=0.05 & abs(DIFFNAME$logFC) >= 1   
DIFFNAME$NUM.SIG.EXP <- colSums(DIFFNAME$SIG.EXP)

save(DIFFNAME,file='DifferentialExpressionDiffName_mixedEffects.RData')
```

Number of differentially expressed genes at FDR <= 0.05 and absolute logFC >= 1
```{r significant.sets1,cache=TRUE}
tmp <- as.data.frame(DIFFNAME$NUM.SIG.EXP)
colnames(tmp) <- 'NUMBER OF SIGNIFICANTLY EXPRESSED GENES'
kable(tmp)
```

#### Gender 
```{r gender, cache=TRUE}
# Post adjusted design matrix
DESIGN = getDesignMatrix(COVARIATES[,c("Gender","Diffname_short","run","Donor_Life_Stage","Reprogramming_Vector_Type"),drop=F],Intercept = F)
DESIGN = DESIGN$design[,linColumnFinder(DESIGN$design)$indepCols]
colnames(DESIGN) <- str_replace_all(colnames(DESIGN),'[[:punct:]]','_')
colnames(DESIGN) <- str_replace_all(colnames(DESIGN),' ','_')

# Calculate correlation between donors
correlation <- duplicateCorrelation(EXPR, design=DESIGN, block=COVARIATES$Donor_ID)

# Fit linear model using mixed effects design
FIT = lmFit(EXPR, design=DESIGN, weights = WEIGHTS, block=COVARIATES$Donor_ID, correlation = correlation$cor)

# Make contrast to check differential expression between different differentiation stages
CONT.NAMES <- colnames(DESIGN)[grep('Gender',colnames(DESIGN))]
CONT.NAMES <- combn(CONT.NAMES,2)
CONT.NAMES <- apply(CONT.NAMES,2,paste,collapse='-')

CONT <- makeContrasts(contrasts=CONT.NAMES,levels=colnames(FIT$coefficients))
colnames(CONT) <- sapply(colnames(CONT),function(x){x <- gsub('Gender','',x);})

# Refit contrasts
FIT.CONTRAST <- contrasts.fit(FIT,CONT)
  
# Estimate moderated t-statistics
FIT.CONTRAST <- eBayes(FIT.CONTRAST)
  
# Obtain all the differential expession combinations
GENDER <- list()
GENDER$logFC <- data.frame(row.names = rownames(EXPR))
GENDER$adj.P.Val <- data.frame(row.names = rownames(EXPR))
  
for (i in colnames(CONT)){
  tmp <- topTable(FIT.CONTRAST, coef=i, number=dim(EXPR)[1])    
  GENDER$logFC[,str_replace_all(i,"Gender","")] <- tmp[rownames(GENDER$logFC),'logFC']
  GENDER$adj.P.Val[,str_replace_all(i,"Gender","")] <- tmp[rownames(GENDER$adj.P.Val),'adj.P.Val']
}

# Get number of significantly differentialy expressed genes with adj.P.Val <= 0.05 and abs(logFC) >= 1
GENDER$SIG.EXP <- GENDER$adj.P.Val<=0.05 & abs(GENDER$logFC) >= 1
GENDER$NUM.SIG.EXP <- colSums(GENDER$SIG.EXP)

save(GENDER,file='DifferentialExpressionGender_mixedEffects.RData')
```

Number of differentially expressed genes at FDR <= 0.05 and absolute logFC >= 1
```{r significant.sets2,cache=TRUE}
tmp <- as.data.frame(GENDER$NUM.SIG.EXP)
colnames(tmp) <- 'NUMBER OF SIGNIFICANTLY EXPRESSED GENES'
kable(tmp)
```

#### Differentiation state and gender

```{r diffstate.gender, cache=TRUE}
# Post adjusted design matrix
COVARIATES.NEW <- COVARIATES
COVARIATES.NEW <- mutate(COVARIATES.NEW,Diff.Gender = paste(Diffname_short,Gender,sep='.'))
COVARIATES.NEW$Diff.Gender <- factor(COVARIATES.NEW$Diff.Gender)

DESIGN = getDesignMatrix(COVARIATES.NEW[,c("Diff.Gender","run","Donor_Life_Stage","Reprogramming_Vector_Type"),drop=F],Intercept = F)
DESIGN = DESIGN$design[,linColumnFinder(DESIGN$design)$indepCols]
colnames(DESIGN) <- str_replace_all(colnames(DESIGN),'[[:punct:]]','_')
colnames(DESIGN) <- str_replace_all(colnames(DESIGN),' ','_')

# Calculate correlation between donors
correlation <- duplicateCorrelation(EXPR, design=DESIGN, block=COVARIATES.NEW$Donor_ID)

# Fit linear model using mixed effects design
FIT = lmFit(EXPR, design=DESIGN, weights = WEIGHTS, block=COVARIATES.NEW$Donor_ID, correlation = correlation$cor)

# Make contrast to check differential expression between different differentiation stages
CONT.NAMES <- colnames(DESIGN)[grep('female',colnames(DESIGN))]
CONT.NAMES <- combn(CONT.NAMES,2)
CONT.NAMES1 <- apply(CONT.NAMES,2,paste,collapse='-')
CONT.NAMES2 <- sapply(CONT.NAMES1,str_replace_all,'female','male')
CONT.NAMES <- paste0('(',CONT.NAMES1,')-(',CONT.NAMES2,')')

CONT <- makeContrasts(contrasts=CONT.NAMES,levels=colnames(FIT$coefficients))
colnames(CONT) <- sapply(colnames(CONT),function(x){x <- gsub('Diff_Gender','',x);})

# Refit contrasts
FIT.CONTRAST <- contrasts.fit(FIT,CONT)
  
# Estimate moderated t-statistics
FIT.CONTRAST <- eBayes(FIT.CONTRAST)
  
# Obtain all the differential expession combinations
DIFFNAME.GENDER <- list()
DIFFNAME.GENDER$logFC <- data.frame(row.names = rownames(EXPR))
DIFFNAME.GENDER$adj.P.Val <- data.frame(row.names = rownames(EXPR))
  
for (i in colnames(CONT)){
  tmp <- topTable(FIT.CONTRAST, coef=i, number=dim(EXPR)[1])    
  DIFFNAME.GENDER$logFC[,str_replace_all(i,"DIFFNAME.GENDER","")] <- tmp[rownames(DIFFNAME.GENDER$logFC),'logFC']
  DIFFNAME.GENDER$adj.P.Val[,str_replace_all(i,"DIFFNAME.GENDER","")] <- tmp[rownames(DIFFNAME.GENDER$adj.P.Val),'adj.P.Val']
}

# Get number of significantly differentialy expressed genes with adj.P.Val <= 0.05 and abs(logFC) >= 1
DIFFNAME.GENDER$SIG.EXP <- DIFFNAME.GENDER$adj.P.Val<=0.05 & abs(DIFFNAME.GENDER$logFC) >= 1   
DIFFNAME.GENDER$NUM.SIG.EXP <- colSums(DIFFNAME.GENDER$SIG.EXP)

save(DIFFNAME.GENDER,file='DifferentialExpressionDiffnameGender_mixedEffects.RData')
```

Number of differentially expressed genes at FDR <= 0.05 and absolute logFC >= 1
```{r significant.sets3,cache=TRUE}
tmp <- as.data.frame(DIFFNAME.GENDER$NUM.SIG.EXP)
colnames(tmp) <- 'NUMBER OF SIGNIFICANTLY EXPRESSED GENES'
kable(tmp)
```

#### Reprogramming vector type

```{r reprogramming.vector, cache=TRUE}
# Get design matrix
NEW.DESIGN <- getDesignMatrix(dplyr::select(COVARIATES,one_of('Reprogramming_Vector_Type','Diffname_short','run','Donor_Life_Stage','Gender')), Intercept = F)
tmp <- linColumnFinder(NEW.DESIGN$design)
NEW.DESIGN$design <- NEW.DESIGN$design[,tmp$indepCols]
colnames(NEW.DESIGN$design) = gsub("[[:punct:]]", "_",colnames(NEW.DESIGN$design))
colnames(NEW.DESIGN$design) = gsub(" ", "_",colnames(NEW.DESIGN$design))

# Calculate correlation between donors
correlation <- duplicateCorrelation(EXPR, design=NEW.DESIGN$design, block=COVARIATES$Donor_ID)
  
# Fit linear model using mixed effects design
FIT = lmFit(EXPR, design=NEW.DESIGN$design, weights = WEIGHTS, block=COVARIATES$Donor_ID, correlation = correlation$cor)
                  
# Make contrast to check differential expression between different differentiation stages
CONT.NAMES <- colnames(NEW.DESIGN$design)[grep('Reprogramming_Vector_Type',colnames(NEW.DESIGN$design))]
CONT.NAMES <- combn(CONT.NAMES,2)
CONT.NAMES <- apply(CONT.NAMES,2,paste,collapse='-')

CONT <- makeContrasts(contrasts=CONT.NAMES,levels=colnames(FIT$coefficients))
colnames(CONT) <- sapply(colnames(CONT),function(x){x <- gsub('Reprogramming_Vector_Type','',x);})

# Refit contrasts
FIT.CONTRAST <- contrasts.fit(FIT,CONT)

# Estimate moderated t-statistics
FIT.CONTRAST <- eBayes(FIT.CONTRAST)

# Obtain all the differential expession combinations
REPVECTOR <- list()
REPVECTOR$logFC <- data.frame(row.names = rownames(EXPR))
REPVECTOR$adj.P.Val <- data.frame(row.names = rownames(EXPR)) 

# Get differential expression
for (i in colnames(CONT)){
  tmp <- topTable(FIT.CONTRAST, coef=i, number=dim(EXPR)[1])
  REPVECTOR$logFC[,i] <- tmp[rownames(REPVECTOR$logFC),'logFC']
  REPVECTOR$adj.P.Val[,i] <- tmp[rownames(REPVECTOR$adj.P.Val),'adj.P.Val']
}

# Get number of significantly differentialy expressed genes with adj.P.Val <= 0.05 and abs(logFC) >= 1
REPVECTOR$SIG.EXP <- REPVECTOR$adj.P.Val<=0.05 & abs(REPVECTOR$logFC) >= 1   
REPVECTOR$NUM.SIG.EXP <- colSums(REPVECTOR$SIG.EXP)

save(REPVECTOR,file='DifferentialExpressionReprogrammingVectorType.RData')
```

Number of differentially expressed genes at FDR <= 0.05 and absolute logFC >= 1
```{r significant.sets4,cache=TRUE}
tmp <- as.data.frame(REPVECTOR$NUM.SIG.EXP)
colnames(tmp) <- 'NUMBER OF SIGNIFICANTLY EXPRESSED GENES'
kable(tmp)
```

#### Reprogramming vector type and gender

```{r reprogramming.vector.gender, cache=TRUE}
# Post adjusted design matrix
COVARIATES.NEW <- COVARIATES
COVARIATES.NEW <- mutate(COVARIATES.NEW,Repvect.Gender = paste(Reprogramming_Vector_Type,Gender,sep='.'))
COVARIATES.NEW$Repvect.Gender <- factor(COVARIATES.NEW$Repvect.Gender)

# Get design matrix
NEW.DESIGN <- getDesignMatrix(dplyr::select(COVARIATES.NEW,one_of('Repvect.Gender','Diffname_short','run','Donor_Life_Stage')), Intercept = F)
tmp <- linColumnFinder(NEW.DESIGN$design)
NEW.DESIGN$design <- NEW.DESIGN$design[,tmp$indepCols]
colnames(NEW.DESIGN$design) = gsub("[[:punct:]]", "_",colnames(NEW.DESIGN$design))
colnames(NEW.DESIGN$design) = gsub(" ", "_",colnames(NEW.DESIGN$design))

# Calculate correlation between donors
correlation <- duplicateCorrelation(EXPR, design=NEW.DESIGN$design, block=COVARIATES$Donor_ID)
  
# Fit linear model using mixed effects design
FIT = lmFit(EXPR, design=NEW.DESIGN$design, weights = WEIGHTS, block=COVARIATES$Donor_ID, correlation = correlation$cor)
                  
# Make contrast to check differential expression between different reprogramming vectors with male and female cells
CONT.NAMES1 <- colnames(NEW.DESIGN$design)[grep('_female',colnames(NEW.DESIGN$design))]
CONT.NAMES1 <- combn(CONT.NAMES1,2)
CONT.NAMES1 <- apply(CONT.NAMES1,2,paste,collapse='-')
CONT.NAMES2 <- sapply(CONT.NAMES1,str_replace_all,'female','male')
CONT.NAMES <- paste0('(',CONT.NAMES1,')-(',CONT.NAMES2,')')

CONT <- makeContrasts(contrasts=CONT.NAMES,levels=colnames(FIT$coefficients))
colnames(CONT) <- sapply(colnames(CONT),function(x){x <- gsub('Repvect_Gender','',x);})

# Refit contrasts
FIT.CONTRAST <- contrasts.fit(FIT,CONT)

# Estimate moderated t-statistics
FIT.CONTRAST <- eBayes(FIT.CONTRAST)

# Obtain all the differential expession combinations
REPVECTOR.GENDER <- list()
REPVECTOR.GENDER$logFC <- data.frame(row.names = rownames(EXPR))
REPVECTOR.GENDER$adj.P.Val <- data.frame(row.names = rownames(EXPR)) 

# Get differential expression
for (i in colnames(CONT)){
  tmp <- topTable(FIT.CONTRAST, coef=i, number=dim(EXPR)[1])
  REPVECTOR.GENDER$logFC[,i] <- tmp[rownames(REPVECTOR.GENDER$logFC),'logFC']
  REPVECTOR.GENDER$adj.P.Val[,i] <- tmp[rownames(REPVECTOR.GENDER$adj.P.Val),'adj.P.Val']
}

# Get number of significantly differentialy expressed genes with adj.P.Val <= 0.05 and abs(logFC) >= 1
REPVECTOR.GENDER$SIG.EXP <- REPVECTOR.GENDER$adj.P.Val<=0.05 & abs(REPVECTOR.GENDER$logFC) >= 1   
REPVECTOR.GENDER$NUM.SIG.EXP <- colSums(REPVECTOR.GENDER$SIG.EXP)

save(REPVECTOR.GENDER,file='DifferentialExpressionReprogrammingVectorType.RData')
```

Number of differentially expressed genes at FDR <= 0.05 and absolute logFC >= 1
```{r significant.sets5,cache=TRUE}
tmp <- as.data.frame(REPVECTOR.GENDER$NUM.SIG.EXP)
colnames(tmp) <- 'NUMBER OF SIGNIFICANTLY EXPRESSED GENES'
kable(tmp)
```

#### Get pvalue and fold changes of some important genes

```{r print.imp.genes}
FC <- join_all(list(rownameToFirstColumn(DIFFNAME$logFC,'GeneNames'),
                    rownameToFirstColumn(DIFFNAME.GENDER$logFC,'GeneNames'),
                    rownameToFirstColumn(REPVECTOR$logFC,'GeneNames'),
                    rownameToFirstColumn(REPVECTOR.GENDER$logFC,'GeneNames')),
               by = 'GeneNames',
               match = 'all')

PVAL <- join_all(list(rownameToFirstColumn(DIFFNAME$adj.P.Val,'GeneNames'),
                      rownameToFirstColumn(DIFFNAME.GENDER$adj.P.Val,'GeneNames'),
                      rownameToFirstColumn(REPVECTOR$adj.P.Val,'GeneNames'),
                      rownameToFirstColumn(REPVECTOR.GENDER$adj.P.Val,'GeneNames')),
                 by = 'GeneNames',
                 match = 'all')

SIG <- join_all(list(rownameToFirstColumn(DIFFNAME$SIG.EXP,'GeneNames'),
                     rownameToFirstColumn(DIFFNAME.GENDER$SIG.EXP,'GeneNames'),
                     rownameToFirstColumn(REPVECTOR$SIG.EXP,'GeneNames'),
                     rownameToFirstColumn(REPVECTOR.GENDER$SIG.EXP,'GeneNames')),
                by = 'GeneNames',
                match = 'all')

print('Fold change:')
kable(filter(FC, GeneNames %in% c('LARGE','FRG1','FRG1B','PEG10','SPEG')))
print('Adjusted Pvalue:')
kable(filter(PVAL, GeneNames %in% c('LARGE','FRG1','FRG1B','PEG10','SPEG')))
```

#### Gender effect in differentially expressed genes

ECDFs of CV between male and female cells at each differentiation stages
```{r gender.effect}
ind <- which(rowSums(SIG[,2:16]) != 0)

GENDER.VAR <- list()
for (Diffname in levels(COVARIATES$Diffname_short)){
  for (Gender in levels(COVARIATES$Gender)){
    TEMP.EXPR <- EXPR[ind, COVARIATES$Diffname_short == Diffname & COVARIATES$Gender == Gender]
    GENDER.VAR[[Diffname]][[Gender]] <- apply(TEMP.EXPR,1,mean)/apply(TEMP.EXPR,1,sd)
  }
  plotdata <- as.data.frame(rbind(cbind(GENDER.VAR[[Diffname]]$male,'male'),cbind(GENDER.VAR[[Diffname]]$female,'female')))
  plotdata[,1] <- as.numeric(as.character(plotdata[,1]))
  colnames(plotdata) <- c('CV','Gender')
  GENDER.VAR[[Diffname]]$plot <- ggplot(plotdata, aes(x = CV)) + stat_ecdf(aes(colour = Gender)) + ggtitle(Diffname)  
  print(GENDER.VAR[[Diffname]]$plot)
}
```

Perform enrichment analysis for differentiation
```{r enrichment.analysis, include=FALSE}
enrichmentAnalysis <- function(ind,PVAL,FC){
  allGenes <- PVAL[row.names(FC),ind] <= 0.05 & abs(FC[,ind]) >= 2
  names(allGenes) <- row.names(FC)
  
  # Calculate probability weighting function based on length of gene
  pwf = nullp(allGenes, 'hg19', 'geneSymbol',plot.fit=F)
  
  # Calculate go enrichment using Wallenius method    
  GO.wall = goseq(pwf, 'hg19', 'geneSymbol', method = "Wallenius")
  GO.wall$over_represented_adj_pvalue = p.adjust(GO.wall$over_represented_pvalue, method = 'bonferroni')
  GO.wall$under_represented_adj_pvalue = p.adjust(GO.wall$under_represented_pvalue, method = 'bonferroni')

  return(GO.wall)
}
row.names(PVAL) <- PVAL$GeneNames
row.names(FC) <- FC$GeneNames

tmp <- enrichmentAnalysis(colnames(PVAL)[2],PVAL,FC)
ENRICH <- tmp[,c("category","over_represented_adj_pvalue","term","ontology")]
rownames(ENRICH) <- tmp$category
ENRICH <- dplyr::rename(ENRICH, "DE-EB" = over_represented_adj_pvalue)

for (cols in colnames(PVAL)[-(1:2)]){
  tmp <- enrichmentAnalysis(cols,PVAL,FC)
  rownames(tmp) <- tmp$category
  ENRICH[,cols] <- tmp[rownames(ENRICH),c("over_represented_adj_pvalue")]
}
save(ENRICH,file='DifferentialExpressionEnrichment.RData')
```

Store files in synapse
 
```{r synapse.store, include=FALSE, eval=TRUE}
if(SYNAPSE_STORE){
  activityName='Differential Expression Analysis of eXpress aligned data with mixed effects model'
  
  CODE <- File('./DiffExpAnal_Express_mixedEffects.Rmd',name = 'Differential Expression Analysis Express Mixed Effects',parentId = parentId)
  CODE <- synStore(CODE, used = ALL_USED_IDs,activityName = activityName, executed='https://github.com/th1vairam/pcbc_c4_analysis/blob/diff_exp/code/Rmd/DiffExpAnal_Express_mixedEffects.Rmd')
      
  write.table(FC,file='./DiffnameShort_FoldChange_Express_mixedEffects.tsv',sep='\t',row.names=F,quote=F)
  FC <- File('./DiffnameShort_FoldChange_Express_mixedEffects.tsv',name = 'Differential Expression Analysis Express Mixed Effects logFC',parentId = parentId)
  FC <- synStore(FC, used = ALL_USED_IDs, activityName = activityName, executed=CODE$properties$id)
    
  write.table(PVAL,file='./DiffnameShort_Pvalue_Express_mixedEffects.tsv',sep='\t',row.names=F,quote=F)
  PVAL <- File('./DiffnameShort_Pvalue_Express_mixedEffects.tsv',name = 'Differential Expression Analysis Express Mixed Effects pvalue',parentId = parentId)
  PVAL <- synStore(PVAL, used = ALL_USED_IDs, activityName = activityName, executed=CODE$properties$id)
  
  write.table(ENRICH,file='./DiffnameShort_Enrichment_Express_mixedEffects.tsv',sep='\t',row.names=F,quote=F)
  ENRICH1 <- File('./DiffnameShort_Enrichment_Express_mixedEffects.tsv',name = 'Differential Expression Analysis Express Mixed Effects enrichment',parentId = parentId)
  ENRICH1 <- synStore(ENRICH1, used = ALL_USED_IDs, activityName = activityName, executed=CODE$properties$id)
}
```