---
title: "Differential expression analysis of minfi curated methylation data with mixed effect modeling"
author: "Thanneer Perumal"
date: '`r date()`'
output: html_document
---
```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
## It is assumed your working directory is where this file is

# Clear R console screen output
cat("\014")

# Load required libraries
library('synapseClient')
library('RColorBrewer')
library('ggplot2')
library('gplots')
library('limma')
library('edgeR')
library('ctv')
library('psych')
library('reshape2')
library('vcd')
library('erer')
library('fpc')
library('data.table')
library('dplyr')
library('plyr')
library('knitr')
library('stringr')
library('minfi')
library('org.Hs.eg.db')

## Needs the dev branch
library(rGithubClient)

## Requires ggdendro
devtools::source_url("https://raw.githubusercontent.com/chr1swallace/random-functions/master/R/ggplot-heatmap.R")

# source('/mnt/Github/knit2synapse-1/R/knitFile2SynapseFolder.R')
# knitFile2SynapseFolder(file='./DiffExpAnal_methylation_mixedEffects.Rmd',entityName = 'Differential Expression Analysis Methylation Mixed Effects Model All', owner = 'syn4489211', overwrite=F)

synapseLogin()

# source utility files from ../R/lib folder
file.sources = list.files('../R/lib',pattern="*.R")
file.sources = sapply(file.sources,function(x){return(paste('../R/lib',x,sep='/'))})
tmp = sapply(file.sources,source,.GlobalEnv)

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE)
```
### Download data
```{r setup, include=FALSE}
# Input Parameters
EXP_ID = 'syn4487666'
COVARIATES_ID = 'syn4487669'
PROBE2ENTRZ_ID = 'syn2775255'

SYNAPSE_STORE = T
parentId = 'syn4231339'

# Specify factor and continuous covarites pool (adjusted covariates will only be the subset of these covariates)
FactorCovariates = c("Diffname_short", "BeadChip", "Row", "Column", "Cell_Line_Type", 
                     "Tissue_of_Origin", "Reprogramming_Gene_Combination", "Culture_Conditions",  
                     "Other_Conditions_During_Reprogramming", "Donor_Life_Stage", "Gender", 
                     "Originating_Lab", "Donor_ID", "Cell_Type_of_Origin_Level2",
                     "Reprogramming_Vector_Type" )
ContCovariates = c("PassageAtThaw", "PassageAtDNAHarvest")
```
Synapse id of count matrix used for the analysis is `r paste(EXP_ID,EXP_OBJ$properties$versionNumber,sep='.')` and the covariates synapse id of meta data table used for the analysis is `r paste(COVARIATES_ID,COVARIATES_OBJ$properties$versionNumber,sep='.')` 

Factor covariates considered for analysis are `r paste(gsub('_','\\\\_',FactorCovariates),collapse=',')`, and continuous covariates considered for the analysis are `r paste(gsub('_','\\\\_',ContCovariates),collapse=',')`

Obtain count matrix and metadata from synapse
```{r getdata, cache=TRUE, include=FALSE}
# Get count matrix
EXP_OBJ = synGet(EXP_ID)
ALL_USED_IDs = EXP_OBJ$properties$id
EXP = fread(getFileLocation(EXP_OBJ))
tmp = read.table(getFileLocation(EXP_OBJ), sep = '\t', header=T, check.names=F, nrow=1)
setnames(EXP, colnames(EXP), colnames(tmp))
EXP = as.data.frame(EXP)
rownames(EXP) = EXP[,1]

# Get metadata
COVARIATES_OBJ = synGet(COVARIATES_ID)
ALL_USED_IDs[length(ALL_USED_IDs)+1] = COVARIATES_OBJ$properties$id
COVARIATES = read.table(getFileLocation(COVARIATES_OBJ), sep='\t', header=T)

# Get probe related metadata
fData = get450KProbeMapping(EXP[['methProbeID']])
```
PDF of expression
```{r density.calculations,cache =TRUE}
tmp = melt(EXP, id = 'methProbeID')
tmp$data = 'Residual'

tmp1 = fread(synGet('syn2233188')@filePath, data.table=F)
setnames(tmp1,'ProbeID','methProbeID')
tmp1 = tmp1[tmp1$methProbeID %in% EXP$methProbeID,colnames(tmp1) %in% colnames(EXP)]
tmp1 = melt(tmp1, id = 'methProbeID')
tmp1$data = 'Beta'

tmp2 = rbindlist(list(tmp1,tmp))
```
```{r density.plots, cache=FALSE, fig.height=6, fig.width=6}
print(ggplot(tmp2, aes(x = value, color= data)) + geom_density())
```
Perform differential expression
```{r diff.exp, cache=TRUE, fig.width=6, fig.height=6}
# Make contrast to check differential expression between different differentiation stages
CONT.NAMES = levels(COVARIATES$Diffname_short)
CONT.NAMES = combn(CONT.NAMES,2)

# Get differentially expressed probes
getDifferentialExpression <- function(x, COVARIATES, EXP, Annotations){  
  
  print(paste0('Comparison: ',x[1],'_vs_',x[2]))
  
  # Find differential expression
  COV = droplevels(dplyr::filter(COVARIATES, Diffname_short == x[1] | Diffname_short == x[2]))
  rownames(COV) = COV$UID
  EXP.IN = as.data.frame(EXP)
  EXP.IN = EXP.IN[, rownames(COV)]
  dmp = dmpFinder(as.matrix(EXP.IN), COV$Diffname_short, type="categorical", shrinkVar = T)
  
  # Pvalue distribution
  tmp <- rownameToFirstColumn(dmp,'methProbeID')
  tmp <- melt(tmp, id = 'methProbeID')
  print(ggplot(dplyr::filter(tmp, variable == 'pval' | variable == 'qval'), aes(x = value, color = variable)) + geom_density()+ggtitle(paste0(x[1],'_vs_',x[2])))
    
  # Plot differential expression data
  dmp = mutate(dmp, Diff.Exp = ifelse(qval <= 0.01,ifelse(intercept >= 0.4,'SIG. UP',ifelse(intercept <= -0.4,'SIG. DOWN','DIFF.EXP')),'NOT.DIFF.EXP'))
  print(ggplot(dmp,aes(x=intercept,y=-log10(qval), color= Diff.Exp))+geom_point()+ggtitle(paste0(x[1],'_vs_',x[2])))
  
  # Merge probe annotations
  dmp = rownameToFirstColumn(dmp, 'methProbeIDs')
  dmp = merge(dmp, 
              dplyr::select(Annotations[Annotations$methProbeIDs %in% rownames(dmp), ],methProbeIDs, nearestTSS, nearestTx),
              all=T, by = 'methProbeIDs')
  
  # Plot top four probes
  dmp = dplyr::arrange(dmp, qval)
  cpgs <- dmp$methProbeIDs[1:4]
  par(mfrow=c(2,2))
  plotCpg(as.matrix(EXP.IN), cpg=cpgs, pheno=COV$Diffname_short, ylab = 'Residual Beta')
  
  # Map probes to genes
  NUM.SIG.PROBES.UP = sum(dmp$qval<=0.01 & dmp$intercept > 0.4)
  NUM.SIG.PROBES.DOWN = sum(dmp$qval<=0.01 & dmp$intercept < -0.4)
  
  NUM.SIG.Tx.UP = length(unique(dmp$nearestTx[dmp$qval<=0.01 & dmp$intercept> 0.4]))
  NUM.SIG.Tx.DOWN = length(unique(dmp$nearestTx[dmp$qval<=0.01 & dmp$intercept< -0.4]))
  
  NUM.SIG.TSS.UP = length(unique(dmp$nearestTSS[dmp$qval<=0.01 & dmp$intercept> 0.4]))
  NUM.SIG.TSS.DOWN = length(unique(dmp$nearestTSS[dmp$qval<=0.01 & dmp$intercept< -0.4]))
  
  gctFormat = data.frame( Comparison = c('SIG.Tx.UP','SIG.Tx.DOWN', 'SIG.TSS.UP', 'SIG.TSS.DOWN'),
                          Gene.Symbols = c(paste(unique(dmp$nearestTx[dmp$qval<=0.01 & dmp$intercept> 0.4]), collapse=','),
                                          paste(unique(dmp$nearestTx[dmp$qval<=0.01 & dmp$intercept< -0.4]), collapse=','),
                                          paste(unique(dmp$nearestTSS[dmp$qval<=0.01 & dmp$intercept> 0.4]), collapse=','),
                                          paste(unique(dmp$nearestTSS[dmp$qval<=0.01 & dmp$intercept< -0.4]), collapse=',')))
  gctFormat$Comparison = paste0('All__',x[1],'_vs_',x[2],'__',gctFormat$Comparison)
  
  comparison = dmp[dmp$qval <= 0.01,]
  comparison = mutate(comparison , Diff.Exp = ifelse(intercept<0,'down','up'))
  comparison$Diff.Exp = paste0('All__',x[1],'_vs_',x[2],'__',comparison$Diff.Exp)
  
  return(list(DEXPP = dmp, 
              NUM.SIG.PROBES.UP = NUM.SIG.PROBES.UP, NUM.SIG.PROBES.DOWN = NUM.SIG.PROBES.DOWN, 
              NUM.SIG.TSS.UP = NUM.SIG.TSS.UP, NUM.SIG.TSS.DOWN = NUM.SIG.TSS.DOWN,
              NUM.SIG.Tx.UP = NUM.SIG.Tx.UP, NUM.SIG.Tx.DOWN = NUM.SIG.Tx.DOWN,
              gctFormat = gctFormat, comparison = comparison))
}

DIFF.EXP = apply(CONT.NAMES,2,getDifferentialExpression, COVARIATES, EXP, fData$Annotation)
names(DIFF.EXP) = apply(CONT.NAMES,2,paste,collapse='_vs_')
```
Differentially expressed probes at qval <= 0.05 and having intercept > 0.4 or < -0.4
```{r print.num.diff}
# Print differentially expressed genes and probes
tmp = sapply(DIFF.EXP, function(x){unlist(x[grep('NUM.SIG.',names(x), value=T)])})
kable(tmp)
```
```{r extract.results}
INT = lapply(DIFF.EXP, function(x){x$DEXPP[,c('methProbeIDs','intercept'),drop=F]})
INT = join_all(INT, by = 'methProbeIDs', match = 'all')
colnames(INT)[2:dim(INT)[2]] = names(DIFF.EXP)

PVAL = lapply(DIFF.EXP, function(x){x$DEXPP[,c('methProbeIDs','pval'),drop=F]})
PVAL = join_all(PVAL, by = 'methProbeIDs', match = 'all')
colnames(PVAL)[2:dim(PVAL)[2]] = names(DIFF.EXP)

QVAL = lapply(DIFF.EXP, function(x){x$DEXPP[,c('methProbeIDs','qval'),drop=F]})
QVAL = join_all(QVAL, by = 'methProbeIDs', match = 'all')
colnames(QVAL)[2:dim(QVAL)[2]] = names(DIFF.EXP)

gctFormat = lapply(DIFF.EXP, function(x){return(x$gctFormat)})
gctFormat = rbindlist(gctFormat)

Comparisons = lapply(DIFF.EXP,function(x){return(x$comparison)})
Comparisons = rbindlist(Comparisons)
```
### Store files in synapse
```{r synapse.store, include = FALSE, eval=TRUE, cache=FALSE}
activityName='Differential Expression Analysis of minfi processed methylation data with mixed effects model'
  
thisFileName <- 'DiffExpAnal_methylation_mixedEffects.Rmd'
  
# Github link
thisRepo <- getRepo(repository = "th1vairam/pcbc_c4_analysis", 
                    ref="branch", 
                    refName='methyl')

thisFile <- getPermlink(repository = thisRepo,
                        repositoryPath=paste0('code/Rmd/', thisFileName))
    
# Make folder and populate with wiki
CODE <- Folder(name = 'Differential Expression Analysis Methylation Mixed Effects Model All', parentId = parentId)
CODE <- synStore(CODE)

# Store results: intercept
write.table(INT,file='./DiffExpAnal_methylation_mixedEffects_intercept.tsv',sep='\t',row.names=F,quote=F)
INT_OBJ <- File('./DiffExpAnal_methylation_mixedEffects_intercept.tsv', name = 'Intercept', parentId = CODE$properties$id)
INT_OBJ <- synStore(INT_OBJ, used = ALL_USED_IDs, activityName = activityName, executed=thisFile)

# Store results: pval
write.table(PVAL,file='./DiffExpAnal_methylation_mixedEffects_PVAL.tsv',sep='\t',row.names=F,quote=F)
PVAL_OBJ <- File('./DiffExpAnal_methylation_mixedEffects_PVAL.tsv', name = 'pvalue', parentId = CODE$properties$id)
PVAL_OBJ <- synStore(PVAL_OBJ, used = ALL_USED_IDs, activityName = activityName, executed=thisFile)

# Store results: qval
write.table(QVAL,file='./DiffExpAnal_methylation_mixedEffects_QVAL.tsv',sep='\t',row.names=F,quote=F)
QVAL_OBJ <- File('./DiffExpAnal_methylation_mixedEffects_QVAL.tsv', name = 'qvalue', parentId = CODE$properties$id)
QVAL_OBJ <- synStore(QVAL_OBJ, used = ALL_USED_IDs, activityName = activityName, executed=thisFile)

# Store results: gctFormat
write.table(gctFormat, file='./DiffExpAnal_methylation_mixedEffects_gctFormat.gct',sep='\t',row.names=F,quote=F)
GCT_OBJ <- File('./DiffExpAnal_methylation_mixedEffects_gctFormat.gct', name = 'Differentially Expressed Genes gct format', parentId = CODE$properties$id)
GCT_OBJ <- synStore(GCT_OBJ, used = ALL_USED_IDs, activityName = activityName, executed=thisFile)

# Store results: gctFormat
write.table(Comparisons, file='./DiffExpAnal_methylation_mixedEffects_comparisons.tsv',sep='\t',row.names=F,quote=F)
COMP_OBJ <- File('./DiffExpAnal_methylation_mixedEffects_comparisons.tsv', name = 'Differentially Expressed Genes comparison list', parentId = CODE$properties$id)
COMP_OBJ <- synStore(COMP_OBJ, used = ALL_USED_IDs, activityName = activityName, executed=thisFile)
```
|  *Results*                           |  *SynapseID*                |
|  ----------------------------------- |   ------------------------  |
| intercept |  `r paste(INT_OBJ$properties$id,INT_OBJ$properties$versionNumber,sep='.')`  |
| qvalue    |  `r paste(QVAL_OBJ$properties$id,QVAL_OBJ$properties$versionNumber,sep='.')` |
| Differentially Expressed Genes (gct format)    |  `r paste(GCT_OBJ$properties$id,GCT_OBJ$properties$versionNumber,sep='.')` |
| Differentially Expressed Genes (comparison list)    |  `r paste(COMP_OBJ$properties$id,COMP_OBJ$properties$versionNumber,sep='.')` |