---
title: "Differential expression analysis for eXpress aligned data (EB's only)"
author: "Thanneer Perumal"
date: "`r date()`"
output: html_document
---

```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
## It is assumed your working directory is where this file is
# Load required libraries
library('synapseClient')
library('limma')
library('edgeR')
library('RColorBrewer')
library('ggplot2')
library('reshape2')
library('knitr')
library('data.table')
library('dplyr')

## Requires ggdendro
# devtools::source_url("https://raw.githubusercontent.com/chr1swallace/random-functions/master/R/ggplot-heatmap.R")

# source('/mnt/Github/Knit2Synapse/knit2synapse.R')
# knit2synapse(file = "./", owner = '', wikiName = "",overwrite=T)

synapseLogin()

# source utility files from ../R/lib folder
file.sources = list.files('../R/lib',pattern="*.R")
file.sources = sapply(file.sources,function(x){return(paste('../R/lib',x,sep='/'))})
tmp = sapply(file.sources,source,.GlobalEnv)

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE)
```

### Download data

```{r setup, include=FALSE}
# Input Parameters
COUNT_ID = 'syn3446250'
METADATA_ID = 'syn3156503'
OUTLIERS_ID = 'NULL'

SYNAPSE_STORE = F
parentId = 'syn3276099'

# Specify factor and continuous covarites pool (adjusted covariates will only be the subset of these covariates)
FactorCovariates = c('Diffname_short','run','lane','index','Cell_Type_of_Origin', 'Cell_Line_of_Origin', 'Tissue_of_Origin', 'Reprogramming_Vector_Type','Reprogramming_Gene_Combination', 'Donor_Life_Stage','Originating_Lab','Gender','Donor_ID','Disease','Race','Ethnicity','Cell_Line_Type')
ContCovariates = NULL
```
Synapse id of count matrix used for the analysis is `r COUNT_ID` and the synapse id of meta data table used for the analysis is `r METADATA_ID`. Outliers were extracted from `r OUTLIERS_ID`

Factor covariates considered for analysis are `r FactorCovariates`, and continuous covariates considered for the analysis are `r ContCovariates`.

Obtain count matrix and metadata from synapse.
```{r getdata, cache=TRUE, include=FALSE}
# Get count matrix
COUNT_OBJ = synGet(COUNT_ID)
ALL_USED_IDs = COUNT_OBJ$properties$id
COUNT = fread(getFileLocation(COUNT_OBJ), data.table=FALSE)
row.names(COUNT) = COUNT[,1]

# Get metadata
METADATA_OBJ = synTableQuery(paste('SELECT * FROM',METADATA_ID,sep=' '))
ALL_USED_IDs[length(ALL_USED_IDs)+1] = METADATA_OBJ@schema
METADATA = METADATA_OBJ@values
```

Preprocess counts matrix and metadata.
```{r preprocessing, include=FALSE}
# Preprocess metadata
METADATA[METADATA == 'N/A'] = 'Not Applicable'
```

### Preprocess data
* Remove somatic samples and samples with type NA.
* Remove samples that failed QC and samples classified as exclude.
* Remove samples with abnormal karyotypes.
* Select only EB state
```{r filtering, echo=TRUE}
#### Pre processing mRNA expression counts and metadata ####
metadata_filtered <- 
  METADATA %>%
  filter(Cell_Type == "PSC") %>%
  filter(Diffname_short == "EB") %>%
  filter(!is.na(Cell_Type)) %>%
  filter(pass_qc) %>%
  filter(!exclude) %>%
  filter(C4_Karyotype_Result != "abnormal" | !is.na(C4_Karyotype_Result))  %>%
  filter(UID %in% colnames(COUNT))

REMOVED_UID <- setdiff(METADATA$UID, metadata_filtered$UID)
METADATA <- metadata_filtered
COUNT <- COUNT[, METADATA$UID]
```

`r length(REMOVED_UID)` samples were removed that did not fit the criteria.

#### Remaining samples:

```{r}
METADATA %>% count(Cell_Line_Type) %>% kable
```


### CPM Normalisation
Determine design matrix for normalisation and differential expression analysis. 

Remove genes that have less than 1 cpm counts in at least 50% of each of the individual differentiation stages.
```{r cpmnormalisation}
tmp <- tapply(colnames(COUNT),
              factor(metadata_filtered$Diffname_short),
              function(cols,COUNT){PROCESSED_COUNTS = getGeneFilteredGeneExprMatrix(COUNT[,cols])},
              COUNT)

ugenes <- c()
for (i in unique(METADATA$Diffname_short)) {
  ugenes <- unique(c(ugenes,tmp[[i]]$filteredExprMatrix$genes[,1]))
}

COUNT <- COUNT[ugenes,,drop=F]
PROCESSED_COUNTS = getGeneFilteredGeneExprMatrix(COUNT,MIN_GENE_CPM=0,
                                                 MIN_SAMPLE_PERCENT_WITH_MIN_GENE_CPM=0)
```

### `Voom` normalization

Initial normalisation using voom (with NULL design):

```{r initial.voom.normalisation}
VOOM.GENE_EXPRESSION.NULL = voom(PROCESSED_COUNTS$filteredExprMatrix, design=NULL, plot=T)
```

Adjusted by `Cell_Line_Type` using voom:

```{r adjusted.voom.normalisation}
# Post adjusted design matrix
covariates <- METADATA %>% mutate(Cell_Line_Type=factor(Cell_Line_Type),
                                  run=factor(run), 
                                  lane=factor(lane),
                                  Donor_ID=factor(Donor_ID))

model.for.fit <- model.matrix(~0 + Cell_Line_Type, data=covariates)

VOOM.GENE_EXPRESSION <- voom(PROCESSED_COUNTS$filteredExprMatrix,
                             design=model.for.fit, plot=T)

```

### Fitting and differential expression

```{r limmafit}
# Fitting linear model
fit <- lmFit(VOOM.GENE_EXPRESSION.NULL$E,
             design = model.for.fit,
             weights=VOOM.GENE_EXPRESSION$weights)

# Make contrast to check differential expression between different differentiation stages
contrastNames <- grep('Cell_Line_Type', colnames(model.for.fit), value=TRUE)
contrastNames <- combn(contrastNames, 2)
contrastNames <- apply(contrastNames, 2, paste, collapse='-')
cont <- makeContrasts(contrasts=contrastNames,
                      levels=colnames(fit$coefficients))
  
# Refit contrasts
fit2 <- contrasts.fit(fit, cont)
  
# Estimate moderated t-statistics
fit2 <- eBayes(fit2)

p.thresh <- 0.05

topGenes <- topTable(fit2, coef="Cell_Line_TypeESC-Cell_Line_TypeiPSC", 
                     adjust.method = "BH", p=p.thresh)
```

Top genes with adjusted p-value less than `r p.thresh`:

```{r}
topGenes %>% 
  select(AveExpr, logFC, adj.P.Val) %>% 
  kable
```


### PCA Plot
Using `voom`-adjusted expression values and plotting the first two principal components.

```{r pcaplot, cache=TRUE, fig.height=15, fig.width=15}
# Find principal components of expression to plot
PC <- prcomp(VOOM.GENE_EXPRESSION$E[rownames(topGenes), ], 
             scale=FALSE, center=FALSE)

# Plot first 2 PCs
plotdata <- data.frame(UID=rownames(PC$rotation), 
                       PC1=PC$rotation[,1], 
                       PC2=PC$rotation[,2])

plotdata <- merge(plotdata, METADATA, by="UID")

p <- ggplot(plotdata, aes(x=PC1, y=PC2))
p <- p + geom_point(aes(color=Cell_Line_Type, shape=Gender), size=5)
p <- p + scale_color_manual(values=c("red", "black"))
p <- p + theme_bw() + theme(legend.position="top")
p
```
