---
title: "Differential expression analysis of minfi curated methylation data - blood samples"
author: "Kenneth Daily"
date: '`r date()`'
output: html_document
---
```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
## It is assumed your working directory is where this file is

# Load required libraries
library('synapseClient')
library('RColorBrewer')
library('ggplot2')
library('gplots')
library('limma')
library('edgeR')
library('ctv')
library('psych')
library('reshape2')
library('vcd')
library('erer')
library('fpc')
library('data.table')
library('plyr')
library('dplyr')
library('knitr')
library('stringr')
library('minfi')
library('org.Hs.eg.db')
library('knit2synapse')
library(pcbcStats)
library(FDb.InfiniumMethylation.hg19)

## Needs the dev branch
library(rGithubClient)

## Requires ggdendro
# knitToFolderEntity(file = "./DiffExpAnal_methylation_mixedEffects.Rmd", 
#                    parentId = 'syn4231339',
#                    entityName = 'Differential Expression Analysis Methylation Mixed Effects Model All', 
#                    overwrite=F)

synapseLogin()

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE)
```
### Download data
```{r setup, include=FALSE}
# Input Parameters
EXP_ID = 'syn2699024'
METADATA_ID <- "syn3156828"
parentId = 'syn4231339'

bloodLines <- c("CD90", "MEP", "GMP", "CD34DMSO", "CD34UNC", "MEG", 
                "SC13-046-32EPOMEP", "SC13-046-32RBC")
```

```{r getdata, cache=TRUE, include=FALSE}
# Get filtered beta values
EXP_OBJ = synGet(EXP_ID)
ALL_USED_IDs = EXP_OBJ$properties$id
EXP = fread(getFileLocation(EXP_OBJ), data.table = F)
row.names(EXP) = EXP[,1]
EXP = dplyr::select(EXP,-(V1))
```

```{r metadata}
METADATA_OBJ = synTableQuery(paste('SELECT * FROM',METADATA_ID,sep=' '))
ALL_USED_IDs[length(ALL_USED_IDs)+1] = METADATA_OBJ@schema
METADATA = METADATA_OBJ@values

METADATA[METADATA == 'N/A'] = NA

# Replace all special characters with blank
myFix <- function(x) str_replace_all(x, '[^[:alnum:]]', '')
METADATA <- METADATA %>%
  dplyr::mutate(blood=ifelse(C4_Cell_Line_ID %in% bloodLines, "blood", "nonblood")) %>% 
  dplyr::mutate_each(funs(myFix), -UID, -C4_Cell_Line_ID, -biologicalSampleName,
                     -public, -exclude, -pass_qc) # fix them but don't touch some columns

# Set rownames
rownames(METADATA) = METADATA$UID

```

```{r}
#### Pre processing mRNA expression counts and metadata ####
metadata_keep <- METADATA %>% 
  filter(UID %in% colnames(EXP))

metadata_filtered <- metadata_keep %>%
  filter(UID != "H9P50") %>%
  filter(UID %in% colnames(EXP)) %>%
  filter(C4_Karyotype_Result != "abnormal")

REMOVED_UID <- setdiff(metadata_keep$UID, metadata_filtered$UID)
METADATA <- metadata_filtered
rownames(METADATA) <- metadata_filtered$UID
EXP <- EXP[, METADATA$UID]

```

```{r}
# Get probe related metadata
fData = get450KProbeMapping(rownames(EXP))
```

#### Synapse id of used matrices

|**Name**| **Synapse ID**|
|Filtered Expression|`r paste(EXP_ID,EXP_OBJ$properties$versionNumber,sep='.')`|

#### Perform differential expression
```{r diff.exp, fig.width=6, fig.height=6}
mat <- model.matrix(~ 0 + blood, METADATA)
colnames(mat) <- c("blood", "nonblood")

# Fit linear model
FIT = lmFit(as.matrix(EXP), design=mat)

CONT <- makeContrasts(contrasts="blood - nonblood", levels=mat)

# Refit contrasts
FIT.CONTRAST <- contrasts.fit(FIT, CONT)

# Estimate moderated t-statistics
FIT.CONTRAST <- eBayes(FIT.CONTRAST)
  
tmp <- topTable(FIT.CONTRAST, number=1e7)

tmp2 <- getUpDownGenes(tmp$adj.P.Val, 
                       tmp$logFC, 
                       rownames(tmp),
                       "blood", 
                       FC_CUTOFF = 0)
```
Number of differentially expressed genes at DE stage for adj.P.Val <= 0.05 and logFC >= 0.5 or <= -0.5
### Store files in synapse
```{r synapse.store, include = FALSE, eval=FALSE, cache=FALSE}
activityName='Differential Expression Analysis of minfi processed methylation data with mixed effects model'
  
thisFileName <- 'DiffExpAnal_methylation_blood.Rmd'
  
# Github link
thisRepo <- getRepo(repository = "Sage-Bionetworks/pcbc_c4_analysis", 
                    ref="branch", 
                    refName='methylblood')

thisFile <- getPermlink(repository = thisRepo,
                        repositoryPath=paste0('code/Rmd/', thisFileName))
    
# Make folder and populate with wiki
CODE <- Folder(name = 'Differential Expression Analysis Methylation Blood', parentId = parentId)
CODE <- synStore(CODE)

# Store results: logFC
write.table(logFC,file='./DiffExpAnal_methylation_blood_topTable.tsv',sep='\t',
            row.names=F, quote=F)

TT_OBJ <- File('./DiffExpAnal_methylation_blood_topTable.tsv', name = 'topTable', 
               parentId = CODE$properties$id)

TT_OBJ <- synStore(FC_OBJ, used = ALL_USED_IDs, 
                   activityName = activityName, executed=thisFile)
```
