---
date: '`r date()`'
---

```{r knit, include=FALSE, eval=FALSE}
synapseClient::synapseLogin()
knit2synapse::knitToFolderEntity(file = "./DiffExpAnal_methylation_blood.Rmd", 
                                 parentId = 'syn4231339',
                                 folderName = 'Differential Expression Analysis Methylation Blood', 
                                 overwrite=F)

```

```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
## It is assumed your working directory is where this file is

# Load required libraries
library('synapseClient')
library('limma')
library('reshape2')
library('data.table')
library('plyr')
library('dplyr')
library('knitr')
library('stringr')
library('minfi')
library('knit2synapse')
library(pcbcStats)
library(FDb.InfiniumMethylation.hg19)

## Needs the dev branch
library(rGithubClient)

synapseLogin()

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=TRUE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = FALSE)
```
### Download data
```{r setup, include=FALSE}
# Input Parameters
EXP_ID = 'syn2699024'
METADATA_ID <- "syn3156828"
parentId = 'syn4231339'

bloodLines <- c("CD90", "MEP", "GMP", "CD34DMSO", "CD34UNC", "MEG", 
                "SC13-046-32EPOMEP", "SC13-046-32RBC")
```

```{r getdata}
# Get filtered beta values
EXP_OBJ = synGet(EXP_ID)
ALL_USED_IDs = EXP_OBJ$properties$id
EXP = fread(getFileLocation(EXP_OBJ), data.table = F)
row.names(EXP) = EXP[,1]
EXP = dplyr::select(EXP,-(V1))
```

```{r metadata}
METADATA_OBJ = synTableQuery(paste('SELECT * FROM',METADATA_ID,sep=' '))
ALL_USED_IDs = c(ALL_USED_IDs, METADATA_OBJ@schema)
METADATA = METADATA_OBJ@values

METADATA[METADATA == 'N/A'] = NA

# Replace all special characters with blank
myFix <- function(x) str_replace_all(x, '[^[:alnum:]]', '')
METADATA <- METADATA %>%
  dplyr::mutate(blood=ifelse(C4_Cell_Line_ID %in% bloodLines, "blood", "nonblood")) %>% 
  dplyr::mutate_each(funs(myFix), -UID, -C4_Cell_Line_ID, -biologicalSampleName,
                     -public, -exclude, -pass_qc) # fix them but don't touch some columns

# Set rownames
rownames(METADATA) = METADATA$UID

```

```{r filter.metadata}
#### Pre processing mRNA expression counts and metadata ####
metadata_keep <- METADATA %>% 
  filter(UID %in% colnames(EXP))

metadata_filtered <- metadata_keep %>%
  filter(UID != "H9P50") %>%
  filter(UID %in% colnames(EXP)) %>%
  filter(C4_Karyotype_Result != "abnormal")

REMOVED_UID <- setdiff(metadata_keep$UID, metadata_filtered$UID)
METADATA <- metadata_filtered
rownames(METADATA) <- metadata_filtered$UID
EXP <- EXP[, METADATA$UID]

```

```{r variance.filter}
exp.variance <- apply(as.matrix(EXP), 1, var)

ind.remove1 <- exp.variance < sort(exp.variance)[round(0.25*length(exp.variance))]
ind.remove2 <- (rowSums(EXP <= 0.25)/dim(EXP)[2] == 1) | (rowSums(EXP >= 0.75)/dim(EXP)[2] == 1)

EXP <- EXP[!ind.remove1 & !ind.remove2,]

```

#### Perform differential expression
```{r diff.exp}
mat <- model.matrix(~ 0 + blood, METADATA)
colnames(mat) <- c("blood", "nonblood")

# Fit linear model
FIT = lmFit(as.matrix(EXP), design=mat)

CONT <- makeContrasts(contrasts="blood - nonblood", levels=mat)

# Refit contrasts
FIT.CONTRAST <- contrasts.fit(FIT, CONT)

# Estimate moderated t-statistics
FIT.CONTRAST <- eBayes(FIT.CONTRAST)

ind = which(METADATA[, 'blood'] == 'blood')
var1Mean = rowMeans(EXP[, METADATA$UID[ind], drop=F], na.rm=T)
var2Mean = rowMeans(EXP[, METADATA$UID[-ind], drop=F], na.rm=T)

tmp <- topTable(FIT.CONTRAST, coef="blood - nonblood", number=1e7, sort.by = "p")
tmp <- mutate(tmp, probeId=rownames(tmp), changeInBeta=var1Mean - var2Mean,
              Comparison=ifelse(logFC > 0, "blood__up", "blood__down")) %>%
  dplyr::select(probeId, everything())

write.csv(tmp, file='./DiffExpAnal_methylation_blood_topTable.csv', row.names=FALSE)
```
Number of differentially expressed genes at adj.P.Val <= 0.05 and abs(logFC) >= 0.5: `r nrow(topTable(FIT.CONTRAST, coef=1, number=Inf, p.value=0.05, lfc=0.5))`.

### Store files in synapse
```{r synapse.store, include=TRUE, eval=TRUE}
activityName='Differential Expression Analysis'
  
thisFileName <- 'DiffExpAnal_methylation_blood.Rmd'
  
# Github link
thisRepo <- getRepo(repository = "Sage-Bionetworks/pcbc_c4_analysis", 
                    ref="branch", 
                    refName='methylblood')

thisFile <- getPermlink(repository = thisRepo,
                        repositoryPath=paste0('code/Rmd/', thisFileName))
    
# Make folder and populate with wiki
CODE <- Folder(name = 'Differential Expression Analysis Methylation Blood', parentId = parentId)
CODE <- synStore(CODE)

TT_OBJ <- File('./DiffExpAnal_methylation_blood_topTable.csv', name = 'topTable', 
               parentId = CODE$properties$id)

TT_OBJ <- synStore(TT_OBJ, used = ALL_USED_IDs, 
                   activityName = activityName, executed=thisFile)
```
