---
title: "Differential expression analysis for eXpress aligned data"
author: "Thanneer Perumal"
date: "`r date()`"
output: html_document
---

```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
## It is assumed your working directory is where this file is
# Load required libraries
library('synapseClient')
library('limma')
library('edgeR')
library('RColorBrewer')
library('ctv')
library('ggplot2')
library('psych')
library('reshape2')
library('gplots')
library('vcd')
library('erer')
library('fpc')
library('data.table')
library('dplyr')
library('stringr')
library('knitr')

## Requires ggdendro
devtools::source_url("https://raw.githubusercontent.com/chr1swallace/random-functions/master/R/ggplot-heatmap.R")

# source('/mnt/Github/knit2synapse-1//R/knitFile2Synapse.R')
# knitFile2Synapse(file = "./DiffExpAnal_Express.Rmd", owner = 'syn3634560', wikiName = "Differential Expression Analysis Express",overwrite=T)

synapseLogin()

# source utility files from ../R/lib folder
file.sources = list.files('../R/lib',pattern="*.R")
file.sources = sapply(file.sources,function(x){return(paste('../R/lib',x,sep='/'))})
tmp = sapply(file.sources,source,.GlobalEnv)

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE)
```

### Download data

```{r setup, include=FALSE}
# Input Parameters
DESIGN_ID = 'syn3471253';
EXPR_ID = 'syn3471238';
WEIGHTS_ID = 'syn3471245';
COVARIATES_ID = 'syn3628878';

SYNAPSE_STORE = T
parentId = 'syn3471223'

# Specify factor and continuous covarites pool (adjusted covariates will only be the subset of these covariates)
FactorCovariates = c('Diffname_short','run','lane','index','Cell_Type_of_Origin', 'Cell_Line_of_Origin', 'Tissue_of_Origin', 'Reprogramming_Vector_Type','Reprogramming_Gene_Combination', 'Donor_Life_Stage','Originating_Lab','Gender','Donor_ID','Disease','Race','Ethnicity','Cell_Line_Type')
ContCovariates = NULL
```
Synapse id of design, expression, weights and covariates matrix used for the analysis are `r DESIGN_ID`,`r EXPR_ID`,`r WEIGHTS_ID`, and `r COVARIATES_ID`

```{r getdata, cache=TRUE, include=FALSE}
# Get design matrix
DESIGN_OBJ = synGet(DESIGN_ID);
ALL_USED_IDs = c(DESIGN_OBJ$properties$id);
DESIGN = read.table(getFileLocation(DESIGN_OBJ),sep='\t',header=T,row.names=1, check.names=FALSE)
colnames(DESIGN) = gsub("[[:punct:]]", "_",colnames(DESIGN))

# Get expression matrix
EXPR_OBJ = synGet(EXPR_ID);
ALL_USED_IDs = c(ALL_USED_IDs,EXPR_OBJ$properties$id);
EXPR = read.table(getFileLocation(EXPR_OBJ),sep='\t',header=T,row.names=1, check.names=FALSE)

# Get weighting matrix
WEIGHTS_OBJ = synGet(WEIGHTS_ID);
ALL_USED_IDs = c(ALL_USED_IDs,WEIGHTS_OBJ$properties$id);
WEIGHTS = read.table(getFileLocation(WEIGHTS_OBJ),sep='\t',header=T,row.names=1, check.names=FALSE)

# Get weighting matrix
COVARIATES_OBJ = synGet(COVARIATES_ID);
ALL_USED_IDs = c(ALL_USED_IDs,COVARIATES_OBJ$properties$id);
COVARIATES = read.table(getFileLocation(COVARIATES_OBJ),sep='\t',header=T,row.names=1, check.names=FALSE)
```

### Calculate residuals
Remove run749, Donor IDD078, index27 from design matrix and residualise run, index and Donor ID information
```{r calculate.residual, echo=FALSE, include=FALSE}
DESIGN <- dplyr::select(DESIGN, -one_of('run749','Donor_IDD078','index27'))
RESIDUAL.EXPR <- calcResiduals(EXPR, DESIGN, varsToAddBackIn=c("Diffname_shortDE","Diffname_shortEB","Diffname_shortECTO","Diffname_shortMESO_EARLY", "Diffname_shortMESO_LATE", "Diffname_shortSC"), sampleWeights=as.matrix(WEIGHTS))
```

Get differentially expressed genes using limma package with following co-efficients `r colnames(DESIGN)` in linear model

Degenerate columns in the new design are `r linColumnFinder(DESIGN)`

#### Differential expression 

Differentiation stages (Male & Female vs Male only vs Female only analysis)
```{r diffstate, cache=TRUE, include=FALSE}
Gender <- c('all','male','female')
diffExpAnal <- function(gend,EXPR,DESIGN,WEIGHTS,COVARIATES){  
  if(gend == 'all'){
    ind <- colnames(EXPR) %in% rownames(COVARIATES)
    print(sum(ind))
    FIT <- lmFit(EXPR[,ind], design = DESIGN[ind,], weights = WEIGHTS[,ind])
  } else {
    ind <- colnames(EXPR) %in% rownames(COVARIATES[COVARIATES$Gender == gend,])
    print(sum(ind))
    FIT <- lmFit(EXPR[,ind], design = DESIGN[ind,], weights = WEIGHTS[,ind])
  }
                
  # Make contrast to check differential expression between different differentiation stages
  CONT.NAMES <- colnames(DESIGN)[grep('Diffname_short',colnames(DESIGN))]
  CONT.NAMES <- combn(CONT.NAMES,2)
  CONT.NAMES <- apply(CONT.NAMES,2,paste,collapse='-')
  
  CONT <- makeContrasts(contrasts=CONT.NAMES,levels=colnames(FIT$coefficients))
  
  # Refit contrasts
  FIT.CONTRAST <- contrasts.fit(FIT,CONT)
  
  # Estimate moderated t-statistics
  FIT.CONTRAST <- eBayes(FIT.CONTRAST)
  
  # Obtain all the differential expession combinations
  DIFF.EXP <- list()
  DIFF.EXP$logFC <- data.frame(row.names = rownames(EXPR))
  DIFF.EXP$adj.P.Val <- data.frame(row.names = rownames(EXPR))
  
  for (i in CONT.NAMES){
    tmp <- topTable(FIT.CONTRAST, coef=i, number=dim(EXPR)[1])    
    DIFF.EXP$logFC[,str_replace_all(i,"Diffname_short","")] <- tmp[rownames(DIFF.EXP$logFC),'logFC']
    DIFF.EXP$adj.P.Val[,str_replace_all(i,"Diffname_short","")] <- tmp[rownames(DIFF.EXP$adj.P.Val),'adj.P.Val']
  }
  
  # Get number of significantly differentialy expressed genes with adj.P.Val <= 0.001 and abs(logFC) >= 2
  DIFF.EXP$SIG.EXP <- DIFF.EXP$adj.P.Val<=0.05 & abs(DIFF.EXP$logFC) >= 2   
  DIFF.EXP$NUM.SIG.EXP <- colSums(DIFF.EXP$SIG.EXP)
  return(DIFF.EXP=DIFF.EXP)
}
DIFFNAME <- tapply(Gender,Gender,diffExpAnal,as.matrix(RESIDUAL.EXPR),as.matrix(DESIGN[,grep('Diffname_short',colnames(DESIGN))]),as.matrix(WEIGHTS),COVARIATES)
save(DIFFNAME,file='DifferentialExpressionDiffName.RData')
```

Number of differentially expressed genes
```{r significant.sets,cache=TRUE}
kable(sapply(DIFFNAME,function(x){return(x$NUM.SIG.EXP)}))
```

Male and female only genes
```{r gender.significant.sets,cache=TRUE}
# Get intersection and difference of gene specific to male and female
NUM.SIG.SETS <- data.frame(row.names=colnames(DIFFNAME[['male']]$SIG.EXP))
for (j in rownames(NUM.SIG.SETS)){
  NUM.SIG.SETS[j,'common'] <- sum(DIFFNAME[['male']]$SIG.EXP[rownames(EXPR),j] & DIFFNAME[['female']]$SIG.EXP[rownames(EXPR),j])
  NUM.SIG.SETS[j,'male.only'] <- sum(DIFFNAME[['male']]$SIG.EXP[rownames(EXPR),j] & !DIFFNAME[['female']]$SIG.EXP[rownames(EXPR),j])
  NUM.SIG.SETS[j,'female.only'] <- sum(!DIFFNAME[['male']]$SIG.EXP[rownames(EXPR),j] & DIFFNAME[['female']]$SIG.EXP[rownames(EXPR),j])
}
kable(NUM.SIG.SETS)
```

Cell line type (iPSc vs hESc)

```{r cell.line.type, cache=TRUE, include=FALSE}
# Since "Cell_Line_TypeiPSC = 1*Diffname_shortDE + 1*Diffname_shortEB + 1*Diffname_shortECTO + 1*Diffname_shortMESO_EARLY + 1*Diffname_shortMESO_LATE + 1*Diffname_shortSC + -1*Cell_Line_TypeESC" we remove Diffname_shortDE

NEW.DESIGN <- getDesignMatrix(dplyr::select(COVARIATES,one_of('Diffname_short','Cell_Line_Type')), Intercept = F)
DESIGN <- NEW.DESIGN$design[,-(grep('Diffname_shortDE',colnames(NEW.DESIGN$design)))]

FIT <- lmFit(RESIDUAL.EXPR, design = DESIGN, weights = as.matrix(WEIGHTS))

# Make contrast to check differential expression between different differentiation stages
CONT.NAMES <- colnames(DESIGN)[grep('Cell_Line_Type',colnames(DESIGN))]
CONT.NAMES <- combn(CONT.NAMES,2)
CONT.NAMES <- apply(CONT.NAMES,2,paste,collapse='-')

CONT <- makeContrasts(contrasts=CONT.NAMES,levels=colnames(FIT$coefficients))

# Refit contrasts
FIT.CONTRAST <- contrasts.fit(FIT,CONT)

# Estimate moderated t-statistics
FIT.CONTRAST <- eBayes(FIT.CONTRAST)

# Obtain all the differential expession combinations
CELL.LINE.TYPE <- list()
CELL.LINE.TYPE$logFC <- data.frame(row.names = rownames(EXPR))
CELL.LINE.TYPE$adj.P.Val <- data.frame(row.names = rownames(EXPR))

for (i in CONT.NAMES){
  tmp <- topTable(FIT.CONTRAST, coef=i, number=dim(EXPR)[1])    
  CELL.LINE.TYPE$logFC[,str_replace_all(i,"Cell_Line_Type","")] <- tmp[rownames(CELL.LINE.TYPE$logFC),'logFC']
  CELL.LINE.TYPE$adj.P.Val[,str_replace_all(i,"Cell_Line_Type","")] <- tmp[rownames(CELL.LINE.TYPE$adj.P.Val),'adj.P.Val']
}

# Get number of significantly differentialy expressed genes with adj.P.Val <= 0.001 and abs(logFC) >= 2
CELL.LINE.TYPE$SIG.EXP <- CELL.LINE.TYPE$adj.P.Val<=0.001 & abs(CELL.LINE.TYPE$logFC) >= 2   
CELL.LINE.TYPE$NUM.SIG.EXP <- colSums(CELL.LINE.TYPE$SIG.EXP)
  
save(CELL.LINE.TYPE,file='DifferentialExpressionCellLineType.RData')
```

Reprogramming vector type

```{r reprogramming.vector, cache=TRUE}
tmp <- linColumnFinder(getDesignMatrix(dplyr::select(COVARIATES,one_of('Diffname_short','Donor_ID','Reprogramming_Vector_Type')),Intercept = F)$design)
print('Some reprogramming vector types are linear combination of donors')
print(tmp$relations)

NEW.DESIGN <- getDesignMatrix(dplyr::select(COVARIATES,one_of('Diffname_short','Reprogramming_Vector_Type')), Intercept = F)
DESIGN <- NEW.DESIGN$design[,-(grep('Diffname_shortDE',colnames(NEW.DESIGN$design)))]
colnames(DESIGN) <- gsub(" ","",colnames(DESIGN))

sigCovars <- runPCAandPlotCorrelations(RESIDUAL.EXPR,getDesignMat2Fact(getDesignMatrix(COVARIATES,Intercept=F)$design, FactorCovariates), 'residual expression')
print(sigCovars$PC_res[[2]]$plotData)

FIT <- lmFit(RESIDUAL.EXPR, design = DESIGN, weights = as.matrix(WEIGHTS))

# Make contrast to check differential expression between different differentiation stages
CONT.NAMES <- colnames(DESIGN)[grep('Reprogramming_Vector_Type',colnames(DESIGN))]
CONT.NAMES <- combn(CONT.NAMES,2)
CONT.NAMES <- apply(CONT.NAMES,2,paste,collapse='-')

CONT <- makeContrasts(contrasts=CONT.NAMES,levels=colnames(FIT$coefficients))

# Refit contrasts
FIT.CONTRAST <- contrasts.fit(FIT,CONT)

# Estimate moderated t-statistics
FIT.CONTRAST <- eBayes(FIT.CONTRAST)

# Obtain all the differential expession combinations
DIFF.EXP.REPROGRAMMING.VECTOR <- list()
DIFF.EXP.REPROGRAMMING.VECTOR$logFC <- data.frame(row.names = rownames(EXPR))
DIFF.EXP.REPROGRAMMING.VECTOR$adj.P.Val <- data.frame(row.names = rownames(EXPR)) 

# Get differential expression
for (i in CONT.NAMES){
  tmp <- topTable(FIT.CONTRAST, coef=i, number=dim(RESIDUAL.EXPR)[1])
  DIFF.EXP.REPROGRAMMING.VECTOR$logFC[,str_replace_all(i,"Reprogramming_Vector_Type","")] <- tmp[rownames(DIFF.EXP.REPROGRAMMING.VECTOR$logFC),'logFC']
  DIFF.EXP.REPROGRAMMING.VECTOR$adj.P.Val[,str_replace_all(i,"Reprogramming_Vector_Type","")] <- tmp[rownames(DIFF.EXP.REPROGRAMMING.VECTOR$adj.P.Val),'adj.P.Val']
}

# Get number of significantly differentialy expressed genes with adj.P.Val <= 0.001 and abs(logFC) >= 2
DIFF.EXP.REPROGRAMMING.VECTOR$SIG.EXP <- DIFF.EXP.REPROGRAMMING.VECTOR$adj.P.Val<=0.05 & abs(DIFF.EXP.REPROGRAMMING.VECTOR$logFC) >= 0   
DIFF.EXP.REPROGRAMMING.VECTOR$NUM.SIG.EXP <- colSums(DIFF.EXP.REPROGRAMMING.VECTOR$SIG.EXP)
save(DIFF.EXP.REPROGRAMMING.VECTOR,file='DifferentialExpressionReprogrammingVectorType.RData')

tmp <- as.data.frame(DIFF.EXP.REPROGRAMMING.VECTOR$NUM.SIG.EXP)
kable(tmp)
```

Pvalue and fold change of some important genes
```{r print.imp.genes}
FC <- merge(rownameToFirstColumn(DIFFNAME$all$logFC,'GeneNames'),rownameToFirstColumn(CELL.LINE.TYPE$logFC,'GeneNames'),by='GeneNames')
PVAL <- merge(rownameToFirstColumn(DIFFNAME$all$adj.P.Val,'GeneNames'),rownameToFirstColumn(CELL.LINE.TYPE$adj.P.Val,'GeneNames'),by='GeneNames')

kable(filter(FC, GeneNames %in% c('LARGE','FRG1','FRG1B','PEG10','SPEG')))

kable(filter(PVAL, GeneNames %in% c('LARGE','FRG1','FRG1B','PEG10','SPEG')))
```

Perform enrichment analysis for differentiation
```{r enrichment.analysis, include=FALSE}
enrichmentAnalysis <- function(ind,PVAL,FC){
  allGenes <- PVAL[row.names(FC),ind] <= 0.05 & abs(FC[,ind]) >= 2
  names(allGenes) <- row.names(FC)
  
  # Calculate probability weighting function based on length of gene
  pwf = nullp(allGenes, 'hg19', 'geneSymbol',plot.fit=F)
  
  # Calculate go enrichment using Wallenius method    
  GO.wall = goseq(pwf, 'hg19', 'geneSymbol', method = "Wallenius")
  GO.wall$over_represented_adj_pvalue = p.adjust(GO.wall$over_represented_pvalue, method = 'bonferroni')
  GO.wall$under_represented_adj_pvalue = p.adjust(GO.wall$under_represented_pvalue, method = 'bonferroni')

  return(GO.wall)
}
row.names(PVAL) <- PVAL$GeneNames
row.names(FC) <- FC$GeneNames

tmp <- enrichmentAnalysis(colnames(PVAL)[2],PVAL,FC)
ENRICH <- tmp[,c("category","over_represented_adj_pvalue","term","ontology")]
rownames(ENRICH) <- tmp$category
ENRICH <- dplyr::rename(ENRICH, "DE-EB" = over_represented_adj_pvalue)

for (cols in colnames(PVAL)[-(1:2)]){
  tmp <- enrichmentAnalysis(cols,PVAL,FC)
  rownames(tmp) <- tmp$category
  ENRICH[,cols] <- tmp[rownames(ENRICH),c("over_represented_adj_pvalue")]
}
save(DIFFNAME,file='DifferentialExpressionDiffName_Enrichment.RData')
```

Store files in synapse
 
```{r synapse.store, include=FALSE}
if(SYNAPSE_STORE){
  activityName='Differential Expression Analysis of eXpress aligned data'
  
  CODE <- File('./DiffExpAnal_Express.Rmd',name = 'Differential Expression Analysis Express',parentId = parentId)
  CODE <- synStore(CODE, used = ALL_USED_IDs,activityName = activityName, executed='https://github.com/th1vairam/pcbc_c4_analysis/blob/394582def33f10389e416d3fa73b3bbe8af969c4/code/Rmd/DiffExpAnal_Express.Rmd')
      
  write.table(FC,file='./DiffnameShort_FoldChange_Express.tsv',sep='\t',row.names=F,quote=F)
  FC <- File('./DiffnameShort_FoldChange_Express.tsv',name = 'Differential Expression Analysis Express logFC',parentId = parentId)
  FC <- synStore(FC, used = ALL_USED_IDs, activityName = activityName, executed=CODE$properties$id)
    
  write.table(PVAL,file='./DiffnameShort_Pvalue_Express.tsv',sep='\t',row.names=F,quote=F)
  PVAL <- File('./DiffnameShort_Pvalue_Express.tsv',name = 'Differential Expression Analysis Express pvalue',parentId = parentId)
  PVAL <- synStore(PVAL, used = ALL_USED_IDs, activityName = activityName, executed=CODE$properties$id)
  
  write.table(ENRICH,file='./DiffnameShort_Enrichment_Express.tsv',sep='\t',row.names=F,quote=F)
  ENRICH1 <- File('./DiffnameShort_Enrichment_Express.tsv',name = 'Differential Expression Analysis Express enrichment',parentId = parentId)
  ENRICH1 <- synStore(ENRICH1, used = ALL_USED_IDs, activityName = activityName, executed=CODE$properties$id)
}
```