---
title: "Differential expression analysis for eXpress aligned data with mixed effects modelling at each diff state"
author: "Thanneer Perumal"
date: "`r date()`"
output: html_document
---
```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
## It is assumed your working directory is where this file is

# Clear R console screen output
cat("\014")  

# Load required libraries
library('synapseClient')
library('RColorBrewer')
library('ggplot2')
library('gplots')
library('limma')
library('edgeR')
library('ctv')
library('psych')
library('reshape2')
library('vcd')
library('erer')
library('fpc')
library('data.table')
library('dplyr')
library('knitr')
library('stringr')

## Needs the dev branch
library(rGithubClient)

## Requires ggdendro
devtools::source_url("https://raw.githubusercontent.com/chr1swallace/random-functions/master/R/ggplot-heatmap.R")

# source('/mnt/Github/knit2synapse-1/R/knitFile2Synapse.R')
# knitFile2Synapse(file = "./DiffExpAnal_Express_mixedEffects_atEachDiffState.Rmd", owner = 'syn4223183', wikiName = "Differential Expression Analysis Express Mixed Effects At Each DiffState",overwrite=F)

synapseLogin()

# source utility files from ../R/lib folder
file.sources = list.files('../R/lib',pattern="*.R")
file.sources = sapply(file.sources,function(x){return(paste('../R/lib',x,sep='/'))})
tmp = sapply(file.sources,source,.GlobalEnv)

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE)
```
### Download data
```{r setup, include=FALSE}
# Input Parameters
COUNT_ID = 'syn3446250'
METADATA_ID = 'syn3156503'

SYNAPSE_STORE = T
parentId = 'syn3471223'

# Specify factor and continuous covarites pool (adjusted covariates will only be the subset of these covariates)
FactorCovariates = c('Diffname_short', 'run', 'lane', 'PassageAtThaw', 'PassageAtHarvest', 'Cell_Line_Type', 'Cell_Line_of_Origin', 'Tissue_of_Origin', 'Reprogramming_Gene_Combination', 'Culture_Conditions', 'Donor_Life_Stage', 'Race', 'Ethnicity' , 'Gender', 'Disease', 'Originating_Lab', 'Donor_ID', 'Cell_Type_of_Origin_Level2', 'Reprogramming_Vector_Type')
ContCovariates = NULL
```
Synapse id of count matrix used for the analysis is `r COUNT_ID` and the synapse id of meta data table used for the analysis is `r METADATA_ID`. 

Factor covariates considered for analysis are  `r paste(gsub('_','\\\\_',FactorCovariates),collapse=',')`, and continuous covariates considered for the analysis are  `r paste(gsub('_','\\\\_',ContCovariates),collapse=',')`.

Obtain count matrix and metadata from synapse.
```{r getdata, cache=TRUE, include=FALSE}
# Get count matrix
COUNT_OBJ = synGet(COUNT_ID)
ALL_USED_IDs = COUNT_OBJ$properties$id
COUNT = fread(getFileLocation(COUNT_OBJ), data.table=FALSE)
row.names(COUNT) = COUNT[,1]

# Get metadata
METADATA_OBJ = synTableQuery(paste('SELECT * FROM',METADATA_ID,sep=' '))
ALL_USED_IDs[length(ALL_USED_IDs)+1] = METADATA_OBJ@schema
METADATA = METADATA_OBJ@values
```
Preprocess counts matrix and metadata.
```{r preprocessing, include=FALSE}
# Preprocess metadata
METADATA[METADATA == 'N/A'] = NA

# Replace all special characters with blank
myFix <- function(x) str_replace_all(x, '[^[:alnum:]]', '')
METADATA <- METADATA %>%
  dplyr::mutate_each(funs(myFix), -UID, -C4_Cell_Line_ID, -biologicalSampleName) # fix them but don't touch some columns

# Set rownames
rownames(METADATA) = METADATA$UID
```
### Preprocess data
* Remove somatic samples and samples with not type.
* Remove samples that failed QC and samples classified as exclude.
* Remove samples with abnormal karyotypes.
```{r filtering, echo=TRUE}
#### Pre processing mRNA expression counts and metadata ####
metadata_filtered <- 
  METADATA %>%
  filter(Diffname_short != "") %>%
  filter(UID %in% colnames(COUNT)) %>%
  filter(Cell_Type == "PSC") %>%  
  filter(pass_qc == "TRUE") %>%
  filter(exclude != "TRUE") %>%
  filter(C4_Karyotype_Result != "abnormal")

REMOVED_UID <- setdiff(colnames(COUNT), metadata_filtered$UID)
METADATA <- METADATA[metadata_filtered$UID,]
COUNT <- COUNT[, METADATA$UID]
```
The following `r length(REMOVED_UID)` samples were removed:

`r paste(gsub('_','\\\\_',REMOVED_UID), collapse= ',')` 

### CPM Normalisation
Determine design matrix for normalisation and differential expression analysis. 

Remove genes that have less than 1 cpm counts in at least 50% of each of the individual differentiation stages.
```{r cpmnormalisation}
tmp <- tapply(colnames(COUNT),
              factor(METADATA$Diffname_short),
              function(cols,COUNT){PROCESSED_COUNTS = getGeneFilteredGeneExprMatrix(COUNT[,cols])},
              COUNT)

ugenes <- c()
for (i in unique(METADATA$Diffname_short)) {
  ugenes <- unique(c(ugenes,tmp[[i]]$filteredExprMatrix$genes[,1]))
}

COUNT <- COUNT[ugenes,,drop=F]
PROCESSED_COUNTS = getGeneFilteredGeneExprMatrix(COUNT,MIN_GENE_CPM=0,
                                                 MIN_SAMPLE_PERCENT_WITH_MIN_GENE_CPM=0)
```
```{r covariates.clustering}
# Find inter relation between factor covariates
COVARIATES = METADATA[,c(FactorCovariates,ContCovariates)]

# Convert factor covariates to factors
COVARIATES[,FactorCovariates] = lapply(COVARIATES[,FactorCovariates], factor)
```
### Residualisation
Fit mixed effect linear model with Diffname short, Gender, Cell Type of Origin Level2, Donor Life Stage, lane, Culture Conditions, run as fixed effects and Donor ID as random effects

Residualise Donor Life Stage, lane, run and Donor ID by adding back Diffname short, Gender, Cell Type of Origin Level2 and Culture Conditions
```{r find.residuals, cache=TRUE}
# Get design matrix
DESIGN = getDesignMatrix(dplyr::select(COVARIATES, Diffname_short, Gender, Cell_Type_of_Origin_Level2, Donor_Life_Stage, lane, Culture_Conditions, run), Intercept = F)
DESIGN = DESIGN$design[,linColumnFinder(DESIGN$design)$indepCols]
  
# Estimate correlation of random effects
EXPR = voom(COUNT, design=DESIGN, plot=F)

# Calculate correlation between donors
correlation <- duplicateCorrelation(EXPR, block=COVARIATES$Donor_ID)
  
# Fit linear model using mixed effects design
FIT = lmFit(EXPR, block=COVARIATES$Donor_ID, correlation = correlation$cor)

# Get residuals  
RESID.EXPR = calcResiduals(EXPR$E, DESIGN, varsToAddBackIn = c( "Diffname_shortDE","Diffname_shortEB","Diffname_shortECTO","Diffname_shortMESO15","Diffname_shortMESO30","Diffname_shortMESO5","Diffname_shortSC","Genderfemale","Cell_Type_of_Origin_Level2blood","Cell_Type_of_Origin_Level2fibroblast","Cell_Type_of_Origin_Level2innercellmass","Culture_ConditionsDMEMF12KOSR01mMBME01mMNEAA01mMLgluatime20ngmlbFGF","Culture_ConditionsE8","Culture_ConditionsFeederandKOSRcultureconditions","Culture_ConditionsiPSCsweregeneratedandintiallymaintainedusingMEFKSRculturesAftertransfertoWiCelliPSCsexpandedinFeederIndependentTeSR1Medium","Culture_ConditionsStandardTESR1MatrigelorKOSRMEFs","Culture_ConditionsStromalnonprimed","Culture_ConditionsStromalprimed"))
```

### Differential Expression Analysis At Each Differentiation Stage
Obtain differentially expressed genes with the following covariates (as obtained from covariates analysis) in the model: 
* Fixed Effects: Reprogramming vector type, Reprogramming gene combination, Gender, Cell Type of Origin Level2 and Culture Conditions
Function to perform differential expression analysis at each of the differentiation stages
```{r differential.expression, cache=TRUE}
getDifferentialExpression <- function(COVAR, EXPR, cols, varOfInterest){
  Name <- paste0(unique(COVAR$Diffname_short),' differentiation stage')
  print(Name)
  
  # Get covariates matrix
  ind <- which(sapply(COVAR,function(x){length(unique(x))})!=1)
  COVAR <- droplevels(COVAR[,ind])
      
  # Get expression matrix
  EXPR <- EXPR[,rownames(COVAR)]
  
  # Post adjusted design matrix
  DM = getDesignMatrix(COVAR[, intersect(cols, colnames(COVAR))], Intercept = F)
  DM$design = DM$design[,linColumnFinder(DM$design)$indepCols]
  
  # Model Fit
  FIT = lmFit(EXPR,DM$design)
  
  # Residuals after normalisation
  RESIDUAL.GENE_EXPRESSION = residuals.MArrayLM(FIT, EXPR)
    
  # Residual covariates to choose from
  residCovars <- setdiff(c(FactorCovariates,ContCovariates), postAdjustCovars)
    
  # Find PC of residual gene expression and significant covariates that are highly correlated with PCs
  residualSigCovars = runPCAandPlotCorrelations(RESIDUAL.GENE_EXPRESSION,
                                                getDesignMatrix(COVAR,Intercept=F)$design,
                                                'all adjusted design(voom-normalized)',
                                                isKeyPlot=TRUE)
  
  if(length(residualSigCovars$significantCovars) != 0)
    print(paste("Follwoing variables needs to be adjusted in",Name,paste(residualSigCovars$significantCovars, collapse=',')))
  
  # Make contrast to check differential expression between different differentiation stages
  CONT.NAMES <- colnames(DM$design)[grep(varOfInterest,colnames(DM$design))]
  CONT.NAMES <- combn(CONT.NAMES,2)
  CONT.NAMES <- apply(CONT.NAMES,2,paste,collapse='-')
  
  CONT <- makeContrasts(contrasts=CONT.NAMES,levels=colnames(FIT$coefficients))
  colnames(CONT) <- sapply(colnames(CONT),function(x){x <- gsub(varOfInterest,'',x);
                                                      x <- gsub('-','_',x)})
  
  # Refit contrasts
  FIT.CONTRAST <- contrasts.fit(FIT,CONT)
  
  # Estimate moderated t-statistics
  FIT.CONTRAST <- eBayes(FIT.CONTRAST)
  
  # Obtain all the differential expession combinations
  DEXP <- list()
  DEXP$logFC <- data.frame(row.names = rownames(EXPR))
  DEXP$adj.P.Val <- data.frame(row.names = rownames(EXPR))
  DEXP$SIG.SETS <- data.frame()
  
  for (i in colnames(CONT)){
    tmp <- topTable(FIT.CONTRAST, coef=i, number=dim(EXPR)[1])    
    DEXP$logFC[,i] <- tmp[rownames(DEXP$logFC),'logFC']
    DEXP$adj.P.Val[,i] <- tmp[rownames(DEXP$adj.P.Val),'adj.P.Val'] 
    
    DEXP$SIG.SETS <- rbind(DEXP$SIG.SETS,
                           getUpDownGenes(DEXP$adj.P.Val[,i], DEXP$logFC[,i], rownames(DEXP$logFC), i))
  }
  
  # Get number of significantly differentialy expressed genes with adj.P.Val <= 0.05 and logFC >= 0
  DEXP$SIG.EXP.POS <- DEXP$adj.P.Val<=0.05 & DEXP$logFC >= 0   
  DEXP$NUM.SIG.EXP.POS <- colSums(DEXP$SIG.EXP.POS)
  
  # Get number of significantly differentialy expressed genes with adj.P.Val <= 0.05 and logFC <= 0
  DEXP$SIG.EXP.NEG <- DEXP$adj.P.Val<=0.05 & DEXP$logFC <= 0    
  DEXP$NUM.SIG.EXP.NEG <- colSums(DEXP$SIG.EXP.NEG)
  
  return(list(FIT = FIT,FIT.CONTRAST = FIT.CONTRAST, DEXP = DEXP))
}
```
### Reprogramming Gene Combinations
Number of significantly expressed genes at an FDR <= 0.05
```{r rep.gene.combo, cache=TRUE}
COV <- split(COVARIATES,COVARIATES$Diffname_short)
REP.GENE.COMBO <- lapply(COV,getDifferentialExpression,RESID.EXPR, cols = c("Reprogramming_Gene_Combination", "Reprogramming_Vector_Type", "Gender", "Cell_Type_of_Origin_Level2", "Culture_Conditions"), varOfInterest = "Reprogramming_Gene_Combination")

tmp <- lapply(REP.GENE.COMBO,function(x){return(data.frame(Contrast = names(x$DEXP$NUM.SIG.EXP.NEG), POS = x$DEXP$NUM.SIG.EXP.POS, NEG = x$DEXP$NUM.SIG.EXP.NEG))})
tmp <- join_all(tmp,by='Contrast',match='all')
colnames(tmp)[2:dim(tmp)[2]] = rbind(paste(names(REP.GENE.COMBO),'UP',sep='.'),paste(names(REP.GENE.COMBO),'DOWN',sep='.'))[1:(length(REP.GENE.COMBO)*2)]
kable(tmp)
```

### Reprogramming Vector Type
Number of significantly expressed genes at an FDR <= 0.05
```{r rep.vect.type, cache=TRUE}
REP.VECT.TYPE <- lapply(COV,getDifferentialExpression,RESID.EXPR, cols = c("Reprogramming_Vector_Type", "Reprogramming_Gene_Combination", "Gender", "Cell_Type_of_Origin_Level2", "Culture_Conditions"), varOfInterest = "Reprogramming_Vector_Type")

tmp <- lapply(REP.VECT.TYPE,function(x){return(data.frame(Contrast = names(x$DEXP$NUM.SIG.EXP.NEG), POS = x$DEXP$NUM.SIG.EXP.POS, NEG = x$DEXP$NUM.SIG.EXP.NEG))})
tmp <- join_all(tmp,by='Contrast',match='all')
colnames(tmp)[2:dim(tmp)[2]] = rbind(paste(names(REP.VECT.TYPE),'UP',sep='.'),paste(names(REP.VECT.TYPE),'DOWN',sep='.'))[1:(length(REP.VECT.TYPE)*2)]
kable(tmp)
```

### Gender
Number of significantly expressed genes at an FDR <= 0.05
```{r gender, cache=TRUE}
GENDER <- lapply(COV,getDifferentialExpression,RESID.EXPR, cols = c("Gender", "Reprogramming_Vector_Type", "Reprogramming_Gene_Combination", "Cell_Type_of_Origin_Level2", "Culture_Conditions"), varOfInterest = "Gender")

tmp <- lapply(GENDER,function(x){return(data.frame(Contrast = names(x$DEXP$NUM.SIG.EXP.NEG), POS = x$DEXP$NUM.SIG.EXP.POS, NEG = x$DEXP$NUM.SIG.EXP.NEG))})
tmp <- join_all(tmp,by='Contrast',match='all')
colnames(tmp)[2:dim(tmp)[2]] = rbind(paste(names(GENDER),'UP',sep='.'),paste(names(GENDER),'DOWN',sep='.'))[1:(length(GENDER)*2)]
kable(tmp)
```

### Cell Type of Origin Level2
Number of significantly expressed genes at an FDR <= 0.05
```{r cell.type.orig, cache=TRUE}
CELL.TYPE.ORIG.L2 <- lapply(COV,getDifferentialExpression,RESID.EXPR, cols = c("Cell_Type_of_Origin_Level2", "Gender", "Reprogramming_Vector_Type", "Reprogramming_Gene_Combination", "Culture_Conditions"), varOfInterest = "Cell_Type_of_Origin_Level2")

tmp <- lapply(CELL.TYPE.ORIG.L2,function(x){return(data.frame(Contrast = names(x$DEXP$NUM.SIG.EXP.NEG), POS = x$DEXP$NUM.SIG.EXP.POS, NEG = x$DEXP$NUM.SIG.EXP.NEG))})
tmp <- join_all(tmp,by='Contrast',match='all')
colnames(tmp)[2:dim(tmp)[2]] = rbind(paste(names(CELL.TYPE.ORIG.L2),'UP',sep='.'),paste(names(CELL.TYPE.ORIG.L2),'DOWN',sep='.'))[1:(length(CELL.TYPE.ORIG.L2)*2)]
kable(tmp)
```

### Culture Conditions
Number of significantly expressed genes at an FDR <= 0.05
```{r culture.cond, cache=TRUE}
CULT.COND <- lapply(COV[-(4)],getDifferentialExpression,RESID.EXPR, cols = c("Culture_Conditions", "Cell_Type_of_Origin_Level2", "Gender", "Reprogramming_Vector_Type", "Reprogramming_Gene_Combination"), varOfInterest = "Culture_Conditions")

tmp <- lapply(CULT.COND,function(x){return(data.frame(Contrast = names(x$DEXP$NUM.SIG.EXP.NEG), POS = x$DEXP$NUM.SIG.EXP.POS, NEG = x$DEXP$NUM.SIG.EXP.NEG))})
tmp <- join_all(tmp,by='Contrast',match='all')
colnames(tmp)[2:dim(tmp)[2]] = rbind(paste(names(CULT.COND),'UP',sep='.'),paste(names(CULT.COND),'DOWN',sep='.'))[1:(length(CULT.COND)*2)]
kable(tmp)
```
Get pvalue and fold changes of some important genes
```{r print.imp.genes}
ALL.RESULTS <- list(REP.GENE.COMBO, REP.VECT.TYPE, GENDER, CELL.TYPE.ORIG.L2, CULT.COND)
FC <- list()
PVAL <- list()
SIG.SETS <- data.frame()

for(i in 1:length(ALL.RESULTS)){
  for (diff_state in names(ALL.RESULTS[[i]])){
    tmp <- ALL.RESULTS[[i]][[diff_state]]$DEXP$logFC
    colnames(tmp) <- paste(diff_state,colnames(tmp),sep='_')
    FC <- c(FC,list(rownameToFirstColumn(tmp,'GeneNames')))
    
    tmp <- ALL.RESULTS[[i]][[diff_state]]$DEXP$adj.P.Val
    colnames(tmp) <- paste(diff_state,colnames(tmp),sep='_')
    PVAL <- c(PVAL,list(rownameToFirstColumn(tmp,'GeneNames')))
    
    tmp <- ALL.RESULTS[[i]][[diff_state]]$DEXP$SIG.SETS
    if (dim(tmp)[1] > 0){
      tmp$Comparison <-  paste(diff_state, tmp$Comparison, sep = '_')
      SIG.SETS <- rbind(SIG.SETS,tmp)
    }
    
  }
}

FC <- join_all(FC,by = 'GeneNames', match = 'all')
PVAL <- join_all(PVAL, by = 'GeneNames', match = 'all')
```
Fold change:
```{r FC}
kable(filter(FC, GeneNames %in% c('LARGE','FRG1','FRG1B','PEG10','SPEG')))
```
Adjusted Pvalue:
```{r PVAL}
kable(filter(PVAL, GeneNames %in% c('LARGE','FRG1','FRG1B','PEG10','SPEG')))
```
### Enrichment Analysis
Perform enrichment analysis for differentiation
```{r enrichment.analysis, include=FALSE}
# Write input files to folder
inputFolderName <- paste0(getwd(),'/GOElite_input')
if (file.exists(inputFolderName))
  unlink(inputFolderName, recursive = T, force = T)
dir.create(inputFolderName, showWarnings = TRUE, recursive = FALSE)

# Function to write input files for GO-Elite
writeInputFiles <- function(cont.name, FC, PVAL, inputFolderName){
  input <- FC[(FC[,cont.name]>=1 & PVAL[,cont.name]<=0.05),c('GeneNames',cont.name),drop=F]
  input <- dplyr::mutate(input,SystemCode = 'Sy')
  input <- input[,c(1,3,2)]
  colnames(input)[3] <- 'logFC'
  
  Genes <- cbind(paste(str_replace_all(cont.name,'[^[:alnum:]]','_'),'up',sep='_'),
                 paste(input$GeneNames,collapse=','))
  
  FName <- paste(str_replace_all(cont.name,'[^[:alnum:]]','_'),'up','txt',sep='.')    
  FName <- write.table(input,paste(inputFolderName,FName,sep='/'),sep='\t',row.names = F,col.names = T, quote=F)
  
  input <- FC[(FC[,cont.name]<=-1 & PVAL[,cont.name]<=0.05),c('GeneNames',cont.name),drop=F]
  input <- dplyr::mutate(input,SystemCode = 'Sy')
  input <- input[,c(1,3,2)]
  colnames(input)[3] <- 'logFC'
  
  Genes <- rbind(Genes,cbind(paste(str_replace_all(cont.name,'[^[:alnum:]]','_'),'down',sep='_'),
                             paste(input$GeneNames,collapse=',')))
  
  FName <- paste(str_replace_all(cont.name,'[^[:alnum:]]','_'),'down','txt',sep='.')  
  FName <- write.table(input,paste(inputFolderName,FName,sep='/'),sep='\t',row.names = F,col.names = T, quote=F)
  
  return(Genes)
}
DEXPP.GENES <- lapply(colnames(FC)[-(1)],writeInputFiles, FC, PVAL, inputFolderName)
DEXPP.GENES <- plyr::ldply(DEXPP.GENES) 
colnames(DEXPP.GENES) <- c('Contrast.Names','Gene.Symbols')

# Write denominator file to folder
denominatorFolderName <- paste0(getwd(),'/GOElite_denominator')
if (file.exists(denominatorFolderName))
  unlink(denominatorFolderName, recursive = T, force = T)
dir.create(denominatorFolderName, showWarnings = TRUE, recursive = FALSE)
tmp <- FC[,c('GeneNames'),drop=F] 
tmp <- dplyr::mutate(tmp,SystemCode = 'Sy')
FName <- 'Denominaor.txt'
write.table(tmp,paste(denominatorFolderName,FName,sep='/'),sep='\t',row.names = F,col.names = T, quote=F)

# Create output dir
outputFolderName <- paste0(getwd(),'/GOElite_output')
if (file.exists(outputFolderName))
  unlink(outputFolderName, recursive = T, force = T)
dir.create(outputFolderName, showWarnings = TRUE, recursive = FALSE)

# system(paste('python',
#              '/mnt/data/GO-Elite_v.1.2.5-Py/GO_Elite.py',
#              '--species','Hs',
#              '--version','EnsMart72Plus',
#              '--input', inputFolderName,
#              '--denom',denominatorFolderName,
#              '--output',outputFolderName))
```
### Results
Store files in synapse 
```{r synapse.store, include = FALSE, eval=TRUE}
activityName='Differential Expression Analysis of eXpress aligned data with mixed effects model at each differentiation stages'
  
thisFileName <- 'DiffExpAnal_Express_mixedEffects_atEachDiffState.Rmd'
  
# Github link
thisRepo <- getRepo(repository = "th1vairam/pcbc_c4_analysis", 
                    ref="branch", 
                    refName='diff_exp')

thisFile <- getPermlink(repository = thisRepo,
                        repositoryPath=paste0('code/Rmd/', thisFileName))

CODE <- File('./DiffExpAnal_Express_mixedEffects_atEachDiffState.Rmd',name = 'Differential Expression Analysis Express Mixed Effects At Each DiffState',parentId = parentId)
CODE <- synStore(CODE, used = ALL_USED_IDs,activityName = activityName, executed=thisFile)
      
write.table(FC,file='./DiffnameShort_FoldChange_Express_mixedEffects_atEachDiffState.tsv',sep='\t',row.names=F,quote=F)
FC <- File('./DiffnameShort_FoldChange_Express_mixedEffects_atEachDiffState.tsv',name = 'Differential Expression Analysis Express Mixed Effects At Each DiffState logFC',parentId = parentId)
FC <- synStore(FC, used = ALL_USED_IDs, activityName = activityName, executed=CODE$properties$id)
    
write.table(PVAL,file='./DiffnameShort_Pvalue_Express_mixedEffects_atEachDiffState.tsv',sep='\t',row.names=F,quote=F)
PVAL <- File('./DiffnameShort_Pvalue_Express_mixedEffects_atEachDiffState.tsv',name = 'Differential Expression Analysis Express Mixed Effects At Each DiffState pvalue',parentId = parentId)
PVAL <- synStore(PVAL, used = ALL_USED_IDs, activityName = activityName, executed=CODE$properties$id)
  
write.table(DEXPP.GENES,file='./DiffnameShort_DiffGenes_Express_mixedEffects_atEachDiffState.gct',sep='\t',row.names=F,quote=F)
DEXPP <- File('./DiffnameShort_DiffGenes_Express_mixedEffects_atEachDiffState.gct',name = 'Differential Expression Analysis Express Mixed Effects At Each DiffState GeneList',parentId = parentId)
annotations(DEXPP) <- list(Pvalue='0.05',Foldchage='0')
DEXPP <- synStore(DEXPP, used = ALL_USED_IDs, activityName = activityName, executed=CODE$properties$id)

write.table(SIG.SETS,file='./DiffnameShort_DiffGenes_Express_mixedEffects_atEachDiffState_comparison.tsv',sep='\t',row.names=F,quote=F)
SIG.SETS <- File('./DiffnameShort_DiffGenes_Express_mixedEffects_atEachDiffState_comparison.tsv',name = 'Differential Expression Analysis Express Mixed Effects At Each DiffState Comparison',parentId = parentId)
annotations(SIG.SETS) <- list(Pvalue='0.05',Foldchage='0')
SIG.SETS <- synStore(SIG.SETS, used = ALL_USED_IDs, activityName = activityName, executed=CODE$properties$id)
```
|  *Results*           |  *SynapseID*                |
|  -------             |   ---------                 |
|  Log fold change     |  `r FC$properties$id`       |
|  Adjusted pvalues                                |  `r PVAL$properties$id`     |
|  Differentially expressed genes (gct format)     |  `r DEXPP$properties$id`     |
|  Differentially expressed genes (for comparison) |  `r SIG.SETS$properties$id`  |