---
title: "Differential expression analysis for eXpress aligned data for iPSc vs hESc at each diff state"
author: "Thanneer Perumal"
date: "`r date()`"
output: html_document
---

```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
## It is assumed your working directory is where this file is

# Clear R console screen output
cat("\014")

# Load required libraries
library('synapseClient')
library('limma')
library('edgeR')
library('RColorBrewer')
library('ctv')
library('ggplot2')
library('psych')
library('reshape2')
library('gplots')
library('vcd')
library('erer')
library('fpc')
library('data.table')
library('dplyr')
library('stringr')
library('knitr')

## Requires ggdendro
devtools::source_url("https://raw.githubusercontent.com/chr1swallace/random-functions/master/R/ggplot-heatmap.R")

# source('/mnt/Github/knit2synapse-1/R/knitFile2Synapse.R')
# knitFile2Synapse(file = "./DiffExpAnal_Express_mixedEffects_atEachDiffState.Rmd", owner = 'syn3801315', wikiName = "Differential Expression Analysis Express Mixed Effects iPScVsHEsc",overwrite=F)

synapseLogin()

# source utility files from ../R/lib folder
file.sources = list.files('../R/lib',pattern="*.R")
file.sources = sapply(file.sources,function(x){return(paste('../R/lib',x,sep='/'))})
tmp = sapply(file.sources,source,.GlobalEnv)

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE)
```

#### Download data

```{r setup, include=FALSE}
# Input Parameters
COUNT_ID = 'syn3446250'
METADATA_ID = 'syn3156503'

SYNAPSE_STORE = T
parentId = 'syn3471223'

# Specify factor and continuous covarites pool (adjusted covariates will only be the subset of these covariates)
FactorCovariates = c('Diffname_short','run','lane','index','Cell_Type_of_Origin', 'Cell_Line_of_Origin', 'Tissue_of_Origin', 'Reprogramming_Vector_Type','Reprogramming_Gene_Combination', 'Donor_Life_Stage','Originating_Lab','Gender','Donor_ID','Disease','Race','Ethnicity','Cell_Line_Type','PassageAtHarvest')
ContCovariates = NULL
```
Synapse id of count matrix used for the analysis is `r COUNT_ID` and the synapse id of meta data table used for the analysis is `r METADATA_ID`. 

Factor covariates considered for analysis are `r FactorCovariates`, and continuous covariates considered for the analysis are `r ContCovariates`.

Obtain count matrix and metadata from synapse.
```{r getdata, cache=TRUE, include=FALSE}
# Get count matrix
COUNT_OBJ = synGet(COUNT_ID)
ALL_USED_IDs = COUNT_OBJ$properties$id
COUNT = fread(getFileLocation(COUNT_OBJ), data.table=FALSE)
row.names(COUNT) = COUNT[,1]

# Get metadata
METADATA_OBJ = synTableQuery(paste('SELECT * FROM',METADATA_ID,sep=' '))
ALL_USED_IDs[length(ALL_USED_IDs)+1] = METADATA_OBJ@schema
METADATA = METADATA_OBJ@values
```
Preprocess counts matrix and metadata.
We separate MESO samples into early (5 day) and late (15 and 30 day).
```{r preprocessing, include=FALSE}
# Preprocess metadata
METADATA[METADATA == 'N/A'] = 'Not Applicable'

# Preprocess passage at harvest information
METADATA$PassageAtHarvest[is.na(METADATA$PassageAtHarvest)] = 'Not Available' 

# Assign new values for MESO-5, 15 and 30 samples
METADATA$Diffname_short[METADATA$Diffname_short == 'MESO-5'] = 'MESO_EARLY'
METADATA$Diffname_short[METADATA$Diffname_short == 'MESO-15' | 
                          METADATA$Diffname_short == 'MESO-30'] = 'MESO_LATE'
rownames(METADATA) = METADATA$UID

```
#### Preprocess data
* Remove somatic samples and samples with type NA.
* Remove samples that failed QC and samples classified as exclude.
* Remove samples with abnormal karyotypes.
```{r filtering, echo=TRUE}
#### Pre processing mRNA expression counts and metadata ####
metadata_filtered <- 
  METADATA %>%
  filter(Diffname_short != "") %>%
  filter(UID %in% colnames(COUNT)) %>%
  filter(Cell_Type == "PSC") %>%
  filter(!is.na(Cell_Type)) %>%
  filter(pass_qc) %>%
  filter(!exclude) %>%
  filter(C4_Karyotype_Result != "abnormal" | !is.na(C4_Karyotype_Result))

REMOVED_UID <- setdiff(METADATA$UID, metadata_filtered$UID)
METADATA <- METADATA[metadata_filtered$UID,]
COUNT <- COUNT[, METADATA$UID]
```

The following `r length(REMOVED_UID)` samples were removed:

`r REMOVED_UID` 

### CPM Normalisation
Determine design matrix for normalisation and differential expression analysis. 

Remove genes that have less than 1 cpm counts in at least 50% of each of the individual differentiation stages.
```{r cpmnormalisation}
tmp <- tapply(colnames(COUNT),
              factor(METADATA$Diffname_short),
              function(cols,COUNT){PROCESSED_COUNTS = getGeneFilteredGeneExprMatrix(COUNT[,cols])},
              COUNT)

ugenes <- c()
for (i in unique(METADATA$Diffname_short)) {
  ugenes <- unique(c(ugenes,tmp[[i]]$filteredExprMatrix$genes[,1]))
}

COUNT <- COUNT[ugenes,,drop=F]
PROCESSED_COUNTS = getGeneFilteredGeneExprMatrix(COUNT,MIN_GENE_CPM=0,
                                                 MIN_SAMPLE_PERCENT_WITH_MIN_GENE_CPM=0)
```

```{r covariates.clustering}
# Find inter relation between factor covariates
COVARIATES = METADATA[,c(FactorCovariates,ContCovariates)]

# Convert factor covariates to factors
COVARIATES[,FactorCovariates] = lapply(COVARIATES[,FactorCovariates], factor)
```

#### Normalisation with Cell_Line_Type
```{r cell.line.type.factorisation, cache=TRUE}
getDifferentialExpression <- function(COVARIATES,COUNTS){
  
  ind <- which(sapply(COVARIATES,function(x){length(unique(x))})!=1)
  COVARIATES <- COVARIATES[,ind]
  
  # Get processed counts data
  COUNTS <- COUNTS[,rownames(COVARIATES)]
  
  # Design matrix
  DESIGN = getDesignMatrix(dplyr::select(COVARIATES,Cell_Line_Type), Intercept = F)
  DESIGN$design = DESIGN$design[,linColumnFinder(DESIGN$design)$indepCols]
    
  # Estimate correlation of random effects
  VOOM.ADJUSTED.GENE_EXPRESSION = voom(COUNTS, design=DESIGN$design, plot=F)
  correlation <- duplicateCorrelation(VOOM.ADJUSTED.GENE_EXPRESSION$E, design=DESIGN$design, block=COVARIATES$Donor_ID)
  
  # Re-calculate voom weights with correlation of random effects
  VOOM.ADJUSTED.GENE_EXPRESSION = voom(COUNTS, design=DESIGN$design, plot=F,
                                       block=COVARIATES$Donor_ID, correlation = correlation$cor)
  
  # Fit linear model with cell line type
  FIT = lmFit(VOOM.ADJUSTED.GENE_EXPRESSION$E, design=DESIGN$design, weights = VOOM.ADJUSTED.GENE_EXPRESSION$weights,
              block=COVARIATES$Donor_ID, correlation = correlation$cor)
  
  # Make contrast to check differential expression between cell line types
  CONT.NAMES <- colnames(DESIGN$design)[grep('Cell_Line_Type',colnames(DESIGN$design))]
  CONT.NAMES <- combn(CONT.NAMES,2)
  CONT.NAMES <- apply(CONT.NAMES,2,paste,collapse='-')

  CONT <- makeContrasts(contrasts=CONT.NAMES,levels=colnames(FIT$coefficients))
  colnames(CONT) <- sapply(colnames(CONT),function(x){x <- gsub('Cell_Line_Type','',x);})

  # Refit contrasts
  FIT.CONTRAST <- contrasts.fit(FIT,CONT)
  
  # Estimate moderated t-statistics
  FIT.CONTRAST <- eBayes(FIT.CONTRAST)
  
  # Obtain all the differential expession combinations
  CELL.LINE.TYPE <- list()
  CELL.LINE.TYPE$logFC <- data.frame(row.names = rownames(VOOM.ADJUSTED.GENE_EXPRESSION$E))
  CELL.LINE.TYPE$adj.P.Val <- data.frame(row.names = rownames(VOOM.ADJUSTED.GENE_EXPRESSION$E))
  
  for (i in colnames(CONT)){
    tmp <- topTable(FIT.CONTRAST, coef=i, number=dim(VOOM.ADJUSTED.GENE_EXPRESSION$E)[1])    
    CELL.LINE.TYPE$logFC[,i] <- tmp[rownames(CELL.LINE.TYPE$logFC),'logFC']
    CELL.LINE.TYPE$adj.P.Val[,i] <- tmp[rownames(CELL.LINE.TYPE$adj.P.Val),'adj.P.Val']
  }

  # Get number of significantly differentialy expressed genes with adj.P.Val <= 0.05 and abs(logFC) >= 1
  CELL.LINE.TYPE$SIG.EXP <- CELL.LINE.TYPE$adj.P.Val<=0.05 & abs(CELL.LINE.TYPE$logFC) >= 0   
  CELL.LINE.TYPE$NUM.SIG.EXP <- colSums(CELL.LINE.TYPE$SIG.EXP)
  
  DESIGN <- getDesignMatrix(COVARIATES,Intercept = F)
  DESIGN <- DESIGN$design
  colnames(DESIGN) <- gsub('[[:punct:]]','_',colnames(DESIGN))
  colnames(DESIGN) <- gsub(' ','_',colnames(DESIGN))
  
  if (CELL.LINE.TYPE$NUM.SIG.EXP !=0 )
    CELL.LINE.TYPE$sigCovars <- runPCAandPlotCorrelations(VOOM.ADJUSTED.GENE_EXPRESSION$E[CELL.LINE.TYPE$SIG.EXP,],COVARIATES,paste0('cell_line_type modeled_',unique(COVARIATES$Cell_Line_Type)))
  else
    CELL.LINE.TYPE$sigCovars <- NULL

  return(CELL.LINE.TYPE)
}

# Split covariates based on differentiation stages
NEW.COVARIATES <- split(COVARIATES,COVARIATES$Diffname_short,drop=T)

# Find differnetially expressed genes b/w ESC and iPSC
RESULTS <- list()
for (COV in names(NEW.COVARIATES)){
  RESULTS[[COV]] <- getDifferentialExpression(NEW.COVARIATES[[COV]],PROCESSED_COUNTS$filteredExprMatrix)
}
```

Number of significantly expressed genes at an FDR <= 0.05 
```{r significant.genes}
tmp <- as.data.frame(sapply(RESULTS,function(x){return(x$NUM.SIG.EXP)}))
colnames(tmp) <- 'Num. of sig. expressed genes'
kable(tmp)
```

Correlation between covariates 
```{r significant.covariates, cache=TRUE, fig.height=10, fig.width=15}
sapply(RESULTS,function(x){print(x$sigCovars$PC_res[[2]]$plotData)})
```

Store results to synapse

```{r synapsestore, echo=FALSE, include=FALSE,eval=TRUE}
if (SYNAPSE_STORE){  
  ActivityName <- 'Differential Expression Analysis of eXpress aligned data at each differentiation stages with mixed effect modeling'
  
  # Code
  CODE <- File('./DiffExpAnal_Express_atEachDiffState.Rmd',name = 'Differential Expression Analysis Express Mixed Effects iPScVshESc',parentId = parentId)
  CODE <- synStore(CODE, used = ALL_USED_IDs, activityName = ActivityName, executed = '')  
}
```