---
title: "Differential expression analysis for eXpress aligned data with mixed effects modelling at each diff state"
author: "Thanneer Perumal"
date: "`r date()`"
output: html_document
---
```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
## It is assumed your working directory is where this file is

# Clear R console screen output
cat("\014")  

# Load required libraries
library('synapseClient')
library('RColorBrewer')
library('ggplot2')
library('gplots')
library('limma')
library('edgeR')
library('ctv')
library('psych')
library('reshape2')
library('vcd')
library('erer')
library('fpc')
library('data.table')
library('dplyr')
library('knitr')
library('stringr')

## Needs the dev branch
library(rGithubClient)

## Requires ggdendro
devtools::source_url("https://raw.githubusercontent.com/chr1swallace/random-functions/master/R/ggplot-heatmap.R")

# source('/mnt/Github/knit2synapse-1/R/knitFile2Synapse.R')
# knitFile2Synapse(file = "./DiffExpAnal_Express_mixedEffects_atEachDiffState.Rmd", owner = 'syn4223183', wikiName = "Differential Expression Analysis Express Mixed Effects At Each DiffState",overwrite=F)

synapseLogin()

# source utility files from ../R/lib folder
file.sources = list.files('../R/lib',pattern="*.R")
file.sources = sapply(file.sources,function(x){return(paste('../R/lib',x,sep='/'))})
tmp = sapply(file.sources,source,.GlobalEnv)

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE)
```
### Download data
```{r setup, include=FALSE}
# Input Parameters
COUNT_ID = 'syn3446250'
METADATA_ID = 'syn3156503'

SYNAPSE_STORE = T
parentId = 'syn3471223'

# Specify factor and continuous covarites pool (adjusted covariates will only be the subset of these covariates)
FactorCovariates = c("Diffname_short", "Gender", "Cell_Type_of_Origin_Level2", "Reprogramming_Gene_Combination", "Reprogramming_Vector_Type", "Disease", "run", "Donor_ID")
ContCovariates = NULL
```
Synapse id of count matrix used for the analysis is `r COUNT_ID` and the synapse id of meta data table used for the analysis is `r METADATA_ID`. 

Factor covariates considered for analysis are `r FactorCovariates`, and continuous covariates considered for the analysis are `r ContCovariates`.

Obtain count matrix and metadata from synapse.
```{r getdata, cache=TRUE, include=FALSE}
# Get count matrix
COUNT_OBJ = synGet(COUNT_ID)
ALL_USED_IDs = COUNT_OBJ$properties$id
COUNT = fread(getFileLocation(COUNT_OBJ), data.table=FALSE)
row.names(COUNT) = COUNT[,1]

# Get metadata
METADATA_OBJ = synTableQuery(paste('SELECT * FROM',METADATA_ID,sep=' '))
ALL_USED_IDs[length(ALL_USED_IDs)+1] = METADATA_OBJ@schema
METADATA = METADATA_OBJ@values
```
Preprocess counts matrix and metadata.
```{r preprocessing, include=FALSE}
# Preprocess metadata
METADATA[METADATA == 'N/A'] = NA

# Replace all special characters with blank
myFix <- function(x) str_replace_all(x, '[^[:alnum:]]', '')
METADATA <- METADATA %>%
  dplyr::mutate_each(funs(myFix), -UID, -C4_Cell_Line_ID, -biologicalSampleName) # fix them but don't touch some columns

# Set rownames
rownames(METADATA) = METADATA$UID

# Preprocess passage at harvest information
# METADATA$PassageAtHarvest[is.na(METADATA$PassageAtHarvest)] = 'Not Available' 

# # Assign new values for MESO-5, 15 and 30 samples
# METADATA$Diffname_short[METADATA$Diffname_short == 'MESO-5'] = 'MESO_EARLY'
# METADATA$Diffname_short[METADATA$Diffname_short == 'MESO-15' | 
#                           METADATA$Diffname_short == 'MESO-30'] = 'MESO_LATE'
```
### Preprocess data
* Remove somatic samples and samples with not type.
* Remove samples that failed QC and samples classified as exclude.
* Remove samples with abnormal karyotypes.
```{r filtering, echo=TRUE}
#### Pre processing mRNA expression counts and metadata ####
metadata_filtered <- 
  METADATA %>%
  filter(Diffname_short != "") %>%
  filter(UID %in% colnames(COUNT)) %>%
  filter(Cell_Type == "PSC") %>%  
  filter(pass_qc == "TRUE") %>%
  filter(exclude != "TRUE") %>%
  filter(C4_Karyotype_Result != "abnormal") %>%
  filter(!is.na(C4_Karyotype_Result))

REMOVED_UID <- setdiff(METADATA$UID, metadata_filtered$UID)
METADATA <- METADATA[metadata_filtered$UID,]
COUNT <- COUNT[, METADATA$UID]
```
The following `r length(REMOVED_UID)` samples were removed:

`r REMOVED_UID` 

### CPM Normalisation
Determine design matrix for normalisation and differential expression analysis. 

Remove genes that have less than 1 cpm counts in at least 50% of each of the individual differentiation stages.
```{r cpmnormalisation}
tmp <- tapply(colnames(COUNT),
              factor(METADATA$Diffname_short),
              function(cols,COUNT){PROCESSED_COUNTS = getGeneFilteredGeneExprMatrix(COUNT[,cols])},
              COUNT)

ugenes <- c()
for (i in unique(METADATA$Diffname_short)) {
  ugenes <- unique(c(ugenes,tmp[[i]]$filteredExprMatrix$genes[,1]))
}

COUNT <- COUNT[ugenes,,drop=F]
PROCESSED_COUNTS = getGeneFilteredGeneExprMatrix(COUNT,MIN_GENE_CPM=0,
                                                 MIN_SAMPLE_PERCENT_WITH_MIN_GENE_CPM=0)
```
```{r covariates.clustering}
# Find inter relation between factor covariates
COVARIATES = METADATA[,c(FactorCovariates,ContCovariates)]

# Convert factor covariates to factors
COVARIATES[,FactorCovariates] = lapply(COVARIATES[,FactorCovariates], factor)
```
### Differential Expression Analysis At Each Differentiation Stage
Obtain differentially expressed genes with the following covariates (as obtained from covariates analysis) in the model: 
* Fixed Effects: Diffname short, Gender, Cell Type of Origin Level2, Reprogramming Gene Combination, Reprogramming Vector Type, Disease, run
* Random Effects: Donor ID

Function to perform differential expression analysis at each of the differentiation stages
```{r differential.expression, cache=TRUE}
getDifferentialExpression <- function(COVARIATES,COUNTS, cols, varOfInterest){
  Name <- paste0(unique(COVARIATES$Diffname_short),' differentiation stage')
  print(Name)
  
  # Get covariates matrix
  ind <- which(sapply(COVARIATES,function(x){length(unique(x))})!=1)
  COVARIATES <- droplevels(COVARIATES[,ind])
    
  # Get processed counts data
  COUNTS <- COUNTS[,rownames(COVARIATES)]
    
  # Post adjusted design matrix
  DESIGN = getDesignMatrix(COVARIATES[,intersect(cols,colnames(COVARIATES))], Intercept = F)
  DESIGN = DESIGN$design[,linColumnFinder(DESIGN$design)$indepCols]
  
  # Estimate correlation of random effects
  EXPR = voom(COUNTS, design=DESIGN, plot=F)
  
  # Calculate correlation between donors
  correlation <- duplicateCorrelation(EXPR, block=COVARIATES$Donor_ID)
  
  # Fit linear model using mixed effects design
  FIT = lmFit(EXPR, block=COVARIATES$Donor_ID, correlation = correlation$cor)
  
  # Make contrast to check differential expression between different differentiation stages
  CONT.NAMES <- colnames(DESIGN)[grep(varOfInterest,colnames(DESIGN))]
  CONT.NAMES <- combn(CONT.NAMES,2)
  CONT.NAMES <- apply(CONT.NAMES,2,paste,collapse='-')
  
  CONT <- makeContrasts(contrasts=CONT.NAMES,levels=colnames(FIT$coefficients))
  colnames(CONT) <- sapply(colnames(CONT),function(x){x <- gsub(varOfInterest,'',x);
                                                      x <- gsub('-','_',x)})
  
  # Refit contrasts
  FIT.CONTRAST <- contrasts.fit(FIT,CONT)
  
  # Estimate moderated t-statistics
  FIT.CONTRAST <- eBayes(FIT.CONTRAST)
  
  # Obtain all the differential expession combinations
  DEXP <- list()
  DEXP$logFC <- data.frame(row.names = rownames(COUNT))
  DEXP$adj.P.Val <- data.frame(row.names = rownames(COUNT))
  DEXP$SIG.SETS <- data.frame()
  
  for (i in colnames(CONT)){
    tmp <- topTable(FIT.CONTRAST, coef=i, number=dim(COUNT)[1])    
    DEXP$logFC[,i] <- tmp[rownames(DEXP$logFC),'logFC']
    DEXP$adj.P.Val[,i] <- tmp[rownames(DEXP$adj.P.Val),'adj.P.Val'] 
    
    DEXP$SIG.SETS <- rbind(DEXP$SIG.SETS,
                           getUpDownGenes(DEXP$adj.P.Val[,i], DEXP$logFC[,i], rownames(DEXP$logFC), i, FC_CUTOFF = 0.01))
  }
  
  # Get number of significantly differentialy expressed genes with adj.P.Val <= 0.05 and logFC >= 0
  DEXP$SIG.EXP.POS <- DEXP$adj.P.Val<=0.05 & DEXP$logFC >= 0   
  DEXP$NUM.SIG.EXP.POS <- colSums(DEXP$SIG.EXP.POS)
  
  # Get number of significantly differentialy expressed genes with adj.P.Val <= 0.05 and logFC <= 0
  DEXP$SIG.EXP.NEG <- DEXP$adj.P.Val<=0.05 & DEXP$logFC <= 0    
  DEXP$NUM.SIG.EXP.NEG <- colSums(DEXP$SIG.EXP.NEG)
  
  return(list(FIT = FIT,FIT.CONTRAST = FIT.CONTRAST, DEXP = DEXP))
}
```
### Cell Type of Origin Level2
Perform differential expression except MESO_LATE due to unavailability of enough Donor replications (at least 2-3 donors need triplicate samples)
Modeling with the following coefficients as fixed effects Diffname short, Gender, Cell Type of Origin Level2, Reprogramming Gene Combination, Reprogramming Vector Type, Disease, and run, and Donor_ID as random effects 
Number of significantly expressed genes at an FDR <= 0.05
```{r cell.type.origin, cache=TRUE}
COV <- split(COVARIATES,COVARIATES$Diffname_short)
CELL.TYPE.ORIG <- lapply(COV[c(2,7)],getDifferentialExpression,COUNT, c("Cell_Type_of_Origin_Level2", "Diffname_short", "Gender", "Reprogramming_Gene_Combination", "Reprogramming_Vector_Type", "Disease", "run"), varOfInterest = "Cell_Type_of_Origin_Level2")

tmp <- lapply(DIFF.EXP,function(x){return(data.frame(Contrast = names(x$DEXP$NUM.SIG.EXP.NEG), POS = x$DEXP$NUM.SIG.EXP.POS, NEG = x$DEXP$NUM.SIG.EXP.NEG))})
tmp <- join_all(tmp,by='Contrast',match='all')
colnames(tmp)[2:dim(tmp)[2]] = rbind(paste(names(DIFF.EXP),'UP',sep='.'),paste(names(DIFF.EXP),'DOWN',sep='.'))[1:(length(DIFF.EXP)*2)]
kable(tmp)
```

Get pvalue and fold changes of some important genes
```{r print.imp.genes}
FC <- join_all(lapply(names(DIFF.EXP), 
                      function(x,DIFF.EXP){ 
                        tmp <- DIFF.EXP[[x]]$DEXP$logFC;
                        colnames(tmp) <- paste(x,colnames(tmp),sep='.');
                        tmp <- rownameToFirstColumn(tmp,'GeneNames');
                        return(tmp)
                      }, DIFF.EXP),
               by = 'GeneNames',
               match = 'all')

PVAL <- join_all(lapply(names(DIFF.EXP),
                        function(x,DIFF.EXP){ 
                          tmp <- DIFF.EXP[[x]]$DEXP$adj.P.Val;
                          colnames(tmp) <- paste(x,colnames(tmp),sep='.');
                          tmp <- rownameToFirstColumn(tmp,'GeneNames');
                          return(tmp)
                        }, DIFF.EXP),
                 by = 'GeneNames',
                 match = 'all')

print('Fold change:')
kable(filter(FC, GeneNames %in% c('LARGE','FRG1','FRG1B','PEG10','SPEG')))
print('Adjusted Pvalue:')
kable(filter(PVAL, GeneNames %in% c('LARGE','FRG1','FRG1B','PEG10','SPEG')))
```

Perform enrichment analysis for differentiation
```{r enrichment.analysis, include=FALSE}
# Write input files to folder
inputFolderName <- paste0(getwd(),'/GOElite_AtEachDiff_input')
if (file.exists(inputFolderName))
  unlink(inputFolderName, recursive = T, force = T)
dir.create(inputFolderName, showWarnings = TRUE, recursive = FALSE)

# Function to write input files for GO-Elite
writeInputFiles <- function(cont.name, FC, PVAL, inputFolderName){
  input <- FC[(FC[,cont.name]>=0 & PVAL[,cont.name]<=0.05),c('GeneNames',cont.name),drop=F]
  input <- dplyr::mutate(input,SystemCode = 'Sy')
  input <- input[,c(1,3,2)]
  colnames(input)[3] <- 'logFC'
  
  Genes <- cbind(paste(str_replace_all(cont.name,'[^[:alnum:]]','_'),'up',sep='_'),
                 paste(input$GeneNames,collapse=','))
  
  FName <- paste(str_replace_all(cont.name,'[^[:alnum:]]','_'),'up','txt',sep='.')    
  FName <- write.table(input,paste(inputFolderName,FName,sep='/'),sep='\t',row.names = F,col.names = T, quote=F)
  
  input <- FC[(FC[,cont.name]<0 & PVAL[,cont.name]<=0.05),c('GeneNames',cont.name),drop=F]
  input <- dplyr::mutate(input,SystemCode = 'Sy')
  input <- input[,c(1,3,2)]
  colnames(input)[3] <- 'logFC'
  
  Genes <- rbind(Genes,cbind(paste(str_replace_all(cont.name,'[^[:alnum:]]','_'),'down',sep='_'),
                             paste(input$GeneNames,collapse=',')))
  
  FName <- paste(str_replace_all(cont.name,'[^[:alnum:]]','_'),'down','txt',sep='.')  
  FName <- write.table(input,paste(inputFolderName,FName,sep='/'),sep='\t',row.names = F,col.names = T, quote=F)
  
  return(Genes)
}
DEXPP.GENES <- lapply(colnames(FC)[-(1)],writeInputFiles, FC, PVAL, inputFolderName)
DEXPP.GENES <- plyr::ldply(DEXPP.GENES) 
colnames(DEXPP.GENES) <- c('Contrast.Names','Gene.Symbols')

# Write denominator file to folder
denominatorFolderName <- paste0(getwd(),'/GOElite_AtEachDiff_denominator')
if (file.exists(denominatorFolderName))
  unlink(denominatorFolderName, recursive = T, force = T)
dir.create(denominatorFolderName, showWarnings = TRUE, recursive = FALSE)
tmp <- FC[,c('GeneNames'),drop=F] 
tmp <- dplyr::mutate(tmp,SystemCode = 'Sy')
FName <- 'Denominaor.txt'
write.table(tmp,paste(denominatorFolderName,FName,sep='/'),sep='\t',row.names = F,col.names = T, quote=F)

# Create output dir
outputFolderName <- paste0(getwd(),'/GOElite_AtEachDiff_output')
if (file.exists(outputFolderName))
  unlink(outputFolderName, recursive = T, force = T)
dir.create(outputFolderName, showWarnings = TRUE, recursive = FALSE)

# system(paste('python',
#              '/mnt/data/GO-Elite_v.1.2.5-Py/GO_Elite.py',
#              '--species','Hs',
#              '--version','EnsMart72Plus',
#              '--input', inputFolderName,
#              '--denom',denominatorFolderName,
#              '--output',outputFolderName))
```


Store files in synapse
 
```{r synapse.store, include = FALSE, eval=TRUE}
if(SYNAPSE_STORE){
  activityName='Differential Expression Analysis of eXpress aligned data with mixed effects model at each differentiation stages'
  
  thisFileName <- 'DiffExpAnal_Express_mixedEffects_atEachDiffState.Rmd'
  
  # Github link
  thisRepo <- getRepo(repository = "th1vairam/pcbc_c4_analysis", 
                      ref="branch", 
                      refName='diff_exp')

  thisFile <- getPermlink(repository = thisRepo,
                          repositoryPath=paste0('code/Rmd/', thisFileName))
  
  
  CODE <- File('./DiffExpAnal_Express_mixedEffects_atEachDiffState.Rmd',name = 'Differential Expression Analysis Express Mixed Effects At Each DiffState',parentId = parentId)
  CODE <- synStore(CODE, used = ALL_USED_IDs,activityName = activityName, executed=thisFile)
      
  write.table(FC,file='./DiffnameShort_FoldChange_Express_mixedEffects_atEachDiffState.tsv',sep='\t',row.names=F,quote=F)
  FC <- File('./DiffnameShort_FoldChange_Express_mixedEffects_atEachDiffState.tsv',name = 'Differential Expression Analysis Express Mixed Effects At Each DiffState logFC',parentId = parentId)
  FC <- synStore(FC, used = ALL_USED_IDs, activityName = activityName, executed=CODE$properties$id)
    
  write.table(PVAL,file='./DiffnameShort_Pvalue_Express_mixedEffects_atEachDiffState.tsv',sep='\t',row.names=F,quote=F)
  PVAL <- File('./DiffnameShort_Pvalue_Express_mixedEffects_atEachDiffState.tsv',name = 'Differential Expression Analysis Express Mixed Effects At Each DiffState pvalue',parentId = parentId)
  PVAL <- synStore(PVAL, used = ALL_USED_IDs, activityName = activityName, executed=CODE$properties$id)
  
  write.table(DEXPP.GENES,file='./DiffnameShort_DiffGenes_Express_mixedEffects_atEachDiffState.tsv',sep='\t',row.names=F,quote=F)
  DEXPP <- File('./DiffnameShort_DiffGenes_Express_mixedEffects_atEachDiffState.tsv',name = 'Differential Expression Analysis Express Mixed Effects At Each DiffState genelist',parentId = parentId)
  annotations(DEXPP) <- list(Pvalue='0.05',Foldchage='0')
  DEXPP <- synStore(DEXPP, used = ALL_USED_IDs, activityName = activityName, executed=CODE$properties$id)
}
```
|  *Results*           |  *SynapseID*                |
|  -------             |   ---------                 |
|  Log fold change     |  `r FC$properties$id`       |
|  Adjusted pvalues    |  `r PVAL$properties$id`     |
|  Differentially expressed genes  |  `r DEXPP$properties$id`  |