---
title: "Enrichment analysis of differential expression data (mRNA, methylation and miRNA) with Enrichr gene sets"
author: "Thanneer Perumal"
date: '`r date()`'
output: html_document
---
```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
## It is assumed your working directory is where this file is

# Clear R console screen output
cat("\014")

# Load required libraries
library(data.table)
library(dplyr)
library(tools)
library(synapseClient)

## Needs the dev branch
library(rGithubClient)

## Requires ggdendro
devtools::source_url("https://raw.githubusercontent.com/chr1swallace/random-functions/master/R/ggplot-heatmap.R")
# knitToFolderEntity(file = "./DiffExpAnal_methylation_mixedEffects_EachDiffState.Rmd", 
#                    parentId = 'syn4231339',
#                    entityName = 'Differential Expression Analysis Methylation Mixed Effects Model EachDiffState', 
#                    overwrite=F)

synapseLogin()

# source utility files from ../R/lib folder
file.sources = list.files('../R/lib',pattern="*.R")
file.sources = sapply(file.sources,function(x){return(paste('../R/lib',x,sep='/'))})
tmp = sapply(file.sources,source,.GlobalEnv)

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE)
```
### Download Gene Lists
```{r get.genesets, include=FALSE}
# Input Parameters
GENE.SETS_ID = 'syn4867851'
ALL_USED_IDs = GENE.SETS_ID

load(synGet(GENE.SETS_ID)@filePath)
```
`r length(GeneSets)` category of genesets were downloaded from synapse
### Filter Gene Sets for mRNA
```{r filter.genesets.mRNA, include=FALSE}
logCPM_ID = 'syn4483934'
ALL_USED_IDs = c(ALL_USED_IDs, logCPM_ID)
logCPM = data.table::fread(synGet(logCPM_ID)@filePath, data.table=F)
genesInBackground = logCPM$GeneName

GeneSets.mRNA = filterGeneSets(GeneSets, genesInBackground)

# Choose gene sets for testing
gsets = c("Achilles_fitness_decrease", "Achilles_fitness_increase", "Allen_Brain_Atlas_down", "Allen_Brain_Atlas_up",
          "BioCarta", "CMAP_down", "CMAP_up", "Cancer_Cell_Line_Encyclopedia", "ChEA", "Cross_Species_Phenotype",
          "Disease_Signatures_from_GEO_down", "Disease_Signatures_from_GEO_up", "Drug_Perturbations_from_GEO",
          "ENCODE_Histone_Modifications_2013", "ESCAPE", "GO_Biological_Process", "GO_Cellular_Component", "GO_Molecular_Function",
          "GeneSigDB", "Genome_Browser_PWMs.1", "HMDB_Metabolites", "HomoloGene", "Human_Gene_Atlas", "KEGG_2015",
          "MGI_Mammalian_Phenotype", "MGI_Mammalian_Phenotype_Level_3", "MGI_Mammalian_Phenotype_Level_4", "MSigDB_Computational",
          "MSigDB_Oncogenic_Signatures", "Mouse_Gene_Atlas", "NCI-60_Cancer_Cell_Lines", "NCI-Nature", 
          "NURSA_Human_Endogenous_Complexome", "OMIM_Disease", "OMIM_Expanded", "PPI_Hub_Proteins", "Pfam_InterPro_Domains",
          "Phosphatase_Substrates_from_DEPOD", "Reactome", "SILAC_Phosphoproteomics", "TF-LOF_Expression_from_GEO", 
          "TargetScan_microRNA", "Tissue_Protein_Expression_from_Human_Proteome_Map", "Tissue_Protein_Expression_from_ProteomicsDB",
          "Transcription_Factor_PPIs", "Virus_Perturbations_from_GEO_down", "Virus_Perturbations_from_GEO_up", "WikiPathways_2015")

GeneSets.mRNA = GeneSets.mRNA[gsets]
```
### Perform mRNA enrichment analysis
```{r mRNA.enrichment}
# Get differential expression (ALL)
DEXP.ALL_ID = 'syn4484232'
ALL_USED_IDs = c(ALL_USED_IDs, DEXP.ALL_ID)
DEXP.ALL = data.table::fread(synGet(DEXP.ALL_ID)@filePath, data.table=F)

# Filter comparisons
Comparisons.ALL = split(DEXP.ALL, factor(DEXP.ALL$Comparison))
Comparisons.ALL = lapply(Comparisons.ALL, function(x){return(unique(x$GeneSymbol[abs(x$logFC) >= log2(1.5) & x$adj.P.value<=0.05]))})

# Get differential expression (EACH)
DEXP.EACH_ID = 'syn4485993'
ALL_USED_IDs = c(ALL_USED_IDs, DEXP.EACH_ID)
DEXP.EACH = data.table::fread(synGet(DEXP.EACH_ID)@filePath, data.table=F)

# Filter comparisons
Comparisons.EACH = split(DEXP.EACH, factor(DEXP.EACH$Comparison))
Comparisons.EACH = lapply(Comparisons.EACH, function(x){return(unique(x$GeneSymbol[abs(x$logFC) >= log2(1.2) & x$adj.P.value<=0.05]))})

Comparisons = c(Comparisons.ALL, Comparisons.EACH)

# Perform enrichment analysis
enrichResults.mRNA = list()
for (name in names(Comparisons)){
  comp = Comparisons[[name]]
  tmp = lapply(GeneSets.mRNA,
               function(x, comp, genesInBackground){
                 tmp = as.data.frame(t(sapply(x, fisherEnrichment, comp, genesInBackground)))
                 tmp$pval = p.adjust(tmp$pval,'fdr')
                 tmp = rownameToFirstColumn(tmp,'GeneSetName')
                 return(tmp)
                 },
               comp, genesInBackground)
  
  for (name1 in names(tmp))
    tmp[[name1]]$category = name1
  
  enrichResults.mRNA[[name]] = rbindlist(tmp)
  writeLines(paste0('Completed ',name))  
}

# Write results to file
for(name in names(enrichResults.mRNA))
  enrichResults.mRNA[[name]]$ComparisonName = name
enrichmentResults = rbindlist(enrichResults.mRNA)
enrichmentResults$ngenes = unlist(enrichmentResults$ngenes)
enrichmentResults$noverlap = unlist(enrichmentResults$noverlap)
enrichmentResults = enrichmentResults[enrichmentResults$pval<=0.1,]
write.table(enrichmentResults, file = 'enrichmentResults_mRNA.tsv', sep='\t', row.names=F, quote=F)

thisFileName <- 'EnrichmentAnalysis_Enrichr.Rmd'
  
# Github link
thisRepo <- getRepo(repository = "th1vairam/pcbc_c4_analysis", 
                    ref="branch", 
                    refName='enrich')

thisFile <- getPermlink(repository = thisRepo,
                        repositoryPath=paste0('code/Rmd/', thisFileName))

# Write to synapse
OBJ <- File('enrichmentResults_mRNA.tsv',
            name = 'Enrichment Results mRNA',
            parentId = 'syn4483762')
OBJ <- synStore(OBJ, 
                used = ALL_USED_IDs, 
                activityName = 'Enrichment Analysis',
                activityDescription = 'Enrichment analysis based on enrichr genesets',
                executed = thisFile)
```
### Filter Gene Sets for methylation
```{r filter.genesets.methyl, include=FALSE}
ALL_USED_IDs = GENE.SETS_FOLDER
BETA_ID = 'syn4487642'
ALL_USED_IDs = c(ALL_USED_IDs, BETA_ID)
BETA = data.table::fread(synGet(BETA_ID)@filePath, data.table=F)
genesInBackground = unique(get450KProbeMapping(BETA$methProbeID)$Annotation$nearestTx)

GeneSets.methyl = filterGeneSets(GeneSets, genesInBackground)

# Choose gene sets for testing
GeneSets.methyl = GeneSets.methyl[gsets]
```
### Perform methylation enrichment analysis
```{r methyl.enrichment}
# Get differential methylation (ALL)
DMETHYL.ALL_ID = 'syn4527629'
ALL_USED_IDs = c(ALL_USED_IDs, DMETHYL.ALL_ID)
DMETHYL.ALL = data.table::fread(synGet(DMETHYL.ALL_ID)@filePath, data.table=F)

# Filter comparisons
Comparisons.ALL = split(DMETHYL.ALL, factor(DMETHYL.ALL$Comparison))
Comparisons.ALL = lapply(Comparisons.ALL, function(x){return(unique(x$nearestTx[abs(x$logFC) >= log2(1.3) & x$adj.P.value<=0.05]))})

# Get differential expression (EACH)
DMETHYL.EACH_ID = 'syn4598861'
ALL_USED_IDs = c(ALL_USED_IDs, DMETHYL.EACH_ID)
DMETHYL.EACH = data.table::fread(synGet(DMETHYL.EACH_ID)@filePath, data.table=F)

# Filter comparisons
Comparisons.EACH = split(DMETHYL.EACH, factor(DMETHYL.EACH$Comparison))
Comparisons.EACH = lapply(Comparisons.EACH, function(x){return(unique(x$nearestTx[abs(x$logFC) >= log2(1.2) & x$adj.P.value<=0.05]))})

Comparisons = c(Comparisons.ALL, Comparisons.EACH)

# Perform enrichment analysis
enrichResults.methyl = list()
for (name in names(Comparisons)){
  comp = Comparisons[[name]]
  tmp = lapply(GeneSets.methyl,
               function(x, comp, genesInBackground){
                 tmp = as.data.frame(t(sapply(x, fisherEnrichment, comp, genesInBackground)))
                 tmp$pval = p.adjust(tmp$pval,'fdr')
                 tmp = rownameToFirstColumn(tmp,'GeneSetName')
                 return(tmp)
                 },
               comp, genesInBackground)
  
  for (name1 in names(tmp))
    tmp[[name1]]$category = name1
  
  enrichResults.methyl[[name]] = rbindlist(tmp)
  writeLines(paste0('Completed ',name))  
}

# Write results to file
for(name in names(enrichResults.methyl))
  enrichResults.methyl[[name]]$ComparisonName = name
enrichmentResults = rbindlist(enrichResults.methyl)
enrichmentResults$ngenes = unlist(enrichmentResults$ngenes)
enrichmentResults$noverlap = unlist(enrichmentResults$noverlap)
enrichmentResults = enrichmentResults[enrichmentResults$pval<=0.1,]
write.table(enrichmentResults, file = 'enrichmentResults_methylation.tsv', sep='\t', row.names=F, quote=F)

thisFileName <- 'EnrichmentAnalysis_Enrichr.Rmd'
  
# Github link
thisRepo <- getRepo(repository = "th1vairam/pcbc_c4_analysis", 
                    ref="branch", 
                    refName='enrich')

thisFile <- getPermlink(repository = thisRepo,
                        repositoryPath=paste0('code/Rmd/', thisFileName))

# Write to synapse
OBJ <- File('enrichmentResults_methylation.tsv',
            name = 'Enrichment Results methylation',
            parentId = 'syn4489211')
OBJ <- synStore(OBJ, 
                used = ALL_USED_IDs, 
                activityName = 'Enrichment Analysis',
                activityDescription = 'Enrichment analysis based on enrichr genesets',
                executed = thisFile)
```