---
title: "Methylation Analysis of minfi data with mixed effect modeling"
author: "Thanneer Perumal"
date: "`r date()`"
output: html_document
---
  
```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
## It is assumed your working directory is where this file is

# Clear R console screen output
cat("\014")  

# Load required libraries
library('synapseClient')
library('RColorBrewer')
library('ggplot2')
library('gplots')
library('limma')
library('edgeR')
library('ctv')
library('psych')
library('reshape2')
library('vcd')
library('erer')
library('fpc')
library('data.table')
library('dplyr')
library('knitr')
library('stringr')
library('minfi')

## Requires ggdendro
devtools::source_url("https://raw.githubusercontent.com/chr1swallace/random-functions/master/R/ggplot-heatmap.R")

# source('/mnt/Github/knit2synapse-1/R/knitFile2Synapse.R')
# knitFile2Synapse(file = "./CovariateAnalysis_Express_mixedEffects.Rmd", name = 'Covariate Analysis Express Mixed Effects', owner = 'syn3943944', overwrite=F)

synapseLogin()

# source utility files from ../R/lib folder
file.sources = list.files('../R/lib',pattern="*.R")
file.sources = sapply(file.sources,function(x){return(paste('../R/lib',x,sep='/'))})
tmp = sapply(file.sources,source,.GlobalEnv)

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE)
```

### Download data

```{r setup, include=FALSE}
# Input Parameters
EXP_ID = 'syn2233188'
METADATA_ID = 'syn3156828'

SYNAPSE_STORE = T
parentId = 'syn4231339'

# Specify factor and continuous covarites pool (adjusted covariates will only be the subset of these covariates)
FactorCovariates = c("Diffname_short","BeadChip","Row","Column","PassageAtThaw","PassageAtDNAHarvest","Cell_Line_Type","Cell_Type_of_Origin","Cell_Line_of_Origin","Tissue_of_Origin","Reprogramming_Vector_Type","Reprogramming_Gene_Combination","Culture_Conditions","Other_Conditions_During_Reprogramming","Donor_Life_Stage","Race","Ethnicity","Gender","Disease","Donor_Phenotype","Originating_Lab","Donor_ID")
ContCovariates = NULL
```
Synapse id of count matrix used for the analysis is `r EXP_ID` and the synapse id of meta data table used for the analysis is `r METADATA_ID`. 

Factor covariates considered for analysis are `r FactorCovariates`, and continuous covariates considered for the analysis are `r ContCovariates`.

Obtain count matrix and metadata from synapse.
```{r getdata, cache=TRUE, include=FALSE}
# Get count matrix
EXP_OBJ = synGet(EXP_ID)
ALL_USED_IDs = EXP_OBJ$properties$id
EXP = fread(getFileLocation(EXP_OBJ), data.table = F)
colnames(EXP) = c('ProbeIds',colnames(read.table(getFileLocation(EXP_OBJ),nrows=1,header=T,check.names = F)))
row.names(EXP) = EXP[,1]
EXP = dplyr::select(EXP,-(ProbeIds))

# Get metadata
METADATA_OBJ = synTableQuery(paste('SELECT * FROM',METADATA_ID,sep=' '))
ALL_USED_IDs[length(ALL_USED_IDs)+1] = METADATA_OBJ@schema
METADATA = METADATA_OBJ@values
```

Preprocess counts matrix and metadata.
We separate MESO samples into early (5 day) and late (15 and 30 day).
```{r preprocessing, include=FALSE}
# Preprocess metadata
METADATA[METADATA == 'N/A'] = 'Not Applicable'

# Preprocess passage at harvest information
METADATA$PassageAtDNAHarvest [is.na(METADATA$PassageAtDNAHarvest)] = 'Not Available' 

# Preprocess passage at Thaw information
METADATA$PassageAtThaw [is.na(METADATA$PassageAtThaw)] = 'Not Available' 

# Assign new values for MESO-5, 15 and 30 samples
METADATA$Diffname_short[METADATA$Diffname_short == 'MESO-5'] = 'MESO_EARLY'
METADATA$Diffname_short[METADATA$Diffname_short == 'MESO-15' | 
                          METADATA$Diffname_short == 'MESO-30'] = 'MESO_LATE'
rownames(METADATA) = METADATA$UID

ind <- match(colnames(EXP),METADATA$UID)
METADATA <- METADATA[ind,]
```
Distribution of diff name short
|           | summary.factor.METADATA.Diffname_short..|
|:----------|----------------------------------------:|
|UID="H9P50"|                                        1|
|DE         |                                       11|
|EB         |                                       22|
|ECTO       |                                       11|
|MESO_EARLY |                                       11|
|SC         |                                       43|

### Preprocess data
* Remove samples with no Diffname short
* Remove somatic samples and samples with no Cell Type
* Remove samples with abnornal or no C4 Karyotype Result
```{r filtering, echo=TRUE}
#### Pre processing mRNA expression counts and metadata ####
metadata_filtered <- 
  METADATA %>%
  filter(Diffname_short != "") %>%
  filter(UID %in% colnames(EXP)) %>%
  filter(Cell_Type == "PSC") %>%
  filter(C4_Karyotype_Result != "abnormal" & C4_Karyotype_Result != '')

REMOVED_UID <- setdiff(METADATA$UID, metadata_filtered$UID)
METADATA <- METADATA[metadata_filtered$UID,]
EXP <- EXP[, METADATA$UID]
```

The following `r length(REMOVED_UID)` samples were removed:

`r REMOVED_UID` 

### Covariate clustering
Determine relationship between covariates. 

```{r covariates.clustering}
# Find inter relation between factor covariates
COVARIATES = METADATA[,c(FactorCovariates,ContCovariates)]

# Convert factor covariates to factors
COVARIATES[,FactorCovariates] = lapply(COVARIATES[,FactorCovariates], factor)
```

Covariate correlation

```{r covariates.correlation, fig.width=10, fig.height=10}
COVARIATES.CORRELATION = getAssociationStatistics(COVARIATES, PVAL = 0.1)
ggheatmap.show(COVARIATES.CORRELATION$plot, col.width=0.3, row.width=0.15)
```

### Probe filtering
Remove probes that have less than 0.2 beta and greater than 0.8 beta in at least 30% of the samples.
```{r gene.filtering, cache=TRUE}
tmp <- EXP>=0.2 & EXP<=0.8
EXP <- EXP[rowSums(tmp)/dim(EXP)[2] >= 0.3,]

```
`r dim(EXP)[1]` number of probes are considered for analysis

Clustering of initial normalised data (with NULL design)

```{r decompse.normalise.data, cache=TRUE, fig.height=10, fig.width=10}
# Find principal components of expression to plot
PC <- prcomp(EXP)

# Determine number of clusters automatically using pamk
pam.cluster <- pamk(t(EXP), krange=2:10)

# Plot first 2 PCs
plotdata <- data.frame(UID=rownames(PC$rotation), 
                       PC1=PC$rotation[,1], 
                       PC2=PC$rotation[,2])

plotdata <- merge(plotdata, METADATA, by="UID")
plotdata <- mutate(plotdata, labels = '');#ifelse(Diffname_short == '',as.character(UID),'')) 

p <- ggplot(plotdata, aes(x=PC1, y=PC2))
p <- p + geom_point(aes(color = Cell_Line_Type, shape=Diffname_short, size=Gender))
p <- p + theme_bw() + scale_size_manual(values = c(5, 2)) + theme(legend.position="top") 
p <- p + geom_text(aes(label= labels), size=4, hjust=0)
p
```

### Significant Covariates
Correlation between pca of unadjusted expression and covariates is used to find significant covariates
```{r preAdjusted.covariates, cache=TRUE}
# Find correlation between PC's of gene expression with covariates
DESIGN = getDesignMatrix(COVARIATES, Intercept = F)
preAdjustedSigCovars = runPCAandPlotCorrelations(EXP, DESIGN$design,'NULL design', isKeyPlot=TRUE)

# Find significant covariates
adjustCovars = designMatVarsToCovars(getCovariatesMap(DM),preAdjustedSigCovars$significantCovars)
```

Significant covariates to adjust at FDR 0.1 are `r adjustCovars`

```{r preAdjustedSigCovars.NULL, fig.width=15, fig.height=8}
preAdjustedSigCovars[["PC_res"]][[1]]$plotData
```

```{r preAdjustedSigCovars.NULL.ALL, fig.width=15, fig.height=8}
preAdjustedSigCovars[["PC_res"]][[2]]$plotData
```

### Normalisation (iterative)
Since many covariates are confounded, re-normalising COVARIATES with an iterative design matrix. Here Donor_ID is chosen as random effect

```{r iterative.adjusted.voom.normalisation, eval=FALSE, include=FALSE}
postAdjustCovars = list()
postAdjustCovars$random = 'Donor_ID'
 
covariatesEffects = designMat2CovEffects(getCovariatesMap(DM),preAdjustedSigCovars$Effects.significantCovars)
covariatesEffects = covariatesEffects[setdiff(names(covariatesEffects),postAdjustCovars$random)]

postAdjustCovars$fixed = names(which.max(covariatesEffects))
residualSigCovars = preAdjustedSigCovars

loopCount = 0 
while(length(residualSigCovars$significantCovars)!=0 && loopCount <= 100){
  print(paste('Using following covariates in the model',
              paste(paste(postAdjustCovars$fixed,collapse=','),'as fixed effects and', paste(postAdjustCovars$random,collapse=','),'as random effects'),
              sep=':'))
  
  # Post adjusted design matrix
  DM1 = getDesignMatrix(COVARIATES[,postAdjustCovars$fixed,drop=F],Intercept = F)
  DM1$design = DM1$design[,linColumnFinder(DM1$design)$indepCols]
    
  # Estimate correlation of random effects
  VOOM.ADJUSTED.GENE_EXPRESSION = voom(PROCESSED_COUNTS$filteredExprMatrix, design=DM1$design, plot=F)
  correlation <- duplicateCorrelation(VOOM.ADJUSTED.GENE_EXPRESSION$E, design=DM1$design, block=COVARIATES$Donor_ID)
  
  # Re-calculate voom weights with correlation of random effects
  VOOM.ADJUSTED.GENE_EXPRESSION = voom(PROCESSED_COUNTS$filteredExprMatrix, design=DM1$design, plot=F,
                                       block=COVARIATES$Donor_ID, correlation = correlation$cor)
  
  tmp = lmFit(VOOM.ADJUSTED.GENE_EXPRESSION$E, design=DM1$design, weights = VOOM.ADJUSTED.GENE_EXPRESSION$weights,
                                        block=COVARIATES$Donor_ID, correlation = correlation$cor)
  
  # Residuals after normalisation
  RESIDUAL.GENE_EXPRESSION = residuals.MArrayLM(tmp,VOOM.ADJUSTED.GENE_EXPRESSION$E)
  
  # Residual covariates to choose from
  residCovars <- setdiff(c(FactorCovariates,ContCovariates), postAdjustCovars$fixed)
  
  # Find PC of residual gene expression and significant covariates that are highly correlated with PCs
  residualSigCovars = runPCAandPlotCorrelations(RESIDUAL.GENE_EXPRESSION, 
                                                dplyr::select(DESIGN,-starts_with('Donor_ID')),
                                                'all adjusted design(voom-normalized)',
                                                isKeyPlot=TRUE)
  
  # Add postadjusted covariates (if any)
  covariatesEffects = designMat2CovEffects(getCovariatesMap(DM),residualSigCovars$Effects.significantCovars)
  covariatesEffects = covariatesEffects[setdiff(names(covariatesEffects),postAdjustCovars$random)]
  
  postAdjustCovars$fixed = c(postAdjustCovars$fixed,names(which.max(covariatesEffects)))
  loopCount = loopCount + 1
}
tmp <- paste('Using following covariates in the final model', paste(paste(postAdjustCovars$fixed,collapse=','),'as fixed effects and', paste(postAdjustCovars$random,collapse=','),'as random effects'))
```

Store files in synapse

```{r synapsestore, echo=FALSE, include=FALSE, eval=TRUE}
if (SYNAPSE_STORE){  
  ActivityName <- 'Methylation Analysis of minfi data with mixed effects modeling'
  
  thisFileName <- 'MethylAnal_Test.Rmd'
  
  # Github link
  thisRepo <- getRepo(repository = "th1vairam/pcbc_c4_analysis", 
                      ref="branch", 
                      refName='diff_exp')

  thisFile <- getPermlink(repository = thisRepo,
                          repositoryPath=paste0('code/Rmd/', thisFileName))
  
  # Code
  CODE <- File('./MetylAnal_Test.Rmd',name = 'Methylation Analysis Mixed Effects',parentId = parentId)
  CODE <- synStore(CODE, used = ALL_USED_IDs, activityName = ActivityName, executed = thisFile)
}
```